{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Code preamble Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import requests\n",
    "import altair as alt\n",
    "from requests.adapters import HTTPAdapter\n",
    "from urllib3.util.retry import Retry"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Writing figure tex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_strings(test_str):\n",
    "    test_str = test_str.replace(\"_\",\"\")\n",
    "    for ele in test_str:\n",
    "        if ele.isdigit():\n",
    "            test_str = test_str.replace(ele,  num_to_word(ele))\n",
    "    return test_str\n",
    "def num_to_word(num):\n",
    "    n_dict = {0:'zero', 1:'one', 2:'two', 3:'three', 4:'four', 5:'five', 6:'six', \n",
    "              7:'seven',8:'eight', 9:'nine', 10:'ten', 11:'eleven', 12:'twelve',\n",
    "              13:'thirteen', 14:'fourteen', 15: 'fifteen'}\n",
    "    return n_dict[int(num)].title()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pathway_figure_tex(label, fname, score, filter):\n",
    "    list_str= [label, fname]\n",
    "    method, fname = [format_strings(i_str) for i_str in list_str]\n",
    "    score_lab = score.title().replace(\" \", \"\")\n",
    "    filter_lab = filter.title().replace(\" \", \"\")\n",
    "    beg_str = \"\\n \\\\newcommand{\\\\figurePathway\"+label+score_lab+filter_lab+\"}{\\\\begin{subfigure}[t]{\\doublefigurewidth} \\n\"\n",
    "    mid_str = \"\\includegraphics[width=\\linewidth] {\"+fname+\"} \\n\"\n",
    "    cap_str = \"\\caption{Pathway enrichment for the clustering method: \"+label+\". \\n Showing the \"+filter+\" based on the \"+score+\".\\label{fig::\"+label+score_lab+filter_lab+\"}}\"\n",
    "    end_str = \"\\end{subfigure} \\n} \\n\"\n",
    "    return(beg_str+mid_str+cap_str+end_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set the directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_dir_clone = \"../../PheWAS-cluster-main/working-example/data\"\n",
    "path_dir = \"../NealesLabData\"\n",
    "tex_dir = \"../../../Documents/Reports/Clustering\"\n",
    "path_plots = \"./Darrous_plots_all/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "clustering_info = pd.read_csv(path_dir+\"/cluster_labels.csv\", index_col=0)\n",
    "clustering_info.head()\n",
    "cluster_method_labels = list(clustering_info.index)\n",
    "cluster_method_labels.remove(\"SpectralProposed\")\n",
    "cluster_method_labels.remove(\"LloydRProposed\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>body_size</th>\n",
       "      <th>1697</th>\n",
       "      <th>1717</th>\n",
       "      <th>20015_irnt</th>\n",
       "      <th>2306</th>\n",
       "      <th>23101_irnt</th>\n",
       "      <th>23102_irnt</th>\n",
       "      <th>23105_irnt</th>\n",
       "      <th>23106_irnt</th>\n",
       "      <th>23107_irnt</th>\n",
       "      <th>...</th>\n",
       "      <th>KmeansCosineMinibatchSixProposed</th>\n",
       "      <th>KmedoidsCosineSixProposed</th>\n",
       "      <th>GMMZeroProposed</th>\n",
       "      <th>GMMOneProposed</th>\n",
       "      <th>BirchZeroProposed</th>\n",
       "      <th>BirchOneProposed</th>\n",
       "      <th>BirchTwoProposed</th>\n",
       "      <th>SpectralProposed</th>\n",
       "      <th>KmeansCosineTwoMinAICProposed</th>\n",
       "      <th>KmedoidsCosineThreeMinAICProposed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs115866895</th>\n",
       "      <td>-1.384558</td>\n",
       "      <td>0.078427</td>\n",
       "      <td>0.064179</td>\n",
       "      <td>1.313715</td>\n",
       "      <td>-0.390765</td>\n",
       "      <td>-1.428121</td>\n",
       "      <td>-1.395080</td>\n",
       "      <td>-1.765438</td>\n",
       "      <td>2.115427</td>\n",
       "      <td>1.865250</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>227</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs4648450</th>\n",
       "      <td>-1.280939</td>\n",
       "      <td>0.667544</td>\n",
       "      <td>0.025040</td>\n",
       "      <td>0.907814</td>\n",
       "      <td>0.041965</td>\n",
       "      <td>-2.116211</td>\n",
       "      <td>-2.090042</td>\n",
       "      <td>-2.321914</td>\n",
       "      <td>2.290245</td>\n",
       "      <td>1.904075</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>252</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs12024554</th>\n",
       "      <td>-1.758457</td>\n",
       "      <td>1.511510</td>\n",
       "      <td>0.407248</td>\n",
       "      <td>4.120851</td>\n",
       "      <td>0.602471</td>\n",
       "      <td>-0.676939</td>\n",
       "      <td>-0.701834</td>\n",
       "      <td>-0.828185</td>\n",
       "      <td>2.797708</td>\n",
       "      <td>2.548867</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1097327</th>\n",
       "      <td>-1.027856</td>\n",
       "      <td>-0.284528</td>\n",
       "      <td>-0.186588</td>\n",
       "      <td>1.438508</td>\n",
       "      <td>-0.030276</td>\n",
       "      <td>-1.195410</td>\n",
       "      <td>-1.167861</td>\n",
       "      <td>-1.501871</td>\n",
       "      <td>1.218832</td>\n",
       "      <td>1.538645</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>300</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs3737992</th>\n",
       "      <td>-0.226169</td>\n",
       "      <td>-1.735855</td>\n",
       "      <td>0.039532</td>\n",
       "      <td>-1.505157</td>\n",
       "      <td>-0.230216</td>\n",
       "      <td>-2.347974</td>\n",
       "      <td>-2.330429</td>\n",
       "      <td>-2.498627</td>\n",
       "      <td>1.603422</td>\n",
       "      <td>1.071926</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 433 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             body_size      1697      1717  20015_irnt      2306  23101_irnt  \\\n",
       "rs115866895  -1.384558  0.078427  0.064179    1.313715 -0.390765   -1.428121   \n",
       "rs4648450    -1.280939  0.667544  0.025040    0.907814  0.041965   -2.116211   \n",
       "rs12024554   -1.758457  1.511510  0.407248    4.120851  0.602471   -0.676939   \n",
       "rs1097327    -1.027856 -0.284528 -0.186588    1.438508 -0.030276   -1.195410   \n",
       "rs3737992    -0.226169 -1.735855  0.039532   -1.505157 -0.230216   -2.347974   \n",
       "\n",
       "             23102_irnt  23105_irnt  23106_irnt  23107_irnt  ...  \\\n",
       "rs115866895   -1.395080   -1.765438    2.115427    1.865250  ...   \n",
       "rs4648450     -2.090042   -2.321914    2.290245    1.904075  ...   \n",
       "rs12024554    -0.701834   -0.828185    2.797708    2.548867  ...   \n",
       "rs1097327     -1.167861   -1.501871    1.218832    1.538645  ...   \n",
       "rs3737992     -2.330429   -2.498627    1.603422    1.071926  ...   \n",
       "\n",
       "             KmeansCosineMinibatchSixProposed  KmedoidsCosineSixProposed  \\\n",
       "rs115866895                                 1                          3   \n",
       "rs4648450                                   1                          3   \n",
       "rs12024554                                  4                          3   \n",
       "rs1097327                                   4                          3   \n",
       "rs3737992                                   1                          1   \n",
       "\n",
       "             GMMZeroProposed  GMMOneProposed  BirchZeroProposed  \\\n",
       "rs115866895                1               4                227   \n",
       "rs4648450                  1               1                252   \n",
       "rs12024554                 0               0                253   \n",
       "rs1097327                  0               0                300   \n",
       "rs3737992                  1               4                195   \n",
       "\n",
       "             BirchOneProposed  BirchTwoProposed  SpectralProposed  \\\n",
       "rs115866895                 0                 0               NaN   \n",
       "rs4648450                   0                 0               NaN   \n",
       "rs12024554                  0                 0               NaN   \n",
       "rs1097327                   0                 0               NaN   \n",
       "rs3737992                   0                 0               NaN   \n",
       "\n",
       "             KmeansCosineTwoMinAICProposed  KmedoidsCosineThreeMinAICProposed  \n",
       "rs115866895                              1                                  1  \n",
       "rs4648450                                1                                  1  \n",
       "rs12024554                               1                                  1  \n",
       "rs1097327                                1                                  1  \n",
       "rs3737992                                1                                  1  \n",
       "\n",
       "[5 rows x 433 columns]"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clust_df = pd.read_csv(path_dir+\"/ours_clusters.csv\", index_col = 0)\n",
    "clust_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the PCAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc_ours_df = pd.read_csv(path_dir+\"/pcas_ours.csv\", index_col=0)\n",
    "pc_ours_with_exp_df = pd.read_csv(path_dir+\"/pcas_ours_with_exp.csv\", index_col=0)\n",
    "pc_pap_df = pd.read_csv(path_dir+\"/pcas_paper.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the Genes-Snps\n",
    "The `Genes` have been found using the function `SNP2GENE` from [FUMA](https://fuma.ctglab.nl/). This requires the `GWAS` data to be uploaded with the following compulsary columns:\n",
    "* Either Chromosome and Position or rsid.\n",
    "* Effect Allele\n",
    "* Sample size.\n",
    "This must be input as a txt file with space separated data and `\\n` separated rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensg</th>\n",
       "      <th>symbol</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>type</th>\n",
       "      <th>entrezID</th>\n",
       "      <th>HUGO</th>\n",
       "      <th>pLI</th>\n",
       "      <th>ncRVIS</th>\n",
       "      <th>posMapSNPs</th>\n",
       "      <th>posMapMaxCADD</th>\n",
       "      <th>minGwasP</th>\n",
       "      <th>IndSigSNPs</th>\n",
       "      <th>GenomicLocus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>1</td>\n",
       "      <td>1550795</td>\n",
       "      <td>1565990</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>142678.0</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>0.000159</td>\n",
       "      <td>0.030973</td>\n",
       "      <td>4</td>\n",
       "      <td>3.164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs115866895</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000189409</td>\n",
       "      <td>MMP23B</td>\n",
       "      <td>1</td>\n",
       "      <td>1567474</td>\n",
       "      <td>1570639</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8510.0</td>\n",
       "      <td>MMP23B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>3.164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs115866895</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000248333</td>\n",
       "      <td>CDK11B</td>\n",
       "      <td>1</td>\n",
       "      <td>1570603</td>\n",
       "      <td>1590473</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>728642.0</td>\n",
       "      <td>CDK11B</td>\n",
       "      <td>0.728294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34</td>\n",
       "      <td>11.560</td>\n",
       "      <td>1.617240e-15</td>\n",
       "      <td>rs115866895</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000189339</td>\n",
       "      <td>SLC35E2B</td>\n",
       "      <td>1</td>\n",
       "      <td>1592939</td>\n",
       "      <td>1624167</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>728661.0</td>\n",
       "      <td>SLC35E2B</td>\n",
       "      <td>0.621002</td>\n",
       "      <td>-0.794611</td>\n",
       "      <td>34</td>\n",
       "      <td>11.560</td>\n",
       "      <td>1.617240e-15</td>\n",
       "      <td>rs115866895</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000157873</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>1</td>\n",
       "      <td>2487078</td>\n",
       "      <td>2496821</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8764.0</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>0.819999</td>\n",
       "      <td>-0.131268</td>\n",
       "      <td>1</td>\n",
       "      <td>1.637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs4648450</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ensg    symbol  chr    start      end  strand            type  \\\n",
       "0  ENSG00000197530      MIB2    1  1550795  1565990       1  protein_coding   \n",
       "1  ENSG00000189409    MMP23B    1  1567474  1570639       1  protein_coding   \n",
       "2  ENSG00000248333    CDK11B    1  1570603  1590473      -1  protein_coding   \n",
       "3  ENSG00000189339  SLC35E2B    1  1592939  1624167      -1  protein_coding   \n",
       "4  ENSG00000157873  TNFRSF14    1  2487078  2496821       1  protein_coding   \n",
       "\n",
       "   entrezID      HUGO       pLI    ncRVIS  posMapSNPs  posMapMaxCADD  \\\n",
       "0  142678.0      MIB2  0.000159  0.030973           4          3.164   \n",
       "1    8510.0    MMP23B       NaN       NaN           5          3.164   \n",
       "2  728642.0    CDK11B  0.728294       NaN          34         11.560   \n",
       "3  728661.0  SLC35E2B  0.621002 -0.794611          34         11.560   \n",
       "4    8764.0  TNFRSF14  0.819999 -0.131268           1          1.637   \n",
       "\n",
       "       minGwasP   IndSigSNPs GenomicLocus  \n",
       "0           NaN  rs115866895            1  \n",
       "1           NaN  rs115866895            1  \n",
       "2  1.617240e-15  rs115866895            1  \n",
       "3  1.617240e-15  rs115866895            1  \n",
       "4           NaN    rs4648450            2  "
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_df = pd.read_csv(path_dir+'/genes.txt', sep=\"\\t\", header=0)\n",
    "genes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_snps = genes_df[[\"symbol\", \"IndSigSNPs\"]]\n",
    "gene_snps.value_counts()\n",
    "full_gene_list = np.unique(gene_snps[\"symbol\"].values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct the Clusters in terms of Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KmeansLloydStdFourProposed</th>\n",
       "      <th>KmeansLloydStdFourDarrous</th>\n",
       "      <th>KmeansLloydStdSixProposed</th>\n",
       "      <th>KmeansLloydStdSixDarrous</th>\n",
       "      <th>KmeansElkanStdFourProposed</th>\n",
       "      <th>KmeansElkanStdFourDarrous</th>\n",
       "      <th>KmedoidsEuclideanProposedFour</th>\n",
       "      <th>KmeansCosineSixProposed</th>\n",
       "      <th>DBScanZeroProposed</th>\n",
       "      <th>DBScanOneProposed</th>\n",
       "      <th>DBScanTwoProposed</th>\n",
       "      <th>KmeansCosineMinibatchSixProposed</th>\n",
       "      <th>KmedoidsCosineSixProposed</th>\n",
       "      <th>GMMZeroProposed</th>\n",
       "      <th>GMMOneProposed</th>\n",
       "      <th>BirchZeroProposed</th>\n",
       "      <th>BirchOneProposed</th>\n",
       "      <th>BirchTwoProposed</th>\n",
       "      <th>KmeansCosineTwoMinAICProposed</th>\n",
       "      <th>KmedoidsCosineThreeMinAICProposed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs115866895</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>227</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs4648450</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>252</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs12024554</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>253</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1097327</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>300</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs3737992</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>195</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs8134638</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>43</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2837398</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>160</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs76040172</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>251</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs6001870</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>323</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs28489620</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>194</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>324 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             KmeansLloydStdFourProposed  KmeansLloydStdFourDarrous  \\\n",
       "rs115866895                           1                          1   \n",
       "rs4648450                             3                          2   \n",
       "rs12024554                            1                          0   \n",
       "rs1097327                             3                          3   \n",
       "rs3737992                             1                          2   \n",
       "...                                 ...                        ...   \n",
       "rs8134638                             0                          0   \n",
       "rs2837398                             0                          0   \n",
       "rs76040172                            3                          1   \n",
       "rs6001870                             3                          2   \n",
       "rs28489620                            3                          2   \n",
       "\n",
       "             KmeansLloydStdSixProposed  KmeansLloydStdSixDarrous  \\\n",
       "rs115866895                          2                         4   \n",
       "rs4648450                            1                         0   \n",
       "rs12024554                           2                         4   \n",
       "rs1097327                            4                         3   \n",
       "rs3737992                            1                         0   \n",
       "...                                ...                       ...   \n",
       "rs8134638                            0                         4   \n",
       "rs2837398                            0                         4   \n",
       "rs76040172                           4                         5   \n",
       "rs6001870                            1                         0   \n",
       "rs28489620                           1                         2   \n",
       "\n",
       "             KmeansElkanStdFourProposed  KmeansElkanStdFourDarrous  \\\n",
       "rs115866895                           1                          1   \n",
       "rs4648450                             3                          2   \n",
       "rs12024554                            1                          0   \n",
       "rs1097327                             3                          3   \n",
       "rs3737992                             1                          2   \n",
       "...                                 ...                        ...   \n",
       "rs8134638                             0                          0   \n",
       "rs2837398                             0                          0   \n",
       "rs76040172                            3                          1   \n",
       "rs6001870                             3                          2   \n",
       "rs28489620                            3                          2   \n",
       "\n",
       "             KmedoidsEuclideanProposedFour  KmeansCosineSixProposed  \\\n",
       "rs115866895                              0                        1   \n",
       "rs4648450                                0                        1   \n",
       "rs12024554                               0                        2   \n",
       "rs1097327                                0                        3   \n",
       "rs3737992                                3                        1   \n",
       "...                                    ...                      ...   \n",
       "rs8134638                                2                        4   \n",
       "rs2837398                                1                        4   \n",
       "rs76040172                               0                        1   \n",
       "rs6001870                                3                        3   \n",
       "rs28489620                               0                        1   \n",
       "\n",
       "             DBScanZeroProposed  DBScanOneProposed  DBScanTwoProposed  \\\n",
       "rs115866895                   0                 -1                 -1   \n",
       "rs4648450                    -1                 -1                 -1   \n",
       "rs12024554                   -1                 -1                 -1   \n",
       "rs1097327                    -1                 -1                 -1   \n",
       "rs3737992                     0                 -1                 -1   \n",
       "...                         ...                ...                ...   \n",
       "rs8134638                    -1                 -1                 -1   \n",
       "rs2837398                    -1                 -1                 -1   \n",
       "rs76040172                    0                 -1                 -1   \n",
       "rs6001870                     0                 -1                 -1   \n",
       "rs28489620                   -1                 -1                 -1   \n",
       "\n",
       "             KmeansCosineMinibatchSixProposed  KmedoidsCosineSixProposed  \\\n",
       "rs115866895                                 1                          3   \n",
       "rs4648450                                   1                          3   \n",
       "rs12024554                                  4                          3   \n",
       "rs1097327                                   4                          3   \n",
       "rs3737992                                   1                          1   \n",
       "...                                       ...                        ...   \n",
       "rs8134638                                   2                          5   \n",
       "rs2837398                                   2                          4   \n",
       "rs76040172                                  1                          3   \n",
       "rs6001870                                   4                          1   \n",
       "rs28489620                                  4                          3   \n",
       "\n",
       "             GMMZeroProposed  GMMOneProposed  BirchZeroProposed  \\\n",
       "rs115866895                1               4                227   \n",
       "rs4648450                  1               1                252   \n",
       "rs12024554                 0               0                253   \n",
       "rs1097327                  0               0                300   \n",
       "rs3737992                  1               4                195   \n",
       "...                      ...             ...                ...   \n",
       "rs8134638                  2               2                 43   \n",
       "rs2837398                  2               2                160   \n",
       "rs76040172                 1               4                251   \n",
       "rs6001870                  0               3                323   \n",
       "rs28489620                 1               1                194   \n",
       "\n",
       "             BirchOneProposed  BirchTwoProposed  \\\n",
       "rs115866895                 0                 0   \n",
       "rs4648450                   0                 0   \n",
       "rs12024554                  0                 0   \n",
       "rs1097327                   0                 0   \n",
       "rs3737992                   0                 0   \n",
       "...                       ...               ...   \n",
       "rs8134638                   1                 1   \n",
       "rs2837398                   1                 1   \n",
       "rs76040172                  0                 0   \n",
       "rs6001870                   0                 0   \n",
       "rs28489620                  0                 0   \n",
       "\n",
       "             KmeansCosineTwoMinAICProposed  KmedoidsCosineThreeMinAICProposed  \n",
       "rs115866895                              1                                  1  \n",
       "rs4648450                                1                                  1  \n",
       "rs12024554                               1                                  1  \n",
       "rs1097327                                1                                  1  \n",
       "rs3737992                                1                                  1  \n",
       "...                                    ...                                ...  \n",
       "rs8134638                                0                                  2  \n",
       "rs2837398                                0                                  2  \n",
       "rs76040172                               1                                  1  \n",
       "rs6001870                                1                                  1  \n",
       "rs28489620                               1                                  1  \n",
       "\n",
       "[324 rows x 20 columns]"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_clusters_df = clust_df[clust_df.columns[clust_df.columns.isin(cluster_method_labels)]]\n",
    "cluster_method_labels = [lab for lab in cluster_method_labels if lab in clust_df.columns]\n",
    "gene_clust_df = pd.DataFrame(columns = cluster_method_labels)\n",
    "snp_clusters_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_clust_dict = {} #pd.DataFrame(index = genes_df.symbol.values)\n",
    "for clust_lab in cluster_method_labels:\n",
    "    # Get the clusters by snps for cluster method clust_lab\n",
    "    clust_single = snp_clusters_df.loc[:, clust_lab]\n",
    "    # Get the possible cluster numbers\n",
    "    clust_nums = clust_single.unique()\n",
    "    gene_clust_single_df = pd.DataFrame(\n",
    "        columns = [clust_lab+\"_\"+str(c_n) for c_n in clust_nums]\n",
    "    )\n",
    "    for c_num in clust_nums:\n",
    "        col_lab = clust_lab+\"_\"+str(c_num)\n",
    "        # snps matching clusters\n",
    "        clust_snps = list(clust_df[clust_single == c_num].index)\n",
    "        # genes prevent in those snps\n",
    "        genes = list(genes_df[genes_df[\"IndSigSNPs\"].isin(clust_snps)].symbol.unique())\n",
    "        # genes marked with 1 if in cluster 0 if not.\n",
    "        gene_clust_c_num_df = pd.DataFrame(\n",
    "            index = genes,\n",
    "            data = {\n",
    "                col_lab: 1\n",
    "            }\n",
    "        )\n",
    "        extra_genes = list(np.setdiff1d(full_gene_list, genes))\n",
    "        gene_clust_not_in_df = pd.DataFrame(\n",
    "            index = extra_genes,\n",
    "            data = {\n",
    "                col_lab: 0\n",
    "            }\n",
    "        )\n",
    "        gene_clust_c_num_df = pd.concat([gene_clust_c_num_df, gene_clust_not_in_df],\n",
    "                                        axis = 0\n",
    "        )\n",
    "        gene_clust_single_df[col_lab] = gene_clust_c_num_df\n",
    "    gene_clust_dict[clust_lab] = gene_clust_single_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensg</th>\n",
       "      <th>symbol</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>type</th>\n",
       "      <th>entrezID</th>\n",
       "      <th>HUGO</th>\n",
       "      <th>pLI</th>\n",
       "      <th>ncRVIS</th>\n",
       "      <th>posMapSNPs</th>\n",
       "      <th>posMapMaxCADD</th>\n",
       "      <th>minGwasP</th>\n",
       "      <th>IndSigSNPs</th>\n",
       "      <th>GenomicLocus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000197530</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>1</td>\n",
       "      <td>1550795</td>\n",
       "      <td>1565990</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>142678.0</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>1.585433e-04</td>\n",
       "      <td>0.030973</td>\n",
       "      <td>4</td>\n",
       "      <td>3.164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs115866895</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000189409</td>\n",
       "      <td>MMP23B</td>\n",
       "      <td>1</td>\n",
       "      <td>1567474</td>\n",
       "      <td>1570639</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8510.0</td>\n",
       "      <td>MMP23B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>3.164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs115866895</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000248333</td>\n",
       "      <td>CDK11B</td>\n",
       "      <td>1</td>\n",
       "      <td>1570603</td>\n",
       "      <td>1590473</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>728642.0</td>\n",
       "      <td>CDK11B</td>\n",
       "      <td>7.282944e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34</td>\n",
       "      <td>11.560</td>\n",
       "      <td>1.617240e-15</td>\n",
       "      <td>rs115866895</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000189339</td>\n",
       "      <td>SLC35E2B</td>\n",
       "      <td>1</td>\n",
       "      <td>1592939</td>\n",
       "      <td>1624167</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>728661.0</td>\n",
       "      <td>SLC35E2B</td>\n",
       "      <td>6.210025e-01</td>\n",
       "      <td>-0.794611</td>\n",
       "      <td>34</td>\n",
       "      <td>11.560</td>\n",
       "      <td>1.617240e-15</td>\n",
       "      <td>rs115866895</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000157873</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>1</td>\n",
       "      <td>2487078</td>\n",
       "      <td>2496821</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>8764.0</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>8.199991e-01</td>\n",
       "      <td>-0.131268</td>\n",
       "      <td>1</td>\n",
       "      <td>1.637</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs4648450</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>ENSG00000196419</td>\n",
       "      <td>XRCC6</td>\n",
       "      <td>22</td>\n",
       "      <td>42017123</td>\n",
       "      <td>42060044</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2547.0</td>\n",
       "      <td>XRCC6</td>\n",
       "      <td>9.968559e-01</td>\n",
       "      <td>-0.128101</td>\n",
       "      <td>2</td>\n",
       "      <td>6.981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs28489620</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>ENSG00000100138</td>\n",
       "      <td>NHP2L1</td>\n",
       "      <td>22</td>\n",
       "      <td>42069934</td>\n",
       "      <td>42086508</td>\n",
       "      <td>-1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4809.0</td>\n",
       "      <td>NHP2L1</td>\n",
       "      <td>7.105853e-01</td>\n",
       "      <td>0.082226</td>\n",
       "      <td>2</td>\n",
       "      <td>6.981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs28489620</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>ENSG00000167077</td>\n",
       "      <td>MEI1</td>\n",
       "      <td>22</td>\n",
       "      <td>42095503</td>\n",
       "      <td>42195460</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>150365.0</td>\n",
       "      <td>MEI1</td>\n",
       "      <td>1.215924e-08</td>\n",
       "      <td>0.500155</td>\n",
       "      <td>13</td>\n",
       "      <td>8.496</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs28489620</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>ENSG00000100147</td>\n",
       "      <td>CCDC134</td>\n",
       "      <td>22</td>\n",
       "      <td>42196683</td>\n",
       "      <td>42222303</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>79879.0</td>\n",
       "      <td>CCDC134</td>\n",
       "      <td>7.411728e-04</td>\n",
       "      <td>0.324960</td>\n",
       "      <td>14</td>\n",
       "      <td>8.496</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs28489620</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>ENSG00000198911</td>\n",
       "      <td>SREBF2</td>\n",
       "      <td>22</td>\n",
       "      <td>42229109</td>\n",
       "      <td>42303312</td>\n",
       "      <td>1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6721.0</td>\n",
       "      <td>SREBF2</td>\n",
       "      <td>9.962716e-01</td>\n",
       "      <td>0.410467</td>\n",
       "      <td>11</td>\n",
       "      <td>8.328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs28489620</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>597 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                ensg    symbol  chr     start       end  strand  \\\n",
       "0    ENSG00000197530      MIB2    1   1550795   1565990       1   \n",
       "1    ENSG00000189409    MMP23B    1   1567474   1570639       1   \n",
       "2    ENSG00000248333    CDK11B    1   1570603   1590473      -1   \n",
       "3    ENSG00000189339  SLC35E2B    1   1592939   1624167      -1   \n",
       "4    ENSG00000157873  TNFRSF14    1   2487078   2496821       1   \n",
       "..               ...       ...  ...       ...       ...     ...   \n",
       "592  ENSG00000196419     XRCC6   22  42017123  42060044       1   \n",
       "593  ENSG00000100138    NHP2L1   22  42069934  42086508      -1   \n",
       "594  ENSG00000167077      MEI1   22  42095503  42195460       1   \n",
       "595  ENSG00000100147   CCDC134   22  42196683  42222303       1   \n",
       "596  ENSG00000198911    SREBF2   22  42229109  42303312       1   \n",
       "\n",
       "               type  entrezID      HUGO           pLI    ncRVIS  posMapSNPs  \\\n",
       "0    protein_coding  142678.0      MIB2  1.585433e-04  0.030973           4   \n",
       "1    protein_coding    8510.0    MMP23B           NaN       NaN           5   \n",
       "2    protein_coding  728642.0    CDK11B  7.282944e-01       NaN          34   \n",
       "3    protein_coding  728661.0  SLC35E2B  6.210025e-01 -0.794611          34   \n",
       "4    protein_coding    8764.0  TNFRSF14  8.199991e-01 -0.131268           1   \n",
       "..              ...       ...       ...           ...       ...         ...   \n",
       "592  protein_coding    2547.0     XRCC6  9.968559e-01 -0.128101           2   \n",
       "593  protein_coding    4809.0    NHP2L1  7.105853e-01  0.082226           2   \n",
       "594  protein_coding  150365.0      MEI1  1.215924e-08  0.500155          13   \n",
       "595  protein_coding   79879.0   CCDC134  7.411728e-04  0.324960          14   \n",
       "596  protein_coding    6721.0    SREBF2  9.962716e-01  0.410467          11   \n",
       "\n",
       "     posMapMaxCADD      minGwasP   IndSigSNPs GenomicLocus  \n",
       "0            3.164           NaN  rs115866895            1  \n",
       "1            3.164           NaN  rs115866895            1  \n",
       "2           11.560  1.617240e-15  rs115866895            1  \n",
       "3           11.560  1.617240e-15  rs115866895            1  \n",
       "4            1.637           NaN    rs4648450            2  \n",
       "..             ...           ...          ...          ...  \n",
       "592          6.981           NaN   rs28489620          314  \n",
       "593          6.981           NaN   rs28489620          314  \n",
       "594          8.496           NaN   rs28489620          314  \n",
       "595          8.496           NaN   rs28489620          314  \n",
       "596          8.328           NaN   rs28489620          314  \n",
       "\n",
       "[597 rows x 16 columns]"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for lab in gene_clust_dict:\n",
    "    gene_clust_dict[lab].to_csv(path_dir+\"/gene_clusters_\" + lab + \".csv\",\n",
    "                     header = True,\n",
    "                     index= True,\n",
    "                     sep = \",\")\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct the PCAs in terms of Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pca_dict = {} #pd.DataFrame(index = genes_df.symbol.values)\n",
    "# Get the possible PCAs\n",
    "pc_gene_ours_df = genes_df.loc[:,[\"IndSigSNPs\",\"symbol\"]]\n",
    "pc_gene_ours_with_exp_df = genes_df.loc[:,[\"IndSigSNPs\",\"symbol\"]]\n",
    "pc_gene_pap_df = genes_df.loc[:,[\"IndSigSNPs\",\"symbol\"]]\n",
    "pc_gene_dict = {\"pap\": pc_gene_pap_df,\n",
    "                \"ours\": pc_gene_ours_df,\n",
    "                \"ours_with_exp\": pc_gene_ours_with_exp_df}\n",
    "pc_dict = {\"pap\": pc_pap_df,\n",
    "            \"ours\": pc_ours_df,\n",
    "            \"ours_with_exp\": pc_ours_with_exp_df}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>IndSigSNPs</th>\n",
       "      <th>symbol</th>\n",
       "      <th>pc_1</th>\n",
       "      <th>pc_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs115866895</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>-6.436117</td>\n",
       "      <td>0.075144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs115866895</td>\n",
       "      <td>MMP23B</td>\n",
       "      <td>-6.436117</td>\n",
       "      <td>0.075144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs115866895</td>\n",
       "      <td>CDK11B</td>\n",
       "      <td>-6.436117</td>\n",
       "      <td>0.075144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs115866895</td>\n",
       "      <td>SLC35E2B</td>\n",
       "      <td>-6.436117</td>\n",
       "      <td>0.075144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4648450</td>\n",
       "      <td>TNFRSF14</td>\n",
       "      <td>-10.560462</td>\n",
       "      <td>-1.307634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>rs28489620</td>\n",
       "      <td>XRCC6</td>\n",
       "      <td>-6.469766</td>\n",
       "      <td>7.800513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>rs28489620</td>\n",
       "      <td>NHP2L1</td>\n",
       "      <td>-6.469766</td>\n",
       "      <td>7.800513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>rs28489620</td>\n",
       "      <td>MEI1</td>\n",
       "      <td>-6.469766</td>\n",
       "      <td>7.800513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>595</th>\n",
       "      <td>rs28489620</td>\n",
       "      <td>CCDC134</td>\n",
       "      <td>-6.469766</td>\n",
       "      <td>7.800513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>rs28489620</td>\n",
       "      <td>SREBF2</td>\n",
       "      <td>-6.469766</td>\n",
       "      <td>7.800513</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>597 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      IndSigSNPs    symbol       pc_1      pc_2\n",
       "0    rs115866895      MIB2  -6.436117  0.075144\n",
       "1    rs115866895    MMP23B  -6.436117  0.075144\n",
       "2    rs115866895    CDK11B  -6.436117  0.075144\n",
       "3    rs115866895  SLC35E2B  -6.436117  0.075144\n",
       "4      rs4648450  TNFRSF14 -10.560462 -1.307634\n",
       "..           ...       ...        ...       ...\n",
       "592   rs28489620     XRCC6  -6.469766  7.800513\n",
       "593   rs28489620    NHP2L1  -6.469766  7.800513\n",
       "594   rs28489620      MEI1  -6.469766  7.800513\n",
       "595   rs28489620   CCDC134  -6.469766  7.800513\n",
       "596   rs28489620    SREBF2  -6.469766  7.800513\n",
       "\n",
       "[597 rows x 4 columns]"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for lab in pc_gene_dict.keys():\n",
    "   pc_gene_df = pc_gene_dict[lab]\n",
    "   pc_df = pc_dict[lab]\n",
    "   pc_gene_df[pc_df.columns] = 0.0\n",
    "   for snp in pc_gene_df.IndSigSNPs:\n",
    "      if \";\" in snp:\n",
    "         snp_list = list(snp.split(\";\"))\n",
    "      else: snp_list = [snp]\n",
    "      for pc_col in pc_df.columns:\n",
    "         for snp in snp_list:\n",
    "            gene_rows = list(pc_gene_df[pc_gene_df[\"IndSigSNPs\"]==snp].index)\n",
    "            pc_gene_df.loc[gene_rows, pc_col] = pc_df.loc[snp, pc_col]\n",
    "pc_gene_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save the gene list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [],
   "source": [
    "pc_gene_df.symbol.to_csv(path_dir+\"/gene_list.csv\", index = False, header=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load the pathways found on reactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1346"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reactome_pathways = pd.read_csv(path_dir+\"/reactome_pathways.csv\")\n",
    "reactome_pathways[\"Submitted entities found\"] = reactome_pathways[\"Submitted entities found\"].str.split(\";\")\n",
    "#reactome_pathways = reactome_pathways.explode(\"Submitted entities found\")\n",
    "len(reactome_pathways[\"Pathway name\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Group the genes by association\n",
    "In order to get pathway enrichment we need to be able to group the genes. For each principal componene group the genes by their association score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [],
   "source": [
    "for lab in pc_gene_dict.keys():\n",
    "   pc_gene_df = pc_gene_dict[lab]\n",
    "   pc_df = pc_dict[lab]\n",
    "   for pc_col in pc_df.columns:\n",
    "     groups = pd.cut(pc_gene_df[pc_col],\n",
    "                     bins = 10,\n",
    "                     labels = [pc_col+\"_%d\"%i for i in range(10)])\n",
    "     groups_ranges = pd.cut(pc_gene_df[pc_col],\n",
    "                     bins = 10)\n",
    "     pc_gene_df[pc_col+\"_bin\"] = groups\n",
    "     pc_gene_df[pc_col+\"_bin_ranges\"] = groups_ranges\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Restructure the dataframe to have the bin labels as columns and genes as the values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pap':            mid           interval\n",
       " pc_1_4 -2.3055    (-3.271, -1.34]\n",
       " pc_1_6  1.5555      (0.59, 2.521]\n",
       " pc_1_3  -4.236   (-5.201, -3.271]\n",
       " pc_1_7   3.486     (2.521, 4.451]\n",
       " pc_1_0 -10.037  (-11.012, -9.062]\n",
       " pc_1_8  5.4165     (4.451, 6.382]\n",
       " pc_1_5  -0.375      (-1.34, 0.59]\n",
       " pc_1_9   7.347     (6.382, 8.312]\n",
       " pc_1_2 -6.1665   (-7.132, -5.201]\n",
       " pc_1_1  -8.097   (-9.062, -7.132]\n",
       " pc_2_6  3.4175     (2.357, 4.478]\n",
       " pc_2_2 -5.0655   (-6.126, -4.005]\n",
       " pc_2_5  1.2965     (0.236, 2.357]\n",
       " pc_2_3  -2.945   (-4.005, -1.885]\n",
       " pc_2_4 -0.8245    (-1.885, 0.236]\n",
       " pc_2_1 -7.1865   (-8.247, -6.126]\n",
       " pc_2_7   5.538     (4.478, 6.598]\n",
       " pc_2_8  7.6585     (6.598, 8.719]\n",
       " pc_2_0 -9.3175  (-10.388, -8.247]\n",
       " pc_2_9  9.7795     (8.719, 10.84],\n",
       " 'ours':             mid            interval\n",
       " pc_1_7    7.346      (5.881, 8.811]\n",
       " pc_1_8  10.2765     (8.811, 11.742]\n",
       " pc_1_2  -7.3075    (-8.773, -5.842]\n",
       " pc_1_4 -1.44645    (-2.912, 0.0191]\n",
       " pc_1_9  13.2075    (11.742, 14.673]\n",
       " pc_1_0  -13.184  (-14.664, -11.704]\n",
       " pc_1_1 -10.2385   (-11.704, -8.773]\n",
       " pc_1_6   4.4155       (2.95, 5.881]\n",
       " pc_1_3   -4.377    (-5.842, -2.912]\n",
       " pc_1_5  1.48455      (0.0191, 2.95]\n",
       " pc_2_5   0.3215      (-0.66, 1.303]\n",
       " pc_2_6    2.284      (1.303, 3.265]\n",
       " pc_2_1  -7.5275    (-8.509, -6.546]\n",
       " pc_2_4   -1.641     (-2.622, -0.66]\n",
       " pc_2_2   -5.565    (-6.546, -4.584]\n",
       " pc_2_7    4.246      (3.265, 5.227]\n",
       " pc_2_3   -3.603    (-4.584, -2.622]\n",
       " pc_2_8   6.2085       (5.227, 7.19]\n",
       " pc_2_0     -9.5   (-10.491, -8.509]\n",
       " pc_2_9    8.171       (7.19, 9.152],\n",
       " 'ours_with_exp':             mid            interval\n",
       " pc_1_2   -7.309     (-8.77, -5.848]\n",
       " pc_1_1  -10.231    (-11.692, -8.77]\n",
       " pc_1_7    7.301       (5.84, 8.762]\n",
       " pc_1_5  1.45707   (-0.00386, 2.918]\n",
       " pc_1_0 -13.1675  (-14.643, -11.692]\n",
       " pc_1_9   13.145    (11.684, 14.606]\n",
       " pc_1_8   10.223     (8.762, 11.684]\n",
       " pc_1_3   -4.387    (-5.848, -2.926]\n",
       " pc_1_6    4.379       (2.918, 5.84]\n",
       " pc_1_4 -1.46493  (-2.926, -0.00386]\n",
       " pc_2_4  -0.3255     (-1.314, 0.663]\n",
       " pc_2_3  -2.3025    (-3.291, -1.314]\n",
       " pc_2_8   7.5825      (6.594, 8.571]\n",
       " pc_2_5   1.6515       (0.663, 2.64]\n",
       " pc_2_7   5.6055      (4.617, 6.594]\n",
       " pc_2_2  -4.2795    (-5.268, -3.291]\n",
       " pc_2_6   3.6285       (2.64, 4.617]\n",
       " pc_2_1  -6.2565    (-7.245, -5.268]\n",
       " pc_2_9   9.5595     (8.571, 10.548]\n",
       " pc_2_0  -8.2435    (-9.242, -7.245]}"
      ]
     },
     "execution_count": 258,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pc_gene_groups_dict = {}\n",
    "bin_range_dict = {}\n",
    "for lab in pc_gene_dict.keys():\n",
    "   pc_gene_df = pc_gene_dict[lab]\n",
    "   bin_groups = {}\n",
    "   bin_ranges = {}\n",
    "   for pc_col in pc_df.columns:\n",
    "      for bin in pc_gene_df[pc_col +\"_bin\"].unique():\n",
    "         genes = pc_gene_df.loc[pc_gene_df[pc_col+\"_bin\"]==bin,\"symbol\"]\n",
    "         bin_groups[bin] = genes\n",
    "         bin_ranges[bin] = {\"mid\": \n",
    "                            pc_gene_df.loc[pc_gene_df[pc_col+\"_bin\"]==bin,pc_col+\"_bin_ranges\"].values[0].mid,\n",
    "                            \"interval\": pc_gene_df.loc[pc_gene_df[pc_col+\"_bin\"]==bin,pc_col+\"_bin_ranges\"].values[0]\n",
    "         }\n",
    "   pc_gene_groups_dict[lab] = pd.DataFrame(bin_groups)\n",
    "   bin_range_dict[lab] = pd.DataFrame(bin_ranges).T\n",
    "   #print(pc_gene_groups_dict[lab])\n",
    "bin_range_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ENRICHR\n",
    "### Get the pathway association for gene sets in each cluster.\n",
    "Using the ENRICHR api get the pathways associated with the sets of SNPS in each cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KmeansLloydStdFourDarrous\n",
      "KmeansLloydStdSixDarrous\n",
      "KmeansElkanStdFourDarrous\n",
      "KmeansLloydStdFourProposed\n",
      "KmeansLloydStdSixProposed\n",
      "KmeansElkanStdFourProposed\n",
      "KmedoidsEuclideanProposedFour\n",
      "KmeansCosineSixProposed\n",
      "DBScanZeroProposed\n",
      "DBScanOneProposed\n",
      "DBScanTwoProposed\n",
      "KmeansCosineMinibatchSixProposed\n",
      "KmedoidsCosineSixProposed\n",
      "GMMZeroProposed\n",
      "GMMOneProposed\n",
      "BirchZeroProposed\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "Error analyzing gene list\n",
      "BirchOneProposed\n",
      "BirchTwoProposed\n",
      "KmeansCosineTwoMinAICProposed\n",
      "KmedoidsCosineThreeMinAICProposed\n",
      "{'KmeansLloydStdFourDarrous':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  https://science.sciencemag.org/content/365/645...   \n",
      "2                  1  10 percent of the longest transcripts of the h...   \n",
      "3                  1                                             ct1346   \n",
      "4                  1                                       inflammation   \n",
      "...              ...                                                ...   \n",
      "3676               3      Blue module gene list CRC (Normal and Cancer)   \n",
      "3677               3  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "3678               3  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3679               3                    MEF-iPS diff ES DNA methylation   \n",
      "3680               3                                     pred all genes   \n",
      "\n",
      "      combined_score  \n",
      "0        1570.396478  \n",
      "1          78.826549  \n",
      "2          64.018741  \n",
      "3          31.637536  \n",
      "4         155.274597  \n",
      "...              ...  \n",
      "3676        0.000647  \n",
      "3677        0.000866  \n",
      "3678        0.000412  \n",
      "3679        0.000056  \n",
      "3680        0.000179  \n",
      "\n",
      "[3681 rows x 3 columns], 'KmeansLloydStdSixDarrous':       cluster_number                                            pathway  \\\n",
      "0                  4  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  4  https://science.sciencemag.org/content/365/645...   \n",
      "2                  4                                 az mlm p <.01+ WES   \n",
      "3                  4         Genes that were expressed and up regulated   \n",
      "4                  4                            microRNA in tubercuosis   \n",
      "...              ...                                                ...   \n",
      "4887               1  downregulated, dataset: 6d5ee3258f0d93c3f9f758...   \n",
      "4888               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "4889               1                                         down in VS   \n",
      "4890               1                      Upregulated genes in WT brain   \n",
      "4891               1  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "\n",
      "      combined_score  \n",
      "0         249.540670  \n",
      "1          77.589755  \n",
      "2          53.588404  \n",
      "3          34.728597  \n",
      "4          96.147289  \n",
      "...              ...  \n",
      "4887        0.001711  \n",
      "4888        0.001356  \n",
      "4889        0.002198  \n",
      "4890        0.000217  \n",
      "4891        0.000090  \n",
      "\n",
      "[4892 rows x 3 columns], 'KmeansElkanStdFourDarrous':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  https://science.sciencemag.org/content/365/645...   \n",
      "2                  1  10 percent of the longest transcripts of the h...   \n",
      "3                  1                                             ct1346   \n",
      "4                  1                                       inflammation   \n",
      "...              ...                                                ...   \n",
      "3676               3      Blue module gene list CRC (Normal and Cancer)   \n",
      "3677               3  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "3678               3  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3679               3                    MEF-iPS diff ES DNA methylation   \n",
      "3680               3                                     pred all genes   \n",
      "\n",
      "      combined_score  \n",
      "0        1570.396478  \n",
      "1          78.826549  \n",
      "2          64.018741  \n",
      "3          31.637536  \n",
      "4         155.274597  \n",
      "...              ...  \n",
      "3676        0.000647  \n",
      "3677        0.000866  \n",
      "3678        0.000412  \n",
      "3679        0.000056  \n",
      "3680        0.000179  \n",
      "\n",
      "[3681 rows x 3 columns], 'KmeansLloydStdFourProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1                                       inflammation   \n",
      "2                  1         Genes that were expressed and up regulated   \n",
      "3                  1  10 percent of the longest transcripts of the h...   \n",
      "4                  1                                      wound healing   \n",
      "...              ...                                                ...   \n",
      "3728               2  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3729               2                                         harvard hw   \n",
      "3730               2                                     pred all genes   \n",
      "3731               2                    MEF-iPS diff ES DNA methylation   \n",
      "3732               2  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      combined_score  \n",
      "0        1154.462333  \n",
      "1         441.018853  \n",
      "2          27.891232  \n",
      "3          21.645860  \n",
      "4          52.483056  \n",
      "...              ...  \n",
      "3728        0.002268  \n",
      "3729        0.000202  \n",
      "3730        0.000363  \n",
      "3731        0.000014  \n",
      "3732        0.000018  \n",
      "\n",
      "[3733 rows x 3 columns], 'KmeansLloydStdSixProposed':       cluster_number                                            pathway  \\\n",
      "0                  2  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  2                                       inflammation   \n",
      "2                  2  10 percent of the longest transcripts of the h...   \n",
      "3                  2         Genes that were expressed and up regulated   \n",
      "4                  2                                      wound healing   \n",
      "...              ...                                                ...   \n",
      "4851               5                                        Sca1+ genes   \n",
      "4852               5                                             Cn1346   \n",
      "4853               5  Meta-analysis of autosomal sex DE genes across...   \n",
      "4854               5  Turquise module gene list CRC (Normal and Cancer)   \n",
      "4855               5                                     pred all genes   \n",
      "\n",
      "      combined_score  \n",
      "0         949.152361  \n",
      "1         543.341293  \n",
      "2          21.411208  \n",
      "3          26.387780  \n",
      "4          66.847971  \n",
      "...              ...  \n",
      "4851        0.003640  \n",
      "4852        0.003683  \n",
      "4853        0.001738  \n",
      "4854        0.000622  \n",
      "4855        0.000752  \n",
      "\n",
      "[4856 rows x 3 columns], 'KmeansElkanStdFourProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1                                       inflammation   \n",
      "2                  1         Genes that were expressed and up regulated   \n",
      "3                  1  10 percent of the longest transcripts of the h...   \n",
      "4                  1                                      wound healing   \n",
      "...              ...                                                ...   \n",
      "3728               2  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3729               2                                         harvard hw   \n",
      "3730               2                                     pred all genes   \n",
      "3731               2                    MEF-iPS diff ES DNA methylation   \n",
      "3732               2  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      combined_score  \n",
      "0        1154.462333  \n",
      "1         441.018853  \n",
      "2          27.891232  \n",
      "3          21.645860  \n",
      "4          52.483056  \n",
      "...              ...  \n",
      "3728        0.002268  \n",
      "3729        0.000202  \n",
      "3730        0.000363  \n",
      "3731        0.000014  \n",
      "3732        0.000018  \n",
      "\n",
      "[3733 rows x 3 columns], 'KmedoidsEuclideanProposedFour':       cluster_number                                            pathway  \\\n",
      "0                  0  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  0                                       inflammation   \n",
      "2                  0  10 percent of the longest transcripts of the h...   \n",
      "3                  0                                        Gdgtdrhyjgk   \n",
      "4                  0                 viral chromatin associated factors   \n",
      "...              ...                                                ...   \n",
      "3706               2                                   wte deg analysis   \n",
      "3707               2  Breast cancer possibly correlated genes after ...   \n",
      "3708               2  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3709               2                    MEF-iPS diff ES DNA methylation   \n",
      "3710               2  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      combined_score  \n",
      "0        1605.217526  \n",
      "1         283.031318  \n",
      "2          26.180243  \n",
      "3          33.230975  \n",
      "4          31.991884  \n",
      "...              ...  \n",
      "3706        0.001036  \n",
      "3707        0.000687  \n",
      "3708        0.000808  \n",
      "3709        0.000252  \n",
      "3710        0.000107  \n",
      "\n",
      "[3711 rows x 3 columns], 'KmeansCosineSixProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  Genes closest to the lead SNPs from Ben Neale'...   \n",
      "2                  1  10 percent of the longest transcripts of the h...   \n",
      "3                  1                                        Gdgtdrhyjgk   \n",
      "4                  1  genes closest to lead SNPs from Neale analysis...   \n",
      "...              ...                                                ...   \n",
      "4916               4  Turquise module gene list CRC (Normal and Cancer)   \n",
      "4917               4                          282+/282-, dataset: 70v5q   \n",
      "4918               4                                         Mizosample   \n",
      "4919               4                          282+/282-, dataset: 70uej   \n",
      "4920               4                                        Sca1+ genes   \n",
      "\n",
      "      combined_score  \n",
      "0        3204.063433  \n",
      "1         303.992195  \n",
      "2          39.366311  \n",
      "3          37.796644  \n",
      "4          65.531825  \n",
      "...              ...  \n",
      "4916        0.002602  \n",
      "4917        0.002281  \n",
      "4918        0.001043  \n",
      "4919        0.000876  \n",
      "4920        0.000006  \n",
      "\n",
      "[4921 rows x 3 columns], 'DBScanZeroProposed':       cluster_number                                            pathway  \\\n",
      "0                  0  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  0  10 percent of the longest transcripts of the h...   \n",
      "2                  0                                       inflammation   \n",
      "3                  0  https://science.sciencemag.org/content/365/645...   \n",
      "4                  0  Genes closest to the lead SNPs from Ben Neale'...   \n",
      "...              ...                                                ...   \n",
      "2978               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "2979               1                         All Gap probes upregulated   \n",
      "2980               1                          pathway enrichment celiac   \n",
      "2981               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "2982               1                    MEF-iPS diff ES DNA methylation   \n",
      "\n",
      "      combined_score  \n",
      "0        7516.478683  \n",
      "1         116.826333  \n",
      "2         350.228425  \n",
      "3          40.974086  \n",
      "4         194.777510  \n",
      "...              ...  \n",
      "2978        0.000262  \n",
      "2979        0.000175  \n",
      "2980        0.000111  \n",
      "2981        0.000012  \n",
      "2982        0.000002  \n",
      "\n",
      "[2983 rows x 3 columns], 'DBScanOneProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1  https://science.sciencemag.org/content/365/645...   \n",
      "3                  1  genes closest to lead SNPs from Neale analysis...   \n",
      "4                  1                                        Gdgtdrhyjgk   \n",
      "...              ...                                                ...   \n",
      "1245               1                                     pred all genes   \n",
      "1246               1  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "1247               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "1248               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "1249               1                    MEF-iPS diff ES DNA methylation   \n",
      "\n",
      "      combined_score  \n",
      "0        9670.786612  \n",
      "1         108.507651  \n",
      "2          52.016270  \n",
      "3          91.678083  \n",
      "4          39.257677  \n",
      "...              ...  \n",
      "1245        0.000026  \n",
      "1246        0.000019  \n",
      "1247        0.000002  \n",
      "1248        0.000001  \n",
      "1249        0.000001  \n",
      "\n",
      "[1250 rows x 3 columns], 'DBScanTwoProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1  https://science.sciencemag.org/content/365/645...   \n",
      "3                  1  genes closest to lead SNPs from Neale analysis...   \n",
      "4                  1                                        Gdgtdrhyjgk   \n",
      "...              ...                                                ...   \n",
      "1245               1                                     pred all genes   \n",
      "1246               1  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "1247               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "1248               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "1249               1                    MEF-iPS diff ES DNA methylation   \n",
      "\n",
      "      combined_score  \n",
      "0        9670.786612  \n",
      "1         108.507651  \n",
      "2          52.016270  \n",
      "3          91.678083  \n",
      "4          39.257677  \n",
      "...              ...  \n",
      "1245        0.000026  \n",
      "1246        0.000019  \n",
      "1247        0.000002  \n",
      "1248        0.000001  \n",
      "1249        0.000001  \n",
      "\n",
      "[1250 rows x 3 columns], 'KmeansCosineMinibatchSixProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1  genes closest to lead SNPs from Neale analysis...   \n",
      "3                  1  Genes closest to the lead SNPs from Ben Neale'...   \n",
      "4                  1  https://science.sciencemag.org/content/365/645...   \n",
      "...              ...                                                ...   \n",
      "4462               5                                       wild vs pale   \n",
      "4463               5  CD1 female mice adults - gestational CdCl2 vs ...   \n",
      "4464               5                                   Msc Role Disease   \n",
      "4465               5                                             ct1346   \n",
      "4466               5                Rheumatiod arthritis trasnscriptome   \n",
      "\n",
      "      combined_score  \n",
      "0        5952.375754  \n",
      "1         106.999496  \n",
      "2          81.248532  \n",
      "3         207.630169  \n",
      "4          38.332528  \n",
      "...              ...  \n",
      "4462        0.339474  \n",
      "4463        0.336937  \n",
      "4464        0.310167  \n",
      "4465        0.287661  \n",
      "4466        0.203385  \n",
      "\n",
      "[4467 rows x 3 columns], 'KmedoidsCosineSixProposed':       cluster_number                                            pathway  \\\n",
      "0                  3  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  3                                       inflammation   \n",
      "2                  3  10 percent of the longest transcripts of the h...   \n",
      "3                  3                                        Gdgtdrhyjgk   \n",
      "4                  3                 viral chromatin associated factors   \n",
      "...              ...                                                ...   \n",
      "4373               4                                      hola tomatito   \n",
      "4374               4                                         scc-cx HDE   \n",
      "4375               4                                      Endometriosis   \n",
      "4376               4                                        Sca1+ genes   \n",
      "4377               4                                              Rthth   \n",
      "\n",
      "      combined_score  \n",
      "0        1605.217526  \n",
      "1         283.031318  \n",
      "2          26.180243  \n",
      "3          33.230975  \n",
      "4          31.991884  \n",
      "...              ...  \n",
      "4373        0.037355  \n",
      "4374        0.032609  \n",
      "4375        0.031014  \n",
      "4376        0.022757  \n",
      "4377        0.014559  \n",
      "\n",
      "[4378 rows x 3 columns], 'GMMZeroProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1                                        Gdgtdrhyjgk   \n",
      "3                  1  The genes about major depressive disorder anal...   \n",
      "4                  1  Genes closest to the lead SNPs from Ben Neale'...   \n",
      "...              ...                                                ...   \n",
      "2969               2                          pathway enrichment celiac   \n",
      "2970               2                    MEF-iPS diff ES DNA methylation   \n",
      "2971               2      Blue module gene list CRC (Normal and Cancer)   \n",
      "2972               2                                     pred all genes   \n",
      "2973               2  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      combined_score  \n",
      "0        4740.450033  \n",
      "1          83.271456  \n",
      "2          51.791646  \n",
      "3          78.605707  \n",
      "4         199.716664  \n",
      "...              ...  \n",
      "2969        0.000545  \n",
      "2970        0.000203  \n",
      "2971        0.000095  \n",
      "2972        0.000074  \n",
      "2973        0.000001  \n",
      "\n",
      "[2974 rows x 3 columns], 'GMMOneProposed':       cluster_number                                            pathway  \\\n",
      "0                  4  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  4                                       inflammation   \n",
      "2                  4  The genes about major depressive disorder anal...   \n",
      "3                  4  10 percent of the longest transcripts of the h...   \n",
      "4                  4         Genes that were expressed and up regulated   \n",
      "...              ...                                                ...   \n",
      "3993               3  Identification of the novel conserved transcri...   \n",
      "3994               3                   GSE140581 p65 LPS peaks from GEO   \n",
      "3995               3   HPV-cervical cancer up-regulated genes vs Normal   \n",
      "3996               3                                         DEG of ZZW   \n",
      "3997               3                                             ct1346   \n",
      "\n",
      "      combined_score  \n",
      "0        1761.487115  \n",
      "1         530.789489  \n",
      "2          52.367357  \n",
      "3          20.339658  \n",
      "4          25.334638  \n",
      "...              ...  \n",
      "3993        0.001449  \n",
      "3994        0.001377  \n",
      "3995        0.001240  \n",
      "3996        0.001004  \n",
      "3997        0.001075  \n",
      "\n",
      "[3998 rows x 3 columns], 'BirchZeroProposed':        cluster_number                                            pathway  \\\n",
      "0                   7                                   High in human RG   \n",
      "1                   7                                           PAH test   \n",
      "2                   7                                    Phosphopeptides   \n",
      "3                   7                                           Als Gene   \n",
      "4                   7                      Nextera rapid capture targets   \n",
      "...               ...                                                ...   \n",
      "18277               4                                   mir target genes   \n",
      "18278               4               TargetScanV7 135b target All targets   \n",
      "18279               4  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "18280               4                          ARP1 WT vs E6 UTX binding   \n",
      "18281               4  shRNA knockdown of ARNT using TRC00003816 plas...   \n",
      "\n",
      "       combined_score  \n",
      "0          627.742397  \n",
      "1          220.093038  \n",
      "2          543.359329  \n",
      "3           81.249580  \n",
      "4           81.249580  \n",
      "...               ...  \n",
      "18277        0.034145  \n",
      "18278        0.033001  \n",
      "18279        0.028588  \n",
      "18280        0.026140  \n",
      "18281        0.021655  \n",
      "\n",
      "[18282 rows x 3 columns], 'BirchOneProposed':       cluster_number                                            pathway  \\\n",
      "0                  0  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  0  10 percent of the longest transcripts of the h...   \n",
      "2                  0                                        Gdgtdrhyjgk   \n",
      "3                  0                                       inflammation   \n",
      "4                  0                                        brain pleio   \n",
      "...              ...                                                ...   \n",
      "2214               1                          pathway enrichment celiac   \n",
      "2215               1                    MEF-iPS diff ES DNA methylation   \n",
      "2216               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "2217               1                                     pred all genes   \n",
      "2218               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      combined_score  \n",
      "0        4196.712210  \n",
      "1          77.340450  \n",
      "2          34.572409  \n",
      "3         145.170113  \n",
      "4          40.849789  \n",
      "...              ...  \n",
      "2214        0.000545  \n",
      "2215        0.000203  \n",
      "2216        0.000095  \n",
      "2217        0.000074  \n",
      "2218        0.000001  \n",
      "\n",
      "[2219 rows x 3 columns], 'BirchTwoProposed':       cluster_number                                            pathway  \\\n",
      "0                  0  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  0  10 percent of the longest transcripts of the h...   \n",
      "2                  0                                        Gdgtdrhyjgk   \n",
      "3                  0                                       inflammation   \n",
      "4                  0                                        brain pleio   \n",
      "...              ...                                                ...   \n",
      "2214               1                          pathway enrichment celiac   \n",
      "2215               1                    MEF-iPS diff ES DNA methylation   \n",
      "2216               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "2217               1                                     pred all genes   \n",
      "2218               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      combined_score  \n",
      "0        4196.712210  \n",
      "1          77.340450  \n",
      "2          34.572409  \n",
      "3         145.170113  \n",
      "4          40.849789  \n",
      "...              ...  \n",
      "2214        0.000545  \n",
      "2215        0.000203  \n",
      "2216        0.000095  \n",
      "2217        0.000074  \n",
      "2218        0.000001  \n",
      "\n",
      "[2219 rows x 3 columns], 'KmeansCosineTwoMinAICProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1                                        Gdgtdrhyjgk   \n",
      "3                  1                                       inflammation   \n",
      "4                  1                                        brain pleio   \n",
      "...              ...                                                ...   \n",
      "2214               0                          pathway enrichment celiac   \n",
      "2215               0                    MEF-iPS diff ES DNA methylation   \n",
      "2216               0      Blue module gene list CRC (Normal and Cancer)   \n",
      "2217               0                                     pred all genes   \n",
      "2218               0  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      combined_score  \n",
      "0        4196.712210  \n",
      "1          77.340450  \n",
      "2          34.572409  \n",
      "3         145.170113  \n",
      "4          40.849789  \n",
      "...              ...  \n",
      "2214        0.000545  \n",
      "2215        0.000203  \n",
      "2216        0.000095  \n",
      "2217        0.000074  \n",
      "2218        0.000001  \n",
      "\n",
      "[2219 rows x 3 columns], 'KmedoidsCosineThreeMinAICProposed':       cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1                                        Gdgtdrhyjgk   \n",
      "3                  1                                       inflammation   \n",
      "4                  1                                        brain pleio   \n",
      "...              ...                                                ...   \n",
      "2872               0                                Osteoclast specific   \n",
      "2873               0                              INTERACTOMIC ANALYSIS   \n",
      "2874               0                                         Mizosample   \n",
      "2875               0                                     kp vs wt 1.5fc   \n",
      "2876               0   HPV-cervical cancer up-regulated genes vs Normal   \n",
      "\n",
      "      combined_score  \n",
      "0        4196.712210  \n",
      "1          77.340450  \n",
      "2          34.572409  \n",
      "3         145.170113  \n",
      "4          40.849789  \n",
      "...              ...  \n",
      "2872        0.003139  \n",
      "2873        0.002761  \n",
      "2874        0.002761  \n",
      "2875        0.001441  \n",
      "2876        0.000145  \n",
      "\n",
      "[2877 rows x 3 columns]}\n"
     ]
    }
   ],
   "source": [
    "ENRICHR_add_URL = 'https://maayanlab.cloud/Enrichr/addList'\n",
    "ENRICHR_URL = 'https://maayanlab.cloud/Enrichr/enrich'\n",
    "gene_set_library = 'Enrichr_Users_Contributed_Lists_2020'\n",
    "query_string = '?userListId=%s&backgroundType=%s'\n",
    "path_clust_OR_dict ={}\n",
    "path_clust_score_dict ={}\n",
    "s= requests.Session()\n",
    "retry = Retry(connect=3, backoff_factor=0.5)\n",
    "adapter = HTTPAdapter(max_retries=retry)\n",
    "s.mount('http://', adapter)\n",
    "s.mount('https://', adapter)\n",
    "user_ids_full_list = {}\n",
    "for c_lab in cluster_method_labels:\n",
    "    print(c_lab)\n",
    "    clust_genes_df = gene_clust_dict[c_lab]\n",
    "    c_num_labs = clust_genes_df.columns\n",
    "    OR_list = []\n",
    "    score_list = []\n",
    "    user_ids_list = []\n",
    "    all_list = []\n",
    "    for c_num_lab in c_num_labs:\n",
    "        try: c_num = int(c_num_lab[-1])\n",
    "        except: break\n",
    "        clust_cnum_genes = clust_genes_df.index[clust_genes_df[c_num_lab] == 1].tolist()\n",
    "        genes_str = '\\n'.join(clust_cnum_genes)\n",
    "        description = c_lab + ' gene list'\n",
    "        payload = {\n",
    "            'list': (None, genes_str),\n",
    "            'description': (None, description)\n",
    "        }\n",
    "        response = s.post(ENRICHR_add_URL, files=payload)\n",
    "        if not response.ok:\n",
    "            #raise Exception\n",
    "            print('Error analyzing gene list')\n",
    "        else:\n",
    "            data = json.loads(response.text)\n",
    "            #print(data)\n",
    "            user_list_id = data['userListId']\n",
    "            #print(user_list_id)\n",
    "            user_ids_list.append(user_list_id)\n",
    "            response = s.get(\n",
    "            ENRICHR_URL + query_string % (user_list_id, gene_set_library)\n",
    "            )\n",
    "            if not response.ok:\n",
    "                #raise Exception\n",
    "                print('Error fetching enrichment results')\n",
    "            else:\n",
    "                data = json.loads(response.text)[gene_set_library]\n",
    "                if len(data)>0:\n",
    "                    for i in range(len(data)):\n",
    "                        rank = data[i][0]\n",
    "                        pathway = data[i][1]\n",
    "                        pval = data[i][2]\n",
    "                        OR = data[i][3]\n",
    "                        score = data[i][4]\n",
    "                        overlap_genes = data[i][5]\n",
    "                        adjust_pval = data[i][6]\n",
    "                        all_row = {\"cluster_number\": c_num,\n",
    "                            \"pathway\": pathway,\n",
    "                            \"rank\": rank,\n",
    "                            \"pval\": pval, \n",
    "                            \"adjusted_pval\": adjust_pval,\n",
    "                            \"Odds_Ratio\" : OR,\n",
    "                            \"combined_score\": score,\n",
    "                            \"n_overlap\": len(overlap_genes),\n",
    "                            \"overlap_genes\": overlap_genes}\n",
    "                        all_list.append(all_row)\n",
    "                        OR_row = {\"cluster_number\": c_num,\n",
    "                            \"pathway\": pathway,\n",
    "                            \"Odds_Ratio\" : OR}\n",
    "                        OR_list.append(OR_row)\n",
    "                        score_row = {\"cluster_number\": c_num,\n",
    "                                \"pathway\": pathway,\n",
    "                                \"combined_score\": score}\n",
    "                        score_list.append(score_row)\n",
    "                else:\n",
    "                    OR_row = {\"cluster_number\": c_num,\n",
    "                              \"pathway\": \"No_pathway\"}\n",
    "                    OR_list.append(OR_row)\n",
    "                    score_row = {\"cluster_number\": c_num,\n",
    "                              \"pathway\": \"No_pathway\"}\n",
    "                    score_list.append(score_row)\n",
    "    user_ids_full_list[c_lab] = user_ids_list\n",
    "    OR_df = pd.DataFrame(OR_list)\n",
    "    score_df = pd.DataFrame(score_list)\n",
    "    all_df = pd.DataFrame(all_list)\n",
    "    OR_df.to_csv(path_dir+\"/clust_paths_OR_\"+c_lab+\".csv\")\n",
    "    score_df.to_csv(path_dir+\"/clust_path_combined_score_\"+c_lab+\".csv\")\n",
    "    all_df.to_csv(path_dir+\"/clust_path_enrichment_\"+c_lab+\".csv\")\n",
    "    path_clust_OR_dict[c_lab] = OR_df\n",
    "    path_clust_score_dict[c_lab] = score_df\n",
    "print(path_clust_score_dict)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get the pathway association for gene sets in each PCA bin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set_library = 'Enrichr_Users_Contributed_Lists_2020'\n",
    "query_string = '?userListId=%s&backgroundType=%s'\n",
    "path_pca_OR_dict ={}\n",
    "path_pca_score_dict ={}\n",
    "path_pca_adj_score_dict = {}\n",
    "s= requests.Session()\n",
    "retry = Retry(connect=3, backoff_factor=0.5)\n",
    "adapter = HTTPAdapter(max_retries=retry)\n",
    "s.mount('http://', adapter)\n",
    "s.mount('https://', adapter)\n",
    "user_ids_full_list = {}\n",
    "for lab in pc_gene_groups_dict.keys():\n",
    "    pc_genes_df = pc_gene_groups_dict[lab]\n",
    "    bin_labs = pc_genes_df.columns\n",
    "    OR_list = []\n",
    "    score_list = []\n",
    "    adj_score_list = []\n",
    "    user_ids_list = []\n",
    "    all_list = []\n",
    "    for bin in bin_labs:\n",
    "        bin_genes = pc_genes_df[bin].dropna(axis=0).tolist()\n",
    "        bin_mid = bin_range_dict[lab].loc[bin,\"mid\"]\n",
    "        genes_str = '\\n'.join(bin_genes)\n",
    "        description = bin + ' gene list'\n",
    "        payload = {\n",
    "            'list': (None, genes_str),\n",
    "            'description': (None, description)\n",
    "        }\n",
    "        response = s.post(ENRICHR_add_URL, files=payload)\n",
    "        if not response.ok:\n",
    "            #raise Exception\n",
    "            print('Error analyzing gene list')\n",
    "        else:\n",
    "            data = json.loads(response.text)\n",
    "            #print(data)\n",
    "            user_list_id = data['userListId']\n",
    "            #print(user_list_id)\n",
    "            user_ids_list.append(user_list_id)\n",
    "            response = s.get(\n",
    "            ENRICHR_URL + query_string % (user_list_id, gene_set_library)\n",
    "            )\n",
    "            if not response.ok:\n",
    "                #raise Exception\n",
    "                print('Error fetching enrichment results')\n",
    "            else:\n",
    "                data = json.loads(response.text)[gene_set_library]\n",
    "                if len(data)>0:\n",
    "                    for i in range(len(data)):\n",
    "                        rank = data[i][0]\n",
    "                        pathway = data[i][1]\n",
    "                        pval = data[i][2]\n",
    "                        OR = data[i][3]\n",
    "                        score = data[i][4]\n",
    "                        adj_score = score*bin_mid\n",
    "                        overlap_genes = data[i][5]\n",
    "                        adjust_pval = data[i][6]\n",
    "                        all_row = {\"bin\": bin,\n",
    "                            \"pathway\": pathway,\n",
    "                            \"rank\": rank,\n",
    "                            \"pval\": pval, \n",
    "                            \"adjusted_pval\": adjust_pval,\n",
    "                            \"Odds_Ratio\" : OR,\n",
    "                            \"combined_score\": score,\n",
    "                            \"adjusted_score\": adj_score,\n",
    "                            \"n_overlap\": len(overlap_genes),\n",
    "                            \"overlap_genes\": overlap_genes}\n",
    "                        all_list.append(all_row)\n",
    "                        OR_row = {\"bin\": bin,\n",
    "                            \"pathway\": pathway,\n",
    "                            \"Odds_Ratio\" : OR}\n",
    "                        OR_list.append(OR_row)\n",
    "                        score_row = {\"bin\": bin,\n",
    "                                \"pathway\": pathway,\n",
    "                                \"combined_score\": score}\n",
    "                        score_list.append(score_row)\n",
    "                        adj_score_row = {\"bin\": bin,\n",
    "                                \"pathway\": pathway,\n",
    "                                \"adjusted_score\": adj_score}\n",
    "                        adj_score_list.append(adj_score_row)\n",
    "                else:\n",
    "                    OR_row = {\"bin\": bin,\n",
    "                              \"pathway\": \"No_pathway\"}\n",
    "                    OR_list.append(OR_row)\n",
    "                    score_row = {\"bin\": bin,\n",
    "                              \"pathway\": \"No_pathway\"}\n",
    "                    score_list.append(score_row)\n",
    "                    adj_score_row = {\"bin\": bin,\n",
    "                              \"pathway\": \"No_pathway\"}\n",
    "                    adj_score_list.append(adj_score_row)\n",
    "    user_ids_full_list[bin] = user_ids_list\n",
    "    OR_df = pd.DataFrame(OR_list)\n",
    "    score_df = pd.DataFrame(score_list)\n",
    "    all_df = pd.DataFrame(all_list)\n",
    "    adj_score_df = pd.DataFrame(adj_score_list)\n",
    "    OR_df.to_csv(path_dir+\"/pca_paths_OR_\"+lab+\".csv\")\n",
    "    score_df.to_csv(path_dir+\"/pca_path_combined_score_\"+lab+\".csv\")\n",
    "    all_df.to_csv(path_dir+\"/pca_path_enrichment_\"+lab+\".csv\")\n",
    "    adj_score_df.to_csv(path_dir + \"/pca_path_adj_score\"+lab+\".csv\")\n",
    "    path_pca_OR_dict[lab] = OR_df\n",
    "    path_pca_score_dict[lab] = score_df\n",
    "    path_pca_adj_score_dict[lab] = adj_score_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "# OFFLINE VERSION - load latest \n",
    "path_clust_OR_dict = {}\n",
    "path_clust_score_dict = {}\n",
    "user_ids_full_list = {}\n",
    "for c_lab in cluster_method_labels:\n",
    "    clust_genes_df = gene_clust_dict[c_lab]\n",
    "    OR_df = pd.read_csv(path_dir+\"/clust_paths_OR_\"+c_lab+\".csv\", index_col=0)\n",
    "    score_df = pd.read_csv(path_dir+\"/clust_path_combined_score_\"+c_lab+\".csv\", index_col=0)\n",
    "    all_df = pd.read_csv(path_dir+\"/clust_path_enrichment_\"+c_lab+\".csv\", index_col=0)\n",
    "    path_clust_OR_dict[c_lab] = OR_df\n",
    "    path_clust_score_dict[c_lab] = score_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter the dominant pathways for each cluster.\n",
    "Get the top fifty pathways overall for each cluster method by `combined score` and `odds_ratio`.\n",
    "Also get the top 5 pathways per cluster by `combined_score` and `odds_ratio`.\n",
    "\n",
    "The odds ratio is computed using this formula:\n",
    "\n",
    "$$ oddsRatio = \\frac{1.0 * a * d}{max(1.0 * b * c, 1)}$$ \n",
    "\n",
    "where: a are the overlapping genes, b are the genes in the annotated set - overlapping genes, c- are the genes in the input set - overlapping genes, and d- are the 20,000 genes (or total genes in the background) - genes in the annotated set - genes in the input set + overlapping genes\n",
    "\n",
    "Finally, the combined score is a combination of the p-value and odds ratio calculated by multiplying the two scores as follows:\n",
    "\n",
    "$$ c = -log(p) * \\text{oddsRatio} $$\n",
    "Where c is the combined score, p is the p-value computed using Fisher's exact test, and oddsRatio is the odds ratio. The combined score provides a compromise between both methods and in several benchmarks we show that it reports the best rankings when compared with the other scoring schemes.\n",
    "\n",
    "Enrichr provides all four options for sorting enriched terms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  https://science.sciencemag.org/content/365/645...   \n",
      "2                  1  10 percent of the longest transcripts of the h...   \n",
      "3                  1                                             ct1346   \n",
      "4                  1                                       inflammation   \n",
      "...              ...                                                ...   \n",
      "3676               3      Blue module gene list CRC (Normal and Cancer)   \n",
      "3677               3  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "3678               3  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3679               3                    MEF-iPS diff ES DNA methylation   \n",
      "3680               3                                     pred all genes   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      21.038178  \n",
      "1       4.110373  \n",
      "2       3.457802  \n",
      "3       2.345616  \n",
      "4      17.413263  \n",
      "...          ...  \n",
      "3676    0.240561  \n",
      "3677    0.328778  \n",
      "3678    0.451822  \n",
      "3679    0.126142  \n",
      "3680    0.437391  \n",
      "\n",
      "[3681 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "c_lab = cluster_method_labels[0]\n",
    "print(path_clust_OR_dict[c_lab])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KmeansLloydStdFourDarrous\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  https://science.sciencemag.org/content/365/645...   \n",
      "2                  1  10 percent of the longest transcripts of the h...   \n",
      "3                  1                                             ct1346   \n",
      "4                  1                                       inflammation   \n",
      "...              ...                                                ...   \n",
      "3676               3      Blue module gene list CRC (Normal and Cancer)   \n",
      "3677               3  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "3678               3  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3679               3                    MEF-iPS diff ES DNA methylation   \n",
      "3680               3                                     pred all genes   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      21.038178  \n",
      "1       4.110373  \n",
      "2       3.457802  \n",
      "3       2.345616  \n",
      "4      17.413263  \n",
      "...          ...  \n",
      "3676    0.240561  \n",
      "3677    0.328778  \n",
      "3678    0.451822  \n",
      "3679    0.126142  \n",
      "3680    0.437391  \n",
      "\n",
      "[3681 rows x 3 columns]\n",
      "KmeansLloydStdSixDarrous\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  4  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  4  https://science.sciencemag.org/content/365/645...   \n",
      "2                  4                                 az mlm p <.01+ WES   \n",
      "3                  4         Genes that were expressed and up regulated   \n",
      "4                  4                            microRNA in tubercuosis   \n",
      "...              ...                                                ...   \n",
      "4887               1  downregulated, dataset: 6d5ee3258f0d93c3f9f758...   \n",
      "4888               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "4889               1                                         down in VS   \n",
      "4890               1                      Upregulated genes in WT brain   \n",
      "4891               1  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0       9.788691  \n",
      "1       4.431897  \n",
      "2       4.337006  \n",
      "3       3.507938  \n",
      "4      12.422306  \n",
      "...          ...  \n",
      "4887    0.204902  \n",
      "4888    0.196770  \n",
      "4889    0.425578  \n",
      "4890    0.148564  \n",
      "4891    0.132257  \n",
      "\n",
      "[4892 rows x 3 columns]\n",
      "KmeansElkanStdFourDarrous\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  https://science.sciencemag.org/content/365/645...   \n",
      "2                  1  10 percent of the longest transcripts of the h...   \n",
      "3                  1                                             ct1346   \n",
      "4                  1                                       inflammation   \n",
      "...              ...                                                ...   \n",
      "3676               3      Blue module gene list CRC (Normal and Cancer)   \n",
      "3677               3  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "3678               3  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3679               3                    MEF-iPS diff ES DNA methylation   \n",
      "3680               3                                     pred all genes   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      21.038178  \n",
      "1       4.110373  \n",
      "2       3.457802  \n",
      "3       2.345616  \n",
      "4      17.413263  \n",
      "...          ...  \n",
      "3676    0.240561  \n",
      "3677    0.328778  \n",
      "3678    0.451822  \n",
      "3679    0.126142  \n",
      "3680    0.437391  \n",
      "\n",
      "[3681 rows x 3 columns]\n",
      "KmeansLloydStdFourProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1                                       inflammation   \n",
      "2                  1         Genes that were expressed and up regulated   \n",
      "3                  1  10 percent of the longest transcripts of the h...   \n",
      "4                  1                                      wound healing   \n",
      "...              ...                                                ...   \n",
      "3728               2  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3729               2                                         harvard hw   \n",
      "3730               2                                     pred all genes   \n",
      "3731               2                    MEF-iPS diff ES DNA methylation   \n",
      "3732               2  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      18.251224  \n",
      "1      28.640828  \n",
      "2       3.091057  \n",
      "3       2.419499  \n",
      "4       7.392817  \n",
      "...          ...  \n",
      "3728    0.551057  \n",
      "3729    0.148748  \n",
      "3730    0.490941  \n",
      "3731    0.108477  \n",
      "3732    0.205006  \n",
      "\n",
      "[3733 rows x 3 columns]\n",
      "KmeansLloydStdSixProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  2  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  2                                       inflammation   \n",
      "2                  2  10 percent of the longest transcripts of the h...   \n",
      "3                  2         Genes that were expressed and up regulated   \n",
      "4                  2                                      wound healing   \n",
      "...              ...                                                ...   \n",
      "4851               5                                        Sca1+ genes   \n",
      "4852               5                                             Cn1346   \n",
      "4853               5  Meta-analysis of autosomal sex DE genes across...   \n",
      "4854               5  Turquise module gene list CRC (Normal and Cancer)   \n",
      "4855               5                                     pred all genes   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      17.691123  \n",
      "1      33.404769  \n",
      "2       2.503425  \n",
      "3       3.160211  \n",
      "4       8.611982  \n",
      "...          ...  \n",
      "4851    0.478981  \n",
      "4852    0.511611  \n",
      "4853    0.272446  \n",
      "4854    0.237482  \n",
      "4855    0.416957  \n",
      "\n",
      "[4856 rows x 3 columns]\n",
      "KmeansElkanStdFourProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1                                       inflammation   \n",
      "2                  1         Genes that were expressed and up regulated   \n",
      "3                  1  10 percent of the longest transcripts of the h...   \n",
      "4                  1                                      wound healing   \n",
      "...              ...                                                ...   \n",
      "3728               2  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3729               2                                         harvard hw   \n",
      "3730               2                                     pred all genes   \n",
      "3731               2                    MEF-iPS diff ES DNA methylation   \n",
      "3732               2  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      18.251224  \n",
      "1      28.640828  \n",
      "2       3.091057  \n",
      "3       2.419499  \n",
      "4       7.392817  \n",
      "...          ...  \n",
      "3728    0.551057  \n",
      "3729    0.148748  \n",
      "3730    0.490941  \n",
      "3731    0.108477  \n",
      "3732    0.205006  \n",
      "\n",
      "[3733 rows x 3 columns]\n",
      "KmedoidsEuclideanProposedFour\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  0  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  0                                       inflammation   \n",
      "2                  0  10 percent of the longest transcripts of the h...   \n",
      "3                  0                                        Gdgtdrhyjgk   \n",
      "4                  0                 viral chromatin associated factors   \n",
      "...              ...                                                ...   \n",
      "3706               2                                   wte deg analysis   \n",
      "3707               2  Breast cancer possibly correlated genes after ...   \n",
      "3708               2  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "3709               2                    MEF-iPS diff ES DNA methylation   \n",
      "3710               2  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      18.798968  \n",
      "1      20.805400  \n",
      "2       2.373728  \n",
      "3       3.032666  \n",
      "4       2.983293  \n",
      "...          ...  \n",
      "3706    0.256050  \n",
      "3707    0.287932  \n",
      "3708    0.508820  \n",
      "3709    0.215704  \n",
      "3710    0.271657  \n",
      "\n",
      "[3711 rows x 3 columns]\n",
      "KmeansCosineSixProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  Genes closest to the lead SNPs from Ben Neale'...   \n",
      "2                  1  10 percent of the longest transcripts of the h...   \n",
      "3                  1                                        Gdgtdrhyjgk   \n",
      "4                  1  genes closest to lead SNPs from Neale analysis...   \n",
      "...              ...                                                ...   \n",
      "4916               4  Turquise module gene list CRC (Normal and Cancer)   \n",
      "4917               4                          282+/282-, dataset: 70v5q   \n",
      "4918               4                                         Mizosample   \n",
      "4919               4                          282+/282-, dataset: 70uej   \n",
      "4920               4                                        Sca1+ genes   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      37.788260  \n",
      "1      24.419720  \n",
      "2       3.374674  \n",
      "3       3.990012  \n",
      "4       7.143791  \n",
      "...          ...  \n",
      "4916    0.287892  \n",
      "4917    0.282298  \n",
      "4918    0.252774  \n",
      "4919    0.290352  \n",
      "4920    0.180007  \n",
      "\n",
      "[4921 rows x 3 columns]\n",
      "DBScanZeroProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  0  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  0  10 percent of the longest transcripts of the h...   \n",
      "2                  0                                       inflammation   \n",
      "3                  0  https://science.sciencemag.org/content/365/645...   \n",
      "4                  0  Genes closest to the lead SNPs from Ben Neale'...   \n",
      "...              ...                                                ...   \n",
      "2978               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "2979               1                         All Gap probes upregulated   \n",
      "2980               1                          pathway enrichment celiac   \n",
      "2981               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "2982               1                    MEF-iPS diff ES DNA methylation   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      52.200359  \n",
      "1       4.602693  \n",
      "2      27.205318  \n",
      "3       3.492887  \n",
      "4      17.751522  \n",
      "...          ...  \n",
      "2978    0.261040  \n",
      "2979    0.207642  \n",
      "2980    0.197931  \n",
      "2981    0.229874  \n",
      "2982    0.090859  \n",
      "\n",
      "[2983 rows x 3 columns]\n",
      "DBScanOneProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1  https://science.sciencemag.org/content/365/645...   \n",
      "3                  1  genes closest to lead SNPs from Neale analysis...   \n",
      "4                  1                                        Gdgtdrhyjgk   \n",
      "...              ...                                                ...   \n",
      "1245               1                                     pred all genes   \n",
      "1246               1  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "1247               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "1248               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "1249               1                    MEF-iPS diff ES DNA methylation   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      30.461852  \n",
      "1       2.709316  \n",
      "2       2.431063  \n",
      "3       4.414697  \n",
      "4       2.385412  \n",
      "...          ...  \n",
      "1245    0.646030  \n",
      "1246    0.496040  \n",
      "1247    0.317589  \n",
      "1248    0.261591  \n",
      "1249    0.250096  \n",
      "\n",
      "[1250 rows x 3 columns]\n",
      "DBScanTwoProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1  https://science.sciencemag.org/content/365/645...   \n",
      "3                  1  genes closest to lead SNPs from Neale analysis...   \n",
      "4                  1                                        Gdgtdrhyjgk   \n",
      "...              ...                                                ...   \n",
      "1245               1                                     pred all genes   \n",
      "1246               1  Bacteria translocation, dataset: 4b799ecdbaaed...   \n",
      "1247               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "1248               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "1249               1                    MEF-iPS diff ES DNA methylation   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      30.461852  \n",
      "1       2.709316  \n",
      "2       2.431063  \n",
      "3       4.414697  \n",
      "4       2.385412  \n",
      "...          ...  \n",
      "1245    0.646030  \n",
      "1246    0.496040  \n",
      "1247    0.317589  \n",
      "1248    0.261591  \n",
      "1249    0.250096  \n",
      "\n",
      "[1250 rows x 3 columns]\n",
      "KmeansCosineMinibatchSixProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1  genes closest to lead SNPs from Neale analysis...   \n",
      "3                  1  Genes closest to the lead SNPs from Ben Neale'...   \n",
      "4                  1  https://science.sciencemag.org/content/365/645...   \n",
      "...              ...                                                ...   \n",
      "4462               5                                       wild vs pale   \n",
      "4463               5  CD1 female mice adults - gestational CdCl2 vs ...   \n",
      "4464               5                                   Msc Role Disease   \n",
      "4465               5                                             ct1346   \n",
      "4466               5                Rheumatiod arthritis trasnscriptome   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      46.729262  \n",
      "1       4.514899  \n",
      "2       7.136804  \n",
      "3      18.570492  \n",
      "4       3.455515  \n",
      "...          ...  \n",
      "4462    0.932856  \n",
      "4463    0.929785  \n",
      "4464    0.896765  \n",
      "4465    0.868038  \n",
      "4466    0.750350  \n",
      "\n",
      "[4467 rows x 3 columns]\n",
      "KmedoidsCosineSixProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  3  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  3                                       inflammation   \n",
      "2                  3  10 percent of the longest transcripts of the h...   \n",
      "3                  3                                        Gdgtdrhyjgk   \n",
      "4                  3                 viral chromatin associated factors   \n",
      "...              ...                                                ...   \n",
      "4373               4                                      hola tomatito   \n",
      "4374               4                                         scc-cx HDE   \n",
      "4375               4                                      Endometriosis   \n",
      "4376               4                                        Sca1+ genes   \n",
      "4377               4                                              Rthth   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      18.798968  \n",
      "1      20.805400  \n",
      "2       2.373728  \n",
      "3       3.032666  \n",
      "4       2.983293  \n",
      "...          ...  \n",
      "4373    0.405394  \n",
      "4374    0.389307  \n",
      "4375    0.383609  \n",
      "4376    0.351091  \n",
      "4377    0.311074  \n",
      "\n",
      "[4378 rows x 3 columns]\n",
      "GMMZeroProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1                                        Gdgtdrhyjgk   \n",
      "3                  1  The genes about major depressive disorder anal...   \n",
      "4                  1  Genes closest to the lead SNPs from Ben Neale'...   \n",
      "...              ...                                                ...   \n",
      "2969               2                          pathway enrichment celiac   \n",
      "2970               2                    MEF-iPS diff ES DNA methylation   \n",
      "2971               2      Blue module gene list CRC (Normal and Cancer)   \n",
      "2972               2                                     pred all genes   \n",
      "2973               2  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      35.143114  \n",
      "1       3.664986  \n",
      "2       3.724960  \n",
      "3       6.309524  \n",
      "4      16.170752  \n",
      "...          ...  \n",
      "2969    0.369609  \n",
      "2970    0.341445  \n",
      "2971    0.322708  \n",
      "2972    0.559603  \n",
      "2973    0.247879  \n",
      "\n",
      "[2974 rows x 3 columns]\n",
      "GMMOneProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  4  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  4                                       inflammation   \n",
      "2                  4  The genes about major depressive disorder anal...   \n",
      "3                  4  10 percent of the longest transcripts of the h...   \n",
      "4                  4         Genes that were expressed and up regulated   \n",
      "...              ...                                                ...   \n",
      "3993               3  Identification of the novel conserved transcri...   \n",
      "3994               3                   GSE140581 p65 LPS peaks from GEO   \n",
      "3995               3   HPV-cervical cancer up-regulated genes vs Normal   \n",
      "3996               3                                         DEG of ZZW   \n",
      "3997               3                                             ct1346   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      23.706340  \n",
      "1      32.830438  \n",
      "2       5.839614  \n",
      "3       2.452579  \n",
      "4       3.101363  \n",
      "...          ...  \n",
      "3993    0.197001  \n",
      "3994    0.195250  \n",
      "3995    0.191726  \n",
      "3996    0.247554  \n",
      "3997    0.322396  \n",
      "\n",
      "[3998 rows x 3 columns]\n",
      "BirchZeroProposed\n",
      "       cluster_number                                            pathway  \\\n",
      "0                   7                                   High in human RG   \n",
      "1                   7                                           PAH test   \n",
      "2                   7                                    Phosphopeptides   \n",
      "3                   7                                           Als Gene   \n",
      "4                   7                      Nextera rapid capture targets   \n",
      "...               ...                                                ...   \n",
      "18277               4                                   mir target genes   \n",
      "18278               4               TargetScanV7 135b target All targets   \n",
      "18279               4  Skin Cutaneous Melanoma under expressed genes ...   \n",
      "18280               4                          ARP1 WT vs E6 UTX binding   \n",
      "18281               4  shRNA knockdown of ARNT using TRC00003816 plas...   \n",
      "\n",
      "       Odds_Ratio  \n",
      "0       87.477876  \n",
      "1       39.152610  \n",
      "2      120.854545  \n",
      "3       19.036072  \n",
      "4       19.036072  \n",
      "...           ...  \n",
      "18277    0.394295  \n",
      "18278    0.390312  \n",
      "18279    0.374177  \n",
      "18280    0.364616  \n",
      "18281    0.345682  \n",
      "\n",
      "[18282 rows x 3 columns]\n",
      "BirchOneProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  0  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  0  10 percent of the longest transcripts of the h...   \n",
      "2                  0                                        Gdgtdrhyjgk   \n",
      "3                  0                                       inflammation   \n",
      "4                  0                                        brain pleio   \n",
      "...              ...                                                ...   \n",
      "2214               1                          pathway enrichment celiac   \n",
      "2215               1                    MEF-iPS diff ES DNA methylation   \n",
      "2216               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "2217               1                                     pred all genes   \n",
      "2218               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      25.210486  \n",
      "1       2.951412  \n",
      "2       2.725184  \n",
      "3      13.129922  \n",
      "4       4.227302  \n",
      "...          ...  \n",
      "2214    0.369609  \n",
      "2215    0.341445  \n",
      "2216    0.322708  \n",
      "2217    0.559603  \n",
      "2218    0.247879  \n",
      "\n",
      "[2219 rows x 3 columns]\n",
      "BirchTwoProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  0  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  0  10 percent of the longest transcripts of the h...   \n",
      "2                  0                                        Gdgtdrhyjgk   \n",
      "3                  0                                       inflammation   \n",
      "4                  0                                        brain pleio   \n",
      "...              ...                                                ...   \n",
      "2214               1                          pathway enrichment celiac   \n",
      "2215               1                    MEF-iPS diff ES DNA methylation   \n",
      "2216               1      Blue module gene list CRC (Normal and Cancer)   \n",
      "2217               1                                     pred all genes   \n",
      "2218               1  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      25.210486  \n",
      "1       2.951412  \n",
      "2       2.725184  \n",
      "3      13.129922  \n",
      "4       4.227302  \n",
      "...          ...  \n",
      "2214    0.369609  \n",
      "2215    0.341445  \n",
      "2216    0.322708  \n",
      "2217    0.559603  \n",
      "2218    0.247879  \n",
      "\n",
      "[2219 rows x 3 columns]\n",
      "KmeansCosineTwoMinAICProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1                                        Gdgtdrhyjgk   \n",
      "3                  1                                       inflammation   \n",
      "4                  1                                        brain pleio   \n",
      "...              ...                                                ...   \n",
      "2214               0                          pathway enrichment celiac   \n",
      "2215               0                    MEF-iPS diff ES DNA methylation   \n",
      "2216               0      Blue module gene list CRC (Normal and Cancer)   \n",
      "2217               0                                     pred all genes   \n",
      "2218               0  Turquise module gene list CRC (Normal and Cancer)   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      25.210486  \n",
      "1       2.951412  \n",
      "2       2.725184  \n",
      "3      13.129922  \n",
      "4       4.227302  \n",
      "...          ...  \n",
      "2214    0.369609  \n",
      "2215    0.341445  \n",
      "2216    0.322708  \n",
      "2217    0.559603  \n",
      "2218    0.247879  \n",
      "\n",
      "[2219 rows x 3 columns]\n",
      "KmedoidsCosineThreeMinAICProposed\n",
      "      cluster_number                                            pathway  \\\n",
      "0                  1  lead SNPs from Neale v2 analysis of UKBB data,...   \n",
      "1                  1  10 percent of the longest transcripts of the h...   \n",
      "2                  1                                        Gdgtdrhyjgk   \n",
      "3                  1                                       inflammation   \n",
      "4                  1                                        brain pleio   \n",
      "...              ...                                                ...   \n",
      "2872               0                                Osteoclast specific   \n",
      "2873               0                              INTERACTOMIC ANALYSIS   \n",
      "2874               0                                         Mizosample   \n",
      "2875               0                                     kp vs wt 1.5fc   \n",
      "2876               0   HPV-cervical cancer up-regulated genes vs Normal   \n",
      "\n",
      "      Odds_Ratio  \n",
      "0      25.210486  \n",
      "1       2.951412  \n",
      "2       2.725184  \n",
      "3      13.129922  \n",
      "4       4.227302  \n",
      "...          ...  \n",
      "2872    0.228486  \n",
      "2873    0.289612  \n",
      "2874    0.289612  \n",
      "2875    0.263283  \n",
      "2876    0.139858  \n",
      "\n",
      "[2877 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "path_tops_dict = {}\n",
    "top_five_lab = \"TopFive\"\n",
    "top_fifty_lab = \"TopFifty\"\n",
    "score_lab = \"CombinedScore\"\n",
    "OR_lab = \"OddsRatio\"\n",
    "for lab in cluster_method_labels:\n",
    "    print(lab)\n",
    "    print(path_clust_OR_dict[lab])\n",
    "    sorted_df = path_clust_OR_dict[lab].sort_values(by=OR_lab, ascending=False)\n",
    "    df_wide = sorted_df.pivot_table(index='pathway', columns='cluster_number', values=OR_lab)\n",
    "    ind_list = df_wide.index\n",
    "    col_list = df_wide.columns\n",
    "    df_wide = pd.DataFrame(data = np.nan_to_num(df_wide),\n",
    "                           index=ind_list,\n",
    "                           columns = col_list)\n",
    "    df_long= pd.melt(df_wide, ignore_index=False, value_name=OR_lab)\n",
    "    df_long= df_long.reset_index().sort_values(by=OR_lab, ascending=False)\n",
    "    top_fifty_byOR_df = df_long.head(50)\n",
    "    top_five_perclust_df = df_long.groupby(\"cluster_number\").head(5)\n",
    "    OR_dict = {top_five_lab:top_five_perclust_df, top_fifty_lab:top_fifty_byOR_df}\n",
    "    sorted_df = path_clust_score_dict[lab].sort_values(by=score_lab, ascending=False)\n",
    "    df_wide = sorted_df.pivot_table(index='pathway', columns='cluster_number', values=score_lab)\n",
    "    ind_list = df_wide.index\n",
    "    col_list = df_wide.columns\n",
    "    df_wide = pd.DataFrame(data = np.nan_to_num(df_wide),\n",
    "                           index=ind_list,\n",
    "                           columns = col_list)\n",
    "    df_long= pd.melt(df_wide, ignore_index=False, value_name=score_lab)\n",
    "    df_long= df_long.reset_index().sort_values(by=score_lab, ascending=False)\n",
    "    top_fifty_byscore_df = df_long.head(50)\n",
    "    top_five_perclust_df = df_long.groupby(\"cluster_number\").head(5)\n",
    "    score_dict = {top_five_lab:top_five_perclust_df, top_fifty_lab:top_fifty_byscore_df}\n",
    "    path_tops_dict[lab] = {OR_lab:OR_dict, score_lab:score_dict}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pap': {'Odds_Ratio': {'top_five':                                                  pathway     bin    Odds_Ratio\n",
       "   13762                                wild and pearl mice  pc_2_0  31570.000000\n",
       "   13763                                       wild vs pale  pc_2_0  31546.000000\n",
       "   23909                            Basal 2 control Vs COPD  pc_2_9    203.551020\n",
       "   2672                                         CancerGenes  pc_1_2    201.808081\n",
       "   22864  Genes associated with osteonecrosis in ALL chi...  pc_2_8    159.752000\n",
       "   ...                                                  ...     ...           ...\n",
       "   18821  557gene litte list, dataset: c31f4fef2910f4f50...  pc_2_5     12.088415\n",
       "   19205  Inflammatory genes with most significant isofo...  pc_2_5     11.050195\n",
       "   20025                                            wodecuo  pc_2_5     10.988914\n",
       "   18155                  Ribavirin-Down Sars-Cov-2-S7-1-Up  pc_2_4     10.802832\n",
       "   17704             CircRNA in mouse developmental tissues  pc_2_4     10.557867\n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin    Odds_Ratio\n",
       "   13762                                wild and pearl mice  pc_2_0  31570.000000\n",
       "   13763                                       wild vs pale  pc_2_0  31546.000000\n",
       "   23909                            Basal 2 control Vs COPD  pc_2_9    203.551020\n",
       "   2672                                         CancerGenes  pc_1_2    201.808081\n",
       "   22864  Genes associated with osteonecrosis in ALL chi...  pc_2_8    159.752000\n",
       "   22626                            Am Not Sure Ture Or Not  pc_2_8    153.600000\n",
       "   14342                PamGene significative kinases, CvsT  pc_2_1    138.708333\n",
       "   24848                                           lungdown  pc_2_9    110.594444\n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8    105.206434\n",
       "   3086                           Predicted Cancer Gene Set  pc_1_2     90.763636\n",
       "   1293   557gene litte list, dataset: c31f4fef2910f4f50...  pc_1_1     90.759091\n",
       "   23630                            mouse aging anti genage  pc_2_8     88.662222\n",
       "   13218  TCGA EYA1 high related genes in BRCA, dataset:...  pc_2_0     83.737288\n",
       "   15862                                      child obesity  pc_2_2     83.200000\n",
       "   24256                                               Lung  pc_2_9     81.454918\n",
       "   14422                                       SCD survival  pc_2_1     77.996094\n",
       "   2834   Genes closest to the lead SNPs from Ben Neale'...  pc_1_2     76.676923\n",
       "   24633                                          clay dawn  pc_2_9     76.573674\n",
       "   14799                                       inflammation  pc_2_1     75.992381\n",
       "   22866  Genes closest to the lead SNPs from Ben Neale'...  pc_2_8     75.249057\n",
       "   1113                                      normalized Als  pc_1_0     73.933333\n",
       "   14918                       plts-Sil-tg only, common 1+4  pc_2_1     73.404412\n",
       "   14704                                         everolimus  pc_2_1     65.671053\n",
       "   15581  Ovarian cancer genes, dataset: 5041765f6f25416...  pc_2_2     62.757353\n",
       "   13772                       +E2, CtS +E2, si10 -E2, si10  pc_2_1     62.384375\n",
       "   16136                                     network of WAS  pc_2_2     62.384375\n",
       "   12804                                       For Compare4  pc_2_0     62.284810\n",
       "   14207                                            KHAZAYI  pc_2_1     61.829457\n",
       "   23345                                           b16 down  pc_2_8     61.320000\n",
       "   2090                                       child obesity  pc_1_1     60.490909\n",
       "   12643  Biogrid human PPI interaction cluterized using...  pc_2_0     59.054054\n",
       "   995         heat map Clust3 genes up with ROSI treatment  pc_1_0     57.011429\n",
       "   21996                           TNBC vs HCC Common genes  pc_2_7     54.972452\n",
       "   21511  Depending on gender, these genes can have vary...  pc_2_7     54.972452\n",
       "   15910                      dentate gyrus gene expression  pc_2_2     54.239130\n",
       "   13567  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_0     52.186170\n",
       "   12566                                8A-WAvsWW upChart-3  pc_2_0     51.765172\n",
       "   24782   heart, dataset: 299bdebd46d940223ab8ad8b6e3afa0f  pc_2_9     50.832232\n",
       "   15851                                cad&sleep gene list  pc_2_2     49.895000\n",
       "   24654                                          cove dawn  pc_2_9     49.143223\n",
       "   14853                    miRNA-regulated in CD4+ T cells  pc_2_1     47.973558\n",
       "   5962                 gene set of breast cancer colombian  pc_1_4     46.610526\n",
       "   13276         Up in vitamin C in endothelial cells human  pc_2_0     44.553531\n",
       "   14965  si4 attached+si4 spheres vs siCNTR attached +s...  pc_2_1     43.004310\n",
       "   14779                                      hsa-let-7f-5p  pc_2_1     42.840860\n",
       "   3177                                       Schizophrenia  pc_1_2     42.167019\n",
       "   22855         Gene set of Glycolysis and TCA cycle genes  pc_2_8     41.892632\n",
       "   15966                  follicular lymphoma mutated genes  pc_2_2     41.568750\n",
       "   15079        ADAD vs control exitarory cells upregulater  pc_2_2     40.225806\n",
       "   3569   mRNAs Blancos de hsa-miR-100-5p predichos por ...  pc_1_2     39.411067},\n",
       "  'combined_score': {'top_five':                                                  pathway     bin  \\\n",
       "   13762                                wild and pearl mice  pc_2_0   \n",
       "   13763                                       wild vs pale  pc_2_0   \n",
       "   18575  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_4   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "   ...                                                  ...     ...   \n",
       "   18254  The genes about major depressive disorder anal...  pc_2_4   \n",
       "   16339                                  ALX1 target genes  pc_2_3   \n",
       "   19205  Inflammatory genes with most significant isofo...  pc_2_5   \n",
       "   19091                                        Gdgtdrhyjgk  pc_2_5   \n",
       "   19104                  Genes Involved In Prostate Cancer  pc_2_5   \n",
       "   \n",
       "          combined_score  \n",
       "   13762    98320.258821  \n",
       "   13763    98066.126316  \n",
       "   18575     2754.675440  \n",
       "   9811      1549.385299  \n",
       "   23583     1391.116257  \n",
       "   ...               ...  \n",
       "   18254       47.588304  \n",
       "   16339       46.472760  \n",
       "   19205       45.303269  \n",
       "   19091       40.485233  \n",
       "   19104       39.977327  \n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin  \\\n",
       "   13762                                wild and pearl mice  pc_2_0   \n",
       "   13763                                       wild vs pale  pc_2_0   \n",
       "   18575  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_4   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "   8559   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_6   \n",
       "   21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "   19827  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_5   \n",
       "   4803   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_3   \n",
       "   2672                                         CancerGenes  pc_1_2   \n",
       "   23909                            Basal 2 control Vs COPD  pc_2_9   \n",
       "   6055   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_4   \n",
       "   22864  Genes associated with osteonecrosis in ALL chi...  pc_2_8   \n",
       "   17323  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_3   \n",
       "   22626                            Am Not Sure Ture Or Not  pc_2_8   \n",
       "   14342                PamGene significative kinases, CvsT  pc_2_1   \n",
       "   15581  Ovarian cancer genes, dataset: 5041765f6f25416...  pc_2_2   \n",
       "   2834   Genes closest to the lead SNPs from Ben Neale'...  pc_1_2   \n",
       "   14799                                       inflammation  pc_2_1   \n",
       "   7307   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_5   \n",
       "   2299   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_1   \n",
       "   24633                                          clay dawn  pc_2_9   \n",
       "   24848                                           lungdown  pc_2_9   \n",
       "   14207                                            KHAZAYI  pc_2_1   \n",
       "   3086                           Predicted Cancer Gene Set  pc_1_2   \n",
       "   1293   557gene litte list, dataset: c31f4fef2910f4f50...  pc_1_1   \n",
       "   23630                            mouse aging anti genage  pc_2_8   \n",
       "   15862                                      child obesity  pc_2_2   \n",
       "   14422                                       SCD survival  pc_2_1   \n",
       "   24256                                               Lung  pc_2_9   \n",
       "   5962                 gene set of breast cancer colombian  pc_1_4   \n",
       "   13218  TCGA EYA1 high related genes in BRCA, dataset:...  pc_2_0   \n",
       "   22866  Genes closest to the lead SNPs from Ben Neale'...  pc_2_8   \n",
       "   1113                                      normalized Als  pc_1_0   \n",
       "   14918                       plts-Sil-tg only, common 1+4  pc_2_1   \n",
       "   1047   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_0   \n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   14779                                      hsa-let-7f-5p  pc_2_1   \n",
       "   24782   heart, dataset: 299bdebd46d940223ab8ad8b6e3afa0f  pc_2_9   \n",
       "   14704                                         everolimus  pc_2_1   \n",
       "   24654                                          cove dawn  pc_2_9   \n",
       "   3551   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_2   \n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "   18366                                      child obesity  pc_2_4   \n",
       "   13772                       +E2, CtS +E2, si10 -E2, si10  pc_2_1   \n",
       "   16136                                     network of WAS  pc_2_2   \n",
       "   16071  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_2   \n",
       "   14891                                     oct2Open peaks  pc_2_1   \n",
       "   2090                                       child obesity  pc_1_1   \n",
       "   23345                                           b16 down  pc_2_8   \n",
       "   \n",
       "          combined_score  \n",
       "   13762    98320.258821  \n",
       "   13763    98066.126316  \n",
       "   18575     2754.675440  \n",
       "   9811      1549.385299  \n",
       "   23583     1391.116257  \n",
       "   8559      1256.503023  \n",
       "   21079     1249.096139  \n",
       "   19827     1178.673349  \n",
       "   4803      1043.932090  \n",
       "   2672      1032.951295  \n",
       "   23909      996.446358  \n",
       "   6055       921.656680  \n",
       "   22864      775.878398  \n",
       "   17323      740.603517  \n",
       "   22626      740.221819  \n",
       "   14342      661.819014  \n",
       "   15581      657.616217  \n",
       "   2834       588.522912  \n",
       "   14799      586.460626  \n",
       "   7307       585.834724  \n",
       "   2299       538.142108  \n",
       "   24633      478.942718  \n",
       "   24848      475.393363  \n",
       "   14207      452.897086  \n",
       "   3086       397.505318  \n",
       "   1293       397.125975  \n",
       "   23630      380.247469  \n",
       "   15862      358.067161  \n",
       "   14422      330.973809  \n",
       "   24256      325.722363  \n",
       "   5962       316.243197  \n",
       "   13218      313.864514  \n",
       "   22866      310.731701  \n",
       "   1113       309.051915  \n",
       "   14918      307.322787  \n",
       "   1047       306.723375  \n",
       "   11063      297.994722  \n",
       "   14779      283.746193  \n",
       "   24782      278.113419  \n",
       "   14704      268.078772  \n",
       "   24654      265.727264  \n",
       "   3551       258.101726  \n",
       "   22331      254.235607  \n",
       "   18366      253.539705  \n",
       "   13772      251.643234  \n",
       "   16136      251.643234  \n",
       "   16071      249.071882  \n",
       "   14891      246.914768  \n",
       "   2090       242.184823  \n",
       "   23345      241.000152  },\n",
       "  'adjusted_score': {'top_five':                                                  pathway     bin  \\\n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "   23909                            Basal 2 control Vs COPD  pc_2_9   \n",
       "   22864  Genes associated with osteonecrosis in ALL chi...  pc_2_8   \n",
       "   22626                            Am Not Sure Ture Or Not  pc_2_8   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   ...                                                  ...     ...   \n",
       "   2268                                 hubs gene alzahimer  pc_1_1   \n",
       "   2248                                heat shock for tumor  pc_1_1   \n",
       "   2249                                          hela cells  pc_1_1   \n",
       "   2250                                           hgfdxgfgh  pc_1_1   \n",
       "   2251                      highly potent targetable genes  pc_1_1   \n",
       "   \n",
       "          adjusted_score  \n",
       "   23583    10653.863857  \n",
       "   23909     9744.747156  \n",
       "   22864     5942.064712  \n",
       "   22626     5668.988802  \n",
       "   9811      5401.157154  \n",
       "   ...               ...  \n",
       "   2268         0.000000  \n",
       "   2248         0.000000  \n",
       "   2249         0.000000  \n",
       "   2250         0.000000  \n",
       "   2251         0.000000  \n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin  \\\n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "   23909                            Basal 2 control Vs COPD  pc_2_9   \n",
       "   22864  Genes associated with osteonecrosis in ALL chi...  pc_2_8   \n",
       "   22626                            Am Not Sure Ture Or Not  pc_2_8   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   24633                                          clay dawn  pc_2_9   \n",
       "   24848                                           lungdown  pc_2_9   \n",
       "   21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "   24256                                               Lung  pc_2_9   \n",
       "   23630                            mouse aging anti genage  pc_2_8   \n",
       "   24782   heart, dataset: 299bdebd46d940223ab8ad8b6e3afa0f  pc_2_9   \n",
       "   24654                                          cove dawn  pc_2_9   \n",
       "   22866  Genes closest to the lead SNPs from Ben Neale'...  pc_2_8   \n",
       "   8559   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_6   \n",
       "   23345                                           b16 down  pc_2_8   \n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   19827  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_5   \n",
       "   11994  The genes about major depressive disorder anal...  pc_1_9   \n",
       "   24696  downregulated, dataset: 6d5ee3258f0d93c3f9f758...  pc_2_9   \n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "   24889                       mouse cells, KO target genes  pc_2_9   \n",
       "   21996                           TNBC vs HCC Common genes  pc_2_7   \n",
       "   21511  Depending on gender, these genes can have vary...  pc_2_7   \n",
       "   24647      common genes from iCLIP and RNA-seq data sets  pc_2_9   \n",
       "   22855         Gene set of Glycolysis and TCA cycle genes  pc_2_8   \n",
       "   24103               Gene expression between wt and ube3a  pc_2_9   \n",
       "   23510                                     h3k27 promoter  pc_2_8   \n",
       "   12315  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_9   \n",
       "   24591                                          arch dawn  pc_2_9   \n",
       "   23425                                    diabetes type 1  pc_2_8   \n",
       "   23172                                      SALL4 related  pc_2_8   \n",
       "   22950                                 IDH DOWNREGULATION  pc_2_8   \n",
       "   24209       Identified by the KroganLab in 2020 (nature)  pc_2_9   \n",
       "   22675      CA1-TRAP vs BG TRAP RNAseq upregulated in CA1  pc_2_8   \n",
       "   11392  Biological insights from 108 schizophrenia-ass...  pc_1_9   \n",
       "   23687                              pre susceptible genes  pc_2_8   \n",
       "   24902  novel genes and significant in both sift and p...  pc_2_9   \n",
       "   12363                           mouse aging anti gomperz  pc_1_9   \n",
       "   23486                         genage all bedoone tekrari  pc_2_8   \n",
       "   11968                                        TF OF H3.2C  pc_1_9   \n",
       "   24619                        capan1 downregulated p<0.05  pc_2_9   \n",
       "   10340                  Genes Involved In Prostate Cancer  pc_1_8   \n",
       "   9602                                       child obesity  pc_1_7   \n",
       "   22486                                 spinal cord injury  pc_2_7   \n",
       "   23545                                     hsa-miR-33a-5p  pc_2_8   \n",
       "   23055                                             Muscle  pc_2_8   \n",
       "   22876                                   Genes in MDK sig  pc_2_8   \n",
       "   24238                             LIN28 shRNA microarray  pc_2_9   \n",
       "   11577                                      GWAS AD GENES  pc_1_9   \n",
       "   21284                       +E2, CtS +E2, si10 -E2, si10  pc_2_7   \n",
       "   \n",
       "          adjusted_score  \n",
       "   23583    10653.863857  \n",
       "   23909     9744.747156  \n",
       "   22864     5942.064712  \n",
       "   22626     5668.988802  \n",
       "   9811      5401.157154  \n",
       "   24633     4683.820306  \n",
       "   24848     4649.109394  \n",
       "   21079     4268.786054  \n",
       "   24256     3185.401848  \n",
       "   23630     2912.125244  \n",
       "   24782     2719.810186  \n",
       "   24654     2598.679774  \n",
       "   22866     2379.738734  \n",
       "   8559      1954.490452  \n",
       "   23345     1845.699666  \n",
       "   11063     1614.088413  \n",
       "   19827     1528.149997  \n",
       "   11994     1504.321180  \n",
       "   24696     1475.225786  \n",
       "   22331     1407.956792  \n",
       "   24889     1245.247555  \n",
       "   21996     1187.700600  \n",
       "   21511     1187.700600  \n",
       "   24647     1160.374247  \n",
       "   22855     1141.933502  \n",
       "   24103     1100.344040  \n",
       "   23510     1088.905715  \n",
       "   12315     1074.635213  \n",
       "   24591     1017.349542  \n",
       "   23425      980.150738  \n",
       "   23172      903.101141  \n",
       "   22950      892.943712  \n",
       "   24209      876.304909  \n",
       "   22675      856.487436  \n",
       "   11392      845.635593  \n",
       "   23687      809.693833  \n",
       "   24902      806.696070  \n",
       "   12363      786.567206  \n",
       "   23486      731.704215  \n",
       "   11968      720.329197  \n",
       "   24619      706.956102  \n",
       "   10340      697.527666  \n",
       "   9602       694.071264  \n",
       "   22486      689.404758  \n",
       "   23545      665.734352  \n",
       "   23055      653.738137  \n",
       "   22876      642.114516  \n",
       "   24238      596.923596  \n",
       "   11577      574.694516  \n",
       "   21284      560.519016  }},\n",
       " 'ours': {'Odds_Ratio': {'top_five':                                                  pathway     bin  Odds_Ratio\n",
       "   5235   Depending on gender, these genes can have vary...  pc_1_4  227.045455\n",
       "   11838                PamGene significative kinases, CvsT  pc_1_9  130.542484\n",
       "   5237                      Diabetic retinopathy Metformin  pc_1_4   83.170833\n",
       "   12641                            Basal 2 control Vs COPD  pc_2_0   81.408163\n",
       "   6588   Genes associated with osteonecrosis in ALL chi...  pc_1_5   79.856000\n",
       "   ...                                                  ...     ...         ...\n",
       "   18575  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_4   14.244772\n",
       "   9342   Phosphoproteomics platelets Perseus ttest: P<0...  pc_1_7   13.899301\n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   13.727779\n",
       "   10311                                           GES21520  pc_1_8   13.241333\n",
       "   9859                            mouse aging anti gomperz  pc_1_7   12.905844\n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin  Odds_Ratio\n",
       "   5235   Depending on gender, these genes can have vary...  pc_1_4  227.045455\n",
       "   11838                PamGene significative kinases, CvsT  pc_1_9  130.542484\n",
       "   5237                      Diabetic retinopathy Metformin  pc_1_4   83.170833\n",
       "   12641                            Basal 2 control Vs COPD  pc_2_0   81.408163\n",
       "   6588   Genes associated with osteonecrosis in ALL chi...  pc_1_5   79.856000\n",
       "   12471                           spinal cord injury model  pc_1_9   78.301961\n",
       "   11598  Genes closest to the lead SNPs from Ben Neale'...  pc_1_9   78.160784\n",
       "   6350                             Am Not Sure Ture Or Not  pc_1_5   76.780769\n",
       "   12470                                 spinal cord injury  pc_1_9   69.083045\n",
       "   12890         HMGB1-interacted proteins pathway analysis  pc_2_0   67.576271\n",
       "   12200                                         everolimus  pc_1_9   61.804954\n",
       "   5854     closest cardiac imaging genes Pirruccello et al  pc_1_4   59.372024\n",
       "   11268                       +E2, CtS +E2, si10 -E2, si10  pc_1_9   58.711765\n",
       "   4594                                       child obesity  pc_1_3   58.119534\n",
       "   5117          BIOID mass spec resuts of M4K4 interactors  pc_1_4   57.988372\n",
       "   14867                           mouse aging anti gomperz  pc_2_1   57.028571\n",
       "   6102                             mouse aging anti genage  pc_1_4   55.405556\n",
       "   14431                                     SLE risk genes  pc_2_1   53.610215\n",
       "   5897                                     diabetes type 1  pc_1_4   53.598383\n",
       "   14390  Rat proximal tubule TF network identified by wTO.  pc_2_1   51.932292\n",
       "   23964             CircRNA in mouse developmental tissues  pc_2_9   51.105128\n",
       "   5413          Human TREM2- CTRL vs TREM2+ CTRL TOP 50 UP  pc_1_4   50.872449\n",
       "   15192                                        CancerGenes  pc_2_2   50.368687\n",
       "   5048                                      54DNs 1 YapTaz  pc_1_4   49.852500\n",
       "   13133  RPE reprogramming, dataset: 2185b5e32d0c3edf0c...  pc_2_0   47.978313\n",
       "   12095                                cad&sleep gene list  pc_1_9   46.957647\n",
       "   23268  Top 10 DEG genes between benign and malign pro...  pc_2_8   45.217687\n",
       "   12349                    miRNA-regulated in CD4+ T cells  pc_1_9   45.149321\n",
       "   7354                             mouse aging anti genage  pc_1_5   44.320000\n",
       "   13104                           Prlr +/- from beta cells  pc_2_0   42.340426\n",
       "   6055   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_4   42.076139\n",
       "   6381                             Basal 2 control Vs COPD  pc_1_5   40.693878\n",
       "   12461  si4 attached+si4 spheres vs siCNTR attached +s...  pc_1_9   40.472617\n",
       "   13824                                      AD candidates  pc_2_1   39.908000\n",
       "   5320                          Gene List Hot Spot Project  pc_1_4   39.378968\n",
       "   2299   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_1   38.642282\n",
       "   7839   Genes associated with T1D from the genomics co...  pc_1_6   37.770833\n",
       "   8139                   Ribavirin-Down Sars-Cov-2-S7-1-Up  pc_1_6   37.770833\n",
       "   5923   e15 limb chirp seq lncRNA top 90 percent coinc...  pc_1_4   37.736742\n",
       "   13409                                    diabetes type 1  pc_2_0   37.171963\n",
       "   13379      common genes from iCLIP and RNA-seq data sets  pc_2_0   36.941948\n",
       "   11298  34 genes associated with Chronic Pain from Tab...  pc_1_9   35.559715\n",
       "   12864                                       Genes of IMF  pc_2_0   33.977778\n",
       "   6630          HMGB1-interacted proteins pathway analysis  pc_1_5   33.779661\n",
       "   12263       heat map Clust3 genes up with ROSI treatment  pc_1_9   33.524370\n",
       "   2030                                           XP8GY-12H  pc_1_1   33.215000\n",
       "   6404                                       CD161-/CD161+  pc_1_5   32.668852\n",
       "   687                   Sofosbuvir-Down Sars-Cov-2-S7-1-Up  pc_1_0   32.575163\n",
       "   17114                                      child obesity  pc_2_3   32.298701\n",
       "   14100  Genes associated with osteonecrosis in ALL chi...  pc_2_1   31.918400},\n",
       "  'combined_score': {'top_five':                                                  pathway     bin  \\\n",
       "   2299   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_1   \n",
       "   17323  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_3   \n",
       "   5235   Depending on gender, these genes can have vary...  pc_1_4   \n",
       "   19827  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_5   \n",
       "   16071  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_2   \n",
       "   ...                                                  ...     ...   \n",
       "   10905                                    diabetes type 1  pc_1_8   \n",
       "   18540  https://science.sciencemag.org/content/365/645...  pc_2_4   \n",
       "   10748  Top 10 DEG genes between benign and malign pro...  pc_1_8   \n",
       "   18268                                  Two different KOs  pc_2_4   \n",
       "   24011                                     Darkblue(High)  pc_2_9   \n",
       "   \n",
       "          combined_score  \n",
       "   2299      2383.501000  \n",
       "   17323     1687.179235  \n",
       "   5235      1185.987693  \n",
       "   19827     1115.516683  \n",
       "   16071     1112.774712  \n",
       "   ...               ...  \n",
       "   10905       58.731468  \n",
       "   18540       52.974091  \n",
       "   10748       49.312541  \n",
       "   18268       46.097000  \n",
       "   24011       45.945026  \n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin  \\\n",
       "   2299   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_1   \n",
       "   17323  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_3   \n",
       "   5235   Depending on gender, these genes can have vary...  pc_1_4   \n",
       "   19827  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_5   \n",
       "   16071  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_2   \n",
       "   21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   3551   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_2   \n",
       "   11598  Genes closest to the lead SNPs from Ben Neale'...  pc_1_9   \n",
       "   8559   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_6   \n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "   11838                PamGene significative kinases, CvsT  pc_1_9   \n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   18575  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_4   \n",
       "   4803   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_3   \n",
       "   6055   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_4   \n",
       "   4594                                       child obesity  pc_1_3   \n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "   14431                                     SLE risk genes  pc_2_1   \n",
       "   5897                                     diabetes type 1  pc_1_4   \n",
       "   14390  Rat proximal tubule TF network identified by wTO.  pc_2_1   \n",
       "   1047   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_0   \n",
       "   5237                      Diabetic retinopathy Metformin  pc_1_4   \n",
       "   23964             CircRNA in mouse developmental tissues  pc_2_9   \n",
       "   12641                            Basal 2 control Vs COPD  pc_2_0   \n",
       "   6588   Genes associated with osteonecrosis in ALL chi...  pc_1_5   \n",
       "   12471                           spinal cord injury model  pc_1_9   \n",
       "   6350                             Am Not Sure Ture Or Not  pc_1_5   \n",
       "   12470                                 spinal cord injury  pc_1_9   \n",
       "   12890         HMGB1-interacted proteins pathway analysis  pc_2_0   \n",
       "   12200                                         everolimus  pc_1_9   \n",
       "   5320                          Gene List Hot Spot Project  pc_1_4   \n",
       "   22242  genes closest to lead SNPs from Neale analysis...  pc_2_7   \n",
       "   5854     closest cardiac imaging genes Pirruccello et al  pc_1_4   \n",
       "   11268                       +E2, CtS +E2, si10 -E2, si10  pc_1_9   \n",
       "   5117          BIOID mass spec resuts of M4K4 interactors  pc_1_4   \n",
       "   16339                                  ALX1 target genes  pc_2_3   \n",
       "   14867                           mouse aging anti gomperz  pc_2_1   \n",
       "   12315  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_9   \n",
       "   13379      common genes from iCLIP and RNA-seq data sets  pc_2_0   \n",
       "   6102                             mouse aging anti genage  pc_1_4   \n",
       "   17114                                      child obesity  pc_2_3   \n",
       "   5413          Human TREM2- CTRL vs TREM2+ CTRL TOP 50 UP  pc_1_4   \n",
       "   15192                                        CancerGenes  pc_2_2   \n",
       "   5048                                      54DNs 1 YapTaz  pc_1_4   \n",
       "   14819  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_1   \n",
       "   63                                     ALX1 target genes  pc_1_0   \n",
       "   14064                                        GE-K9-4-bed  pc_2_1   \n",
       "   6779                                              Muscle  pc_1_5   \n",
       "   13133  RPE reprogramming, dataset: 2185b5e32d0c3edf0c...  pc_2_0   \n",
       "   \n",
       "          combined_score  \n",
       "   2299      2383.501000  \n",
       "   17323     1687.179235  \n",
       "   5235      1185.987693  \n",
       "   19827     1115.516683  \n",
       "   16071     1112.774712  \n",
       "   21079     1112.760229  \n",
       "   9811      1008.030640  \n",
       "   3551       970.661202  \n",
       "   11598      869.471496  \n",
       "   8559       744.082054  \n",
       "   22331      715.949354  \n",
       "   11838      615.425022  \n",
       "   11063      569.115377  \n",
       "   18575      533.155575  \n",
       "   4803       471.036572  \n",
       "   6055       468.707158  \n",
       "   4594       418.371875  \n",
       "   23583      388.904326  \n",
       "   14431      379.864290  \n",
       "   5897       369.465372  \n",
       "   14390      364.868168  \n",
       "   1047       361.426390  \n",
       "   5237       355.828339  \n",
       "   23964      355.743323  \n",
       "   12641      342.389562  \n",
       "   6588       339.687615  \n",
       "   12471      332.541203  \n",
       "   6350       323.727778  \n",
       "   12470      285.311114  \n",
       "   12890      271.978647  \n",
       "   12200      248.793415  \n",
       "   5320       247.958031  \n",
       "   22242      235.786413  \n",
       "   5854       234.725283  \n",
       "   11268      233.502245  \n",
       "   5117       227.933516  \n",
       "   16339      225.776110  \n",
       "   14867      225.039974  \n",
       "   12315      223.617983  \n",
       "   13379      219.763896  \n",
       "   6102       215.340551  \n",
       "   17114      196.119266  \n",
       "   5413       193.520870  \n",
       "   15192      191.609531  \n",
       "   5048       188.663684  \n",
       "   14819      187.676075  \n",
       "   63         184.884133  \n",
       "   14064      182.768292  \n",
       "   6779       178.491985  \n",
       "   13133      177.101900  },\n",
       "  'adjusted_score': {'top_five':                                                  pathway     bin  \\\n",
       "   11598  Genes closest to the lead SNPs from Ben Neale'...  pc_1_9   \n",
       "   11838                PamGene significative kinases, CvsT  pc_1_9   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   12471                           spinal cord injury model  pc_1_9   \n",
       "   ...                                                  ...     ...   \n",
       "   3429                                   endothelial cells  pc_1_2   \n",
       "   3631                                         p53 targets  pc_1_2   \n",
       "   3532                                        injured treg  pc_1_2   \n",
       "   3535                                         interesting  pc_1_2   \n",
       "   3537                                           just down  pc_1_2   \n",
       "   \n",
       "          adjusted_score  \n",
       "   11598    11483.544778  \n",
       "   11838     8128.225981  \n",
       "   9811      7404.993081  \n",
       "   11063     5848.514176  \n",
       "   12471     4392.037941  \n",
       "   ...               ...  \n",
       "   3429         0.000000  \n",
       "   3631         0.000000  \n",
       "   3532         0.000000  \n",
       "   3535         0.000000  \n",
       "   3537         0.000000  \n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin  \\\n",
       "   11598  Genes closest to the lead SNPs from Ben Neale'...  pc_1_9   \n",
       "   11838                PamGene significative kinases, CvsT  pc_1_9   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   12471                           spinal cord injury model  pc_1_9   \n",
       "   12470                                 spinal cord injury  pc_1_9   \n",
       "   12200                                         everolimus  pc_1_9   \n",
       "   8559   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_6   \n",
       "   11268                       +E2, CtS +E2, si10 -E2, si10  pc_1_9   \n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "   12315  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_9   \n",
       "   23964             CircRNA in mouse developmental tissues  pc_2_9   \n",
       "   21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "   12095                                cad&sleep gene list  pc_1_9   \n",
       "   12349                    miRNA-regulated in CD4+ T cells  pc_1_9   \n",
       "   12461  si4 attached+si4 spheres vs siCNTR attached +s...  pc_1_9   \n",
       "   11298  34 genes associated with Chronic Pain from Tab...  pc_1_9   \n",
       "   12263       heat map Clust3 genes up with ROSI treatment  pc_1_9   \n",
       "   10580                                     PCa signatures  pc_1_8   \n",
       "   11640                                         HO36 group  pc_1_9   \n",
       "   11847                                       Pi Gene list  pc_1_9   \n",
       "   10179                                         Cac vs. WT  pc_1_8   \n",
       "   23268  Top 10 DEG genes between benign and malign pro...  pc_2_8   \n",
       "   22242  genes closest to lead SNPs from Neale analysis...  pc_2_7   \n",
       "   12083                     bladder cancer, dataset: 7w0x7  pc_1_9   \n",
       "   11973                         THP-MVs Sil CvsCL down-sig  pc_1_9   \n",
       "   12460                          shortened in laminB1 over  pc_1_9   \n",
       "   12485                              triple dominant group  pc_1_9   \n",
       "   12316                            lengthened laminB1 over  pc_1_9   \n",
       "   9569                   autism confidence gene from SFARI  pc_1_7   \n",
       "   12239                                     gfp as control  pc_1_9   \n",
       "   23897         BIOID mass spec resuts of M4K4 interactors  pc_2_9   \n",
       "   11638         HMGB1-interacted proteins pathway analysis  pc_1_9   \n",
       "   9776   https://science.sciencemag.org/content/365/645...  pc_1_7   \n",
       "   12275                                      hsa-let-7f-5p  pc_1_9   \n",
       "   11773                                          Microarry  pc_1_9   \n",
       "   11272  10 percent of the longest transcripts of the h...  pc_1_9   \n",
       "   9584                                         brain pleio  pc_1_7   \n",
       "   10905                                    diabetes type 1  pc_1_8   \n",
       "   24118  Genes closest to the lead SNPs from Ben Neale'...  pc_2_9   \n",
       "   22983                                     LFQ-upCtrl 1.5  pc_2_8   \n",
       "   11698            Intersection Among Different Timepoints  pc_1_9   \n",
       "   8139                   Ribavirin-Down Sars-Cov-2-S7-1-Up  pc_1_6   \n",
       "   7839   Genes associated with T1D from the genomics co...  pc_1_6   \n",
       "   23700                                      ptx2 pulldown  pc_2_8   \n",
       "   12156                                        depresssion  pc_1_9   \n",
       "   12413                      plts-Sil-WTonly Common in 2+3  pc_1_9   \n",
       "   23374                                      child obesity  pc_2_8   \n",
       "   23739                           spinal cord injury model  pc_2_8   \n",
       "   \n",
       "          adjusted_score  \n",
       "   11598    11483.544778  \n",
       "   11838     8128.225981  \n",
       "   9811      7404.993081  \n",
       "   11063     5848.514176  \n",
       "   12471     4392.037941  \n",
       "   12470     3768.246535  \n",
       "   12200     3285.939025  \n",
       "   8559      3285.494307  \n",
       "   11268     3083.980900  \n",
       "   22331     3039.920956  \n",
       "   12315     2953.434504  \n",
       "   23964     2906.778690  \n",
       "   21079     2541.544362  \n",
       "   23583     2414.512507  \n",
       "   12095     2335.424208  \n",
       "   12349     2223.235490  \n",
       "   12461     1937.305872  \n",
       "   11298     1644.151922  \n",
       "   12263     1525.111766  \n",
       "   10580     1403.407993  \n",
       "   11640     1284.584725  \n",
       "   11847     1269.157851  \n",
       "   10179     1048.714219  \n",
       "   23268     1038.688334  \n",
       "   22242     1001.149111  \n",
       "   12083      960.867594  \n",
       "   11973      936.212423  \n",
       "   12460      890.014995  \n",
       "   12485      888.198049  \n",
       "   12316      808.444006  \n",
       "   9569       796.254526  \n",
       "   12239      790.015973  \n",
       "   23897      782.265999  \n",
       "   11638      772.288423  \n",
       "   9776       733.913562  \n",
       "   12275      707.670704  \n",
       "   11773      645.207967  \n",
       "   11272      622.132303  \n",
       "   9584       609.717166  \n",
       "   10905      603.553934  \n",
       "   24118      595.786656  \n",
       "   22983      595.505140  \n",
       "   11698      591.480426  \n",
       "   8139       591.235830  \n",
       "   7839       591.235830  \n",
       "   23700      565.069723  \n",
       "   12156      559.946034  \n",
       "   12413      559.691523  \n",
       "   23374      545.114590  \n",
       "   23739      545.114590  }},\n",
       " 'ours_with_exp': {'Odds_Ratio': {'top_five':                                                  pathway     bin  Odds_Ratio\n",
       "   6487   Depending on gender, these genes can have vary...  pc_1_5  227.045455\n",
       "   570                  PamGene significative kinases, CvsT  pc_1_0   96.458937\n",
       "   5336   Genes associated with osteonecrosis in ALL chi...  pc_1_4   88.733333\n",
       "   5098                             Am Not Sure Ture Or Not  pc_1_4   85.316239\n",
       "   6489                      Diabetic retinopathy Metformin  pc_1_5   83.170833\n",
       "   ...                                                  ...     ...         ...\n",
       "   3131                   Ribavirin-Down Sars-Cov-2-S7-1-Up  pc_1_2   14.275144\n",
       "   19778                                          hgfdxgfgh  pc_2_5   14.027523\n",
       "   1547                                            GES21520  pc_1_1   13.796528\n",
       "   19827  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_5   13.757042\n",
       "   3082   Phosphoproteomics platelets Perseus ttest: P<0...  pc_1_2   13.176393\n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin  Odds_Ratio\n",
       "   6487   Depending on gender, these genes can have vary...  pc_1_5  227.045455\n",
       "   570                  PamGene significative kinases, CvsT  pc_1_0   96.458937\n",
       "   5336   Genes associated with osteonecrosis in ALL chi...  pc_1_4   88.733333\n",
       "   5098                             Am Not Sure Ture Or Not  pc_1_4   85.316239\n",
       "   6489                      Diabetic retinopathy Metformin  pc_1_5   83.170833\n",
       "   23909                            Basal 2 control Vs COPD  pc_2_9   81.408163\n",
       "   24158         HMGB1-interacted proteins pathway analysis  pc_2_9   67.576271\n",
       "   8350                                       child obesity  pc_1_6   63.298413\n",
       "   7106     closest cardiac imaging genes Pirruccello et al  pc_1_5   59.372024\n",
       "   6369          BIOID mass spec resuts of M4K4 interactors  pc_1_5   57.988372\n",
       "   1203                            spinal cord injury model  pc_1_0   57.857971\n",
       "   23631                           mouse aging anti gomperz  pc_2_8   57.028571\n",
       "   330    Genes closest to the lead SNPs from Ben Neale'...  pc_1_0   55.812325\n",
       "   7354                             mouse aging anti genage  pc_1_5   55.405556\n",
       "   23195                                     SLE risk genes  pc_2_8   53.610215\n",
       "   7149                                     diabetes type 1  pc_1_5   53.598383\n",
       "   23154  Rat proximal tubule TF network identified by wTO.  pc_2_8   51.932292\n",
       "   12696             CircRNA in mouse developmental tissues  pc_2_0   51.105128\n",
       "   1202                                  spinal cord injury  pc_1_0   51.046036\n",
       "   6665          Human TREM2- CTRL vs TREM2+ CTRL TOP 50 UP  pc_1_5   50.872449\n",
       "   6300                                      54DNs 1 YapTaz  pc_1_5   49.852500\n",
       "   6102                             mouse aging anti genage  pc_1_4   49.246914\n",
       "   21452                                        CancerGenes  pc_2_7   49.246914\n",
       "   24401  RPE reprogramming, dataset: 2185b5e32d0c3edf0c...  pc_2_9   47.978313\n",
       "   932                                           everolimus  pc_1_0   45.668192\n",
       "   5129                             Basal 2 control Vs COPD  pc_1_4   45.217687\n",
       "   14504  Top 10 DEG genes between benign and malign pro...  pc_2_1   45.217687\n",
       "   4083   Genes associated with T1D from the genomics co...  pc_1_3   43.747807\n",
       "   0                           +E2, CtS +E2, si10 -E2, si10  pc_1_0   43.382609\n",
       "   24372                           Prlr +/- from beta cells  pc_2_9   42.340426\n",
       "   7307   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_5   42.076139\n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   41.972143\n",
       "   22588                                      AD candidates  pc_2_8   39.908000\n",
       "   6572                          Gene List Hot Spot Project  pc_1_5   39.378968\n",
       "   7175   e15 limb chirp seq lncRNA top 90 percent coinc...  pc_1_5   37.736742\n",
       "   5378          HMGB1-interacted proteins pathway analysis  pc_1_4   37.534840\n",
       "   24677                                    diabetes type 1  pc_2_9   37.171963\n",
       "   24647      common genes from iCLIP and RNA-seq data sets  pc_2_9   36.941948\n",
       "   4594                                       child obesity  pc_1_3   34.992982\n",
       "   827                                  cad&sleep gene list  pc_1_0   34.697391\n",
       "   5527                                              Muscle  pc_1_4   34.215517\n",
       "   24132                                       Genes of IMF  pc_2_9   33.977778\n",
       "   20870                                      child obesity  pc_2_6   33.843537\n",
       "   1081                     miRNA-regulated in CD4+ T cells  pc_1_0   33.361204\n",
       "   11955                 Sofosbuvir-Down Sars-Cov-2-S7-1-Up  pc_1_9   32.575163\n",
       "   10794                                          XP8GY-12H  pc_1_8   32.140323\n",
       "   22864  Genes associated with osteonecrosis in ALL chi...  pc_2_8   31.918400\n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   31.905939\n",
       "   6758                                            Mayocyte  pc_1_5   30.725309\n",
       "   11331                                  ALX1 target genes  pc_1_9   30.649231},\n",
       "  'combined_score': {'top_five':                                                  pathway     bin  \\\n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "   17323  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_3   \n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "   6487   Depending on gender, these genes can have vary...  pc_1_5   \n",
       "   ...                                                  ...     ...   \n",
       "   1984   Top 10 DEG genes between benign and malign pro...  pc_1_1   \n",
       "   19792  https://science.sciencemag.org/content/365/645...  pc_2_5   \n",
       "   12743                                     Darkblue(High)  pc_2_0   \n",
       "   19520                                  Two different KOs  pc_2_5   \n",
       "   3630              p1 top200 marker genes, dataset: 5py8m  pc_1_2   \n",
       "   \n",
       "          combined_score  \n",
       "   11063     2855.934408  \n",
       "   21079     1659.493161  \n",
       "   17323     1299.498324  \n",
       "   22331     1281.928606  \n",
       "   6487      1185.987693  \n",
       "   ...               ...  \n",
       "   1984        52.104532  \n",
       "   19792       49.412289  \n",
       "   12743       45.945026  \n",
       "   19520       44.245257  \n",
       "   3630        41.865275  \n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin  \\\n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "   17323  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_3   \n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "   6487   Depending on gender, these genes can have vary...  pc_1_5   \n",
       "   3551   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_2   \n",
       "   18575  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_4   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   4803   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_3   \n",
       "   2299   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_1   \n",
       "   16071  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_2   \n",
       "   330    Genes closest to the lead SNPs from Ben Neale'...  pc_1_0   \n",
       "   19827  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_5   \n",
       "   7307   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_5   \n",
       "   8350                                       child obesity  pc_1_6   \n",
       "   8559   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_6   \n",
       "   570                  PamGene significative kinases, CvsT  pc_1_0   \n",
       "   14819  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_1   \n",
       "   5336   Genes associated with osteonecrosis in ALL chi...  pc_1_4   \n",
       "   23195                                     SLE risk genes  pc_2_8   \n",
       "   7149                                     diabetes type 1  pc_1_5   \n",
       "   5098                             Am Not Sure Ture Or Not  pc_1_4   \n",
       "   23154  Rat proximal tubule TF network identified by wTO.  pc_2_8   \n",
       "   12315  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_9   \n",
       "   6489                      Diabetic retinopathy Metformin  pc_1_5   \n",
       "   12696             CircRNA in mouse developmental tissues  pc_2_0   \n",
       "   23909                            Basal 2 control Vs COPD  pc_2_9   \n",
       "   24158         HMGB1-interacted proteins pathway analysis  pc_2_9   \n",
       "   6572                          Gene List Hot Spot Project  pc_1_5   \n",
       "   20095                                  ALX1 target genes  pc_2_6   \n",
       "   7106     closest cardiac imaging genes Pirruccello et al  pc_1_5   \n",
       "   1203                            spinal cord injury model  pc_1_0   \n",
       "   6369          BIOID mass spec resuts of M4K4 interactors  pc_1_5   \n",
       "   23631                           mouse aging anti gomperz  pc_2_8   \n",
       "   24647      common genes from iCLIP and RNA-seq data sets  pc_2_9   \n",
       "   1047   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_0   \n",
       "   7354                             mouse aging anti genage  pc_1_5   \n",
       "   20870                                      child obesity  pc_2_6   \n",
       "   5527                                              Muscle  pc_1_4   \n",
       "   1202                                  spinal cord injury  pc_1_0   \n",
       "   6665          Human TREM2- CTRL vs TREM2+ CTRL TOP 50 UP  pc_1_5   \n",
       "   6300                                      54DNs 1 YapTaz  pc_1_5   \n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "   21452                                        CancerGenes  pc_2_7   \n",
       "   6102                             mouse aging anti genage  pc_1_4   \n",
       "   11331                                  ALX1 target genes  pc_1_9   \n",
       "   22828                                        GE-K9-4-bed  pc_2_8   \n",
       "   24401  RPE reprogramming, dataset: 2185b5e32d0c3edf0c...  pc_2_9   \n",
       "   15982  genes closest to lead SNPs from Neale analysis...  pc_2_2   \n",
       "   932                                           everolimus  pc_1_0   \n",
       "   \n",
       "          combined_score  \n",
       "   11063     2855.934408  \n",
       "   21079     1659.493161  \n",
       "   17323     1299.498324  \n",
       "   22331     1281.928606  \n",
       "   6487      1185.987693  \n",
       "   3551      1119.191860  \n",
       "   18575     1089.789244  \n",
       "   9811       838.272542  \n",
       "   4803       625.174035  \n",
       "   2299       613.940419  \n",
       "   16071      596.076121  \n",
       "   330        570.802923  \n",
       "   19827      506.303941  \n",
       "   7307       468.707158  \n",
       "   8350       465.981899  \n",
       "   8559       436.878970  \n",
       "   570        427.123227  \n",
       "   14819      388.904326  \n",
       "   5336       385.851421  \n",
       "   23195      379.864290  \n",
       "   7149       369.465372  \n",
       "   5098       367.791702  \n",
       "   23154      364.868168  \n",
       "   12315      361.426390  \n",
       "   6489       355.828339  \n",
       "   12696      355.743323  \n",
       "   23909      342.389562  \n",
       "   24158      271.978647  \n",
       "   6572       247.958031  \n",
       "   20095      240.564118  \n",
       "   7106       234.725283  \n",
       "   1203       229.202680  \n",
       "   6369       227.933516  \n",
       "   23631      225.039974  \n",
       "   24647      219.763896  \n",
       "   1047       217.444235  \n",
       "   7354       215.340551  \n",
       "   20870      208.531429  \n",
       "   5527       207.514658  \n",
       "   1202       196.263596  \n",
       "   6665       193.520870  \n",
       "   6300       188.663684  \n",
       "   23583      187.676075  \n",
       "   21452      186.270813  \n",
       "   6102       186.270813  \n",
       "   11331      184.884133  \n",
       "   22828      182.768292  \n",
       "   24401      177.101900  \n",
       "   15982      173.644394  \n",
       "   932        170.827423  },\n",
       "  'adjusted_score': {'top_five':                                                  pathway     bin  \\\n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "   12315  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_9   \n",
       "   ...                                                  ...     ...   \n",
       "   6053                                       laminB1 over3  pc_1_4   \n",
       "   382                               Hashimoto, 1.2 fold up  pc_1_0   \n",
       "   381                                    Harvard home work  pc_1_0   \n",
       "   380        HYDROXYCHLOROQUINE-UP SARS-COV-2-UP-Cluster-1  pc_1_0   \n",
       "   379             HYDROXYCHLOROQUINE-UP SARS-COV-2-S7-1-UP  pc_1_0   \n",
       "   \n",
       "          adjusted_score  \n",
       "   11063    29196.217453  \n",
       "   22331     7185.850801  \n",
       "   9811      6120.227832  \n",
       "   21079     6021.470935  \n",
       "   12315     4750.949892  \n",
       "   ...               ...  \n",
       "   6053         0.000000  \n",
       "   382          0.000000  \n",
       "   381          0.000000  \n",
       "   380          0.000000  \n",
       "   379          0.000000  \n",
       "   \n",
       "   [100 rows x 3 columns],\n",
       "   'top_fifty':                                                  pathway     bin  \\\n",
       "   11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "   22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "   9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "   21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "   12315  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_9   \n",
       "   23909                            Basal 2 control Vs COPD  pc_2_9   \n",
       "   23195                                     SLE risk genes  pc_2_8   \n",
       "   23154  Rat proximal tubule TF network identified by wTO.  pc_2_8   \n",
       "   24158         HMGB1-interacted proteins pathway analysis  pc_2_9   \n",
       "   11331                                  ALX1 target genes  pc_1_9   \n",
       "   24647      common genes from iCLIP and RNA-seq data sets  pc_2_9   \n",
       "   8350                                       child obesity  pc_1_6   \n",
       "   11927                                     SLE risk genes  pc_1_9   \n",
       "   8559   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_6   \n",
       "   6487   Depending on gender, these genes can have vary...  pc_1_5   \n",
       "   23631                           mouse aging anti gomperz  pc_2_8   \n",
       "   24401  RPE reprogramming, dataset: 2185b5e32d0c3edf0c...  pc_2_9   \n",
       "   12295                                       inflammation  pc_1_9   \n",
       "   24209       Identified by the KroganLab in 2020 (nature)  pc_2_9   \n",
       "   11955                 Sofosbuvir-Down Sars-Cov-2-S7-1-Up  pc_1_9   \n",
       "   24372                           Prlr +/- from beta cells  pc_2_9   \n",
       "   23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "   22828                                        GE-K9-4-bed  pc_2_8   \n",
       "   24619                        capan1 downregulated p<0.05  pc_2_9   \n",
       "   11703                                            KHAZAYI  pc_1_9   \n",
       "   24677                                    diabetes type 1  pc_2_9   \n",
       "   24501  TNBC vs Normal Differential Gene Expression An...  pc_2_9   \n",
       "   10794                                          XP8GY-12H  pc_1_8   \n",
       "   12106                                      child obesity  pc_1_9   \n",
       "   22588                                      AD candidates  pc_2_8   \n",
       "   24132                                       Genes of IMF  pc_2_9   \n",
       "   24544         Up in vitamin C in endothelial cells human  pc_2_9   \n",
       "   11051                                   kc exo mass spec  pc_1_8   \n",
       "   22660  Biological insights from 108 schizophrenia-ass...  pc_2_8   \n",
       "   21452                                        CancerGenes  pc_2_7   \n",
       "   11918                                       SCD survival  pc_1_9   \n",
       "   10020  10 percent of the longest transcripts of the h...  pc_1_8   \n",
       "   11043                                       inflammation  pc_1_8   \n",
       "   24400  RPE reprogramming, dataset: 1642a68a2db75afc3f...  pc_2_9   \n",
       "   12414                       plts-Sil-tg only, common 1+4  pc_1_9   \n",
       "   24367                                       Pi Gene list  pc_2_9   \n",
       "   22158                                             damage  pc_2_7   \n",
       "   20095                                  ALX1 target genes  pc_2_6   \n",
       "   24633                                          clay dawn  pc_2_9   \n",
       "   24879                                         modulatity  pc_2_9   \n",
       "   24296                    ModularityInvasive and Stage2Ac  pc_2_9   \n",
       "   19827  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_5   \n",
       "   22864  Genes associated with osteonecrosis in ALL chi...  pc_2_8   \n",
       "   23040           Methylated-Diferentially Expressed Genes  pc_2_8   \n",
       "   23942                                   CHDI mouse liver  pc_2_9   \n",
       "   \n",
       "          adjusted_score  \n",
       "   11063    29196.217453  \n",
       "   22331     7185.850801  \n",
       "   9811      6120.227832  \n",
       "   21079     6021.470935  \n",
       "   12315     4750.949892  \n",
       "   23909     3273.073021  \n",
       "   23195     2880.320981  \n",
       "   23154     2766.612886  \n",
       "   24158     2599.979875  \n",
       "   11331     2430.301930  \n",
       "   24647     2100.832960  \n",
       "   8350      2040.534735  \n",
       "   11927     1927.710242  \n",
       "   8559      1913.093011  \n",
       "   6487      1728.067087  \n",
       "   23631     1706.365600  \n",
       "   24401     1693.005612  \n",
       "   12295     1639.581538  \n",
       "   24209     1576.678269  \n",
       "   11955     1456.985613  \n",
       "   24372     1444.809113  \n",
       "   23583     1423.053842  \n",
       "   22828     1385.840574  \n",
       "   24619     1265.565727  \n",
       "   11703     1240.747071  \n",
       "   24677     1223.444666  \n",
       "   24501     1112.628728  \n",
       "   10794     1109.892266  \n",
       "   12106     1095.598866  \n",
       "   22588     1093.411097  \n",
       "   24132     1089.956636  \n",
       "   24544     1064.114781  \n",
       "   11051     1054.297727  \n",
       "   22660     1046.979203  \n",
       "   21452     1044.141042  \n",
       "   11918     1008.013892  \n",
       "   10020     1006.771621  \n",
       "   11043      973.237161  \n",
       "   24400      940.791718  \n",
       "   12414      931.782116  \n",
       "   24367      931.482218  \n",
       "   22158      925.769618  \n",
       "   20095      872.886902  \n",
       "   24633      859.305326  \n",
       "   24879      854.306704  \n",
       "   24296      846.405380  \n",
       "   19827      836.160958  \n",
       "   22864      823.573727  \n",
       "   23040      803.040771  \n",
       "   23942      801.500779  }}}"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_pca_tops_dict = {}\n",
    "adj_score_lab = \"adjusted_score\"\n",
    "for lab in pc_gene_groups_dict.keys():\n",
    "    sorted_df = path_pca_OR_dict[lab].sort_values(by=OR_lab, ascending=False)\n",
    "    df_wide = sorted_df.pivot_table(index='pathway', columns='bin', values=OR_lab)\n",
    "    ind_list = df_wide.index\n",
    "    col_list = df_wide.columns\n",
    "    df_wide = pd.DataFrame(data = np.nan_to_num(df_wide),\n",
    "                           index=ind_list,\n",
    "                           columns = col_list)\n",
    "    df_long= pd.melt(df_wide, ignore_index=False, value_name=OR_lab)\n",
    "    df_long= df_long.reset_index()\n",
    "    df_long_sorted = df_long.sort_values(by=OR_lab, ascending = False)\n",
    "    top_fifty_byOR_df = df_long_sorted.head(50)\n",
    "    top_five_perclust_df = df_long_sorted.groupby(\"bin\").head(5)\n",
    "    OR_dict = {top_five_lab:top_five_perclust_df, top_fifty_lab:top_fifty_byOR_df}\n",
    "    sorted_df = path_pca_score_dict[lab].sort_values(by=score_lab, ascending=False)\n",
    "    df_wide = sorted_df.pivot_table(index='pathway', columns='bin', values=score_lab)\n",
    "    ind_list = df_wide.index\n",
    "    col_list = df_wide.columns\n",
    "    df_wide = pd.DataFrame(data = np.nan_to_num(df_wide),\n",
    "                           index=ind_list,\n",
    "                           columns = col_list)\n",
    "    df_long= pd.melt(df_wide, ignore_index=False, value_name=score_lab)\n",
    "    df_long= df_long.reset_index()\n",
    "    df_long_sorted = df_long.sort_values(by=score_lab, ascending = False)\n",
    "    top_fifty_byscore_df = df_long_sorted.head(50)\n",
    "    top_five_perclust_df = df_long_sorted.groupby(\"bin\").head(5)\n",
    "    score_dict = {top_five_lab:top_five_perclust_df, top_fifty_lab:top_fifty_byscore_df}\n",
    "    sorted_df = path_pca_adj_score_dict[lab].sort_values(by=adj_score_lab, ascending=False)\n",
    "    df_wide = sorted_df.pivot_table(index='pathway', columns='bin', values=adj_score_lab)\n",
    "    ind_list = df_wide.index\n",
    "    col_list = df_wide.columns\n",
    "    df_wide = pd.DataFrame(data = np.nan_to_num(df_wide),\n",
    "                           index=ind_list,\n",
    "                           columns = col_list)\n",
    "    df_long= pd.melt(df_wide, ignore_index=False, value_name=adj_score_lab)\n",
    "    df_long= df_long.reset_index()\n",
    "    df_long_sorted = df_long.sort_values(by=adj_score_lab, ascending = False)\n",
    "    top_fifty_byscore_df = df_long_sorted.head(50)\n",
    "    top_five_perclust_df = df_long_sorted.groupby(\"bin\").head(5)\n",
    "    adj_score_dict = {top_five_lab:top_five_perclust_df, top_fifty_lab:top_fifty_byscore_df}\n",
    "    path_pca_tops_dict[lab] = {OR_lab:OR_dict, score_lab:score_dict, adj_score_lab: adj_score_dict}\n",
    "path_pca_tops_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>bin</th>\n",
       "      <th>adjusted_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11063</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_1_8</td>\n",
       "      <td>29196.217453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22331</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_2_7</td>\n",
       "      <td>7185.850801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9811</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_1_7</td>\n",
       "      <td>6120.227832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21079</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_2_6</td>\n",
       "      <td>6021.470935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12315</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>4750.949892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23909</th>\n",
       "      <td>Basal 2 control Vs COPD</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>3273.073021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23195</th>\n",
       "      <td>SLE risk genes</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>2880.320981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23154</th>\n",
       "      <td>Rat proximal tubule TF network identified by wTO.</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>2766.612886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24158</th>\n",
       "      <td>HMGB1-interacted proteins pathway analysis</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>2599.979875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11331</th>\n",
       "      <td>ALX1 target genes</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>2430.301930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24647</th>\n",
       "      <td>common genes from iCLIP and RNA-seq data sets</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>2100.832960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8350</th>\n",
       "      <td>child obesity</td>\n",
       "      <td>pc_1_6</td>\n",
       "      <td>2040.534735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11927</th>\n",
       "      <td>SLE risk genes</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>1927.710242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8559</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_1_6</td>\n",
       "      <td>1913.093011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6487</th>\n",
       "      <td>Depending on gender, these genes can have vary...</td>\n",
       "      <td>pc_1_5</td>\n",
       "      <td>1728.067087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23631</th>\n",
       "      <td>mouse aging anti gomperz</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>1706.365600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24401</th>\n",
       "      <td>RPE reprogramming, dataset: 2185b5e32d0c3edf0c...</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>1693.005612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12295</th>\n",
       "      <td>inflammation</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>1639.581538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24209</th>\n",
       "      <td>Identified by the KroganLab in 2020 (nature)</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>1576.678269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11955</th>\n",
       "      <td>Sofosbuvir-Down Sars-Cov-2-S7-1-Up</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>1456.985613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24372</th>\n",
       "      <td>Prlr +/- from beta cells</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>1444.809113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23583</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>1423.053842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22828</th>\n",
       "      <td>GE-K9-4-bed</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>1385.840574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24619</th>\n",
       "      <td>capan1 downregulated p&lt;0.05</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>1265.565727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11703</th>\n",
       "      <td>KHAZAYI</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>1240.747071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24677</th>\n",
       "      <td>diabetes type 1</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>1223.444666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24501</th>\n",
       "      <td>TNBC vs Normal Differential Gene Expression An...</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>1112.628728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10794</th>\n",
       "      <td>XP8GY-12H</td>\n",
       "      <td>pc_1_8</td>\n",
       "      <td>1109.892266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12106</th>\n",
       "      <td>child obesity</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>1095.598866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22588</th>\n",
       "      <td>AD candidates</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>1093.411097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24132</th>\n",
       "      <td>Genes of IMF</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>1089.956636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24544</th>\n",
       "      <td>Up in vitamin C in endothelial cells human</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>1064.114781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11051</th>\n",
       "      <td>kc exo mass spec</td>\n",
       "      <td>pc_1_8</td>\n",
       "      <td>1054.297727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22660</th>\n",
       "      <td>Biological insights from 108 schizophrenia-ass...</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>1046.979203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21452</th>\n",
       "      <td>CancerGenes</td>\n",
       "      <td>pc_2_7</td>\n",
       "      <td>1044.141042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11918</th>\n",
       "      <td>SCD survival</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>1008.013892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10020</th>\n",
       "      <td>10 percent of the longest transcripts of the h...</td>\n",
       "      <td>pc_1_8</td>\n",
       "      <td>1006.771621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11043</th>\n",
       "      <td>inflammation</td>\n",
       "      <td>pc_1_8</td>\n",
       "      <td>973.237161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24400</th>\n",
       "      <td>RPE reprogramming, dataset: 1642a68a2db75afc3f...</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>940.791718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12414</th>\n",
       "      <td>plts-Sil-tg only, common 1+4</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>931.782116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24367</th>\n",
       "      <td>Pi Gene list</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>931.482218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22158</th>\n",
       "      <td>damage</td>\n",
       "      <td>pc_2_7</td>\n",
       "      <td>925.769618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20095</th>\n",
       "      <td>ALX1 target genes</td>\n",
       "      <td>pc_2_6</td>\n",
       "      <td>872.886902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24633</th>\n",
       "      <td>clay dawn</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>859.305326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24879</th>\n",
       "      <td>modulatity</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>854.306704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24296</th>\n",
       "      <td>ModularityInvasive and Stage2Ac</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>846.405380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19827</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_2_5</td>\n",
       "      <td>836.160958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22864</th>\n",
       "      <td>Genes associated with osteonecrosis in ALL chi...</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>823.573727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23040</th>\n",
       "      <td>Methylated-Diferentially Expressed Genes</td>\n",
       "      <td>pc_2_8</td>\n",
       "      <td>803.040771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23942</th>\n",
       "      <td>CHDI mouse liver</td>\n",
       "      <td>pc_2_9</td>\n",
       "      <td>801.500779</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 pathway     bin  \\\n",
       "11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "12315  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_9   \n",
       "23909                            Basal 2 control Vs COPD  pc_2_9   \n",
       "23195                                     SLE risk genes  pc_2_8   \n",
       "23154  Rat proximal tubule TF network identified by wTO.  pc_2_8   \n",
       "24158         HMGB1-interacted proteins pathway analysis  pc_2_9   \n",
       "11331                                  ALX1 target genes  pc_1_9   \n",
       "24647      common genes from iCLIP and RNA-seq data sets  pc_2_9   \n",
       "8350                                       child obesity  pc_1_6   \n",
       "11927                                     SLE risk genes  pc_1_9   \n",
       "8559   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_6   \n",
       "6487   Depending on gender, these genes can have vary...  pc_1_5   \n",
       "23631                           mouse aging anti gomperz  pc_2_8   \n",
       "24401  RPE reprogramming, dataset: 2185b5e32d0c3edf0c...  pc_2_9   \n",
       "12295                                       inflammation  pc_1_9   \n",
       "24209       Identified by the KroganLab in 2020 (nature)  pc_2_9   \n",
       "11955                 Sofosbuvir-Down Sars-Cov-2-S7-1-Up  pc_1_9   \n",
       "24372                           Prlr +/- from beta cells  pc_2_9   \n",
       "23583  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_8   \n",
       "22828                                        GE-K9-4-bed  pc_2_8   \n",
       "24619                        capan1 downregulated p<0.05  pc_2_9   \n",
       "11703                                            KHAZAYI  pc_1_9   \n",
       "24677                                    diabetes type 1  pc_2_9   \n",
       "24501  TNBC vs Normal Differential Gene Expression An...  pc_2_9   \n",
       "10794                                          XP8GY-12H  pc_1_8   \n",
       "12106                                      child obesity  pc_1_9   \n",
       "22588                                      AD candidates  pc_2_8   \n",
       "24132                                       Genes of IMF  pc_2_9   \n",
       "24544         Up in vitamin C in endothelial cells human  pc_2_9   \n",
       "11051                                   kc exo mass spec  pc_1_8   \n",
       "22660  Biological insights from 108 schizophrenia-ass...  pc_2_8   \n",
       "21452                                        CancerGenes  pc_2_7   \n",
       "11918                                       SCD survival  pc_1_9   \n",
       "10020  10 percent of the longest transcripts of the h...  pc_1_8   \n",
       "11043                                       inflammation  pc_1_8   \n",
       "24400  RPE reprogramming, dataset: 1642a68a2db75afc3f...  pc_2_9   \n",
       "12414                       plts-Sil-tg only, common 1+4  pc_1_9   \n",
       "24367                                       Pi Gene list  pc_2_9   \n",
       "22158                                             damage  pc_2_7   \n",
       "20095                                  ALX1 target genes  pc_2_6   \n",
       "24633                                          clay dawn  pc_2_9   \n",
       "24879                                         modulatity  pc_2_9   \n",
       "24296                    ModularityInvasive and Stage2Ac  pc_2_9   \n",
       "19827  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_5   \n",
       "22864  Genes associated with osteonecrosis in ALL chi...  pc_2_8   \n",
       "23040           Methylated-Diferentially Expressed Genes  pc_2_8   \n",
       "23942                                   CHDI mouse liver  pc_2_9   \n",
       "\n",
       "       adjusted_score  \n",
       "11063    29196.217453  \n",
       "22331     7185.850801  \n",
       "9811      6120.227832  \n",
       "21079     6021.470935  \n",
       "12315     4750.949892  \n",
       "23909     3273.073021  \n",
       "23195     2880.320981  \n",
       "23154     2766.612886  \n",
       "24158     2599.979875  \n",
       "11331     2430.301930  \n",
       "24647     2100.832960  \n",
       "8350      2040.534735  \n",
       "11927     1927.710242  \n",
       "8559      1913.093011  \n",
       "6487      1728.067087  \n",
       "23631     1706.365600  \n",
       "24401     1693.005612  \n",
       "12295     1639.581538  \n",
       "24209     1576.678269  \n",
       "11955     1456.985613  \n",
       "24372     1444.809113  \n",
       "23583     1423.053842  \n",
       "22828     1385.840574  \n",
       "24619     1265.565727  \n",
       "11703     1240.747071  \n",
       "24677     1223.444666  \n",
       "24501     1112.628728  \n",
       "10794     1109.892266  \n",
       "12106     1095.598866  \n",
       "22588     1093.411097  \n",
       "24132     1089.956636  \n",
       "24544     1064.114781  \n",
       "11051     1054.297727  \n",
       "22660     1046.979203  \n",
       "21452     1044.141042  \n",
       "11918     1008.013892  \n",
       "10020     1006.771621  \n",
       "11043      973.237161  \n",
       "24400      940.791718  \n",
       "12414      931.782116  \n",
       "24367      931.482218  \n",
       "22158      925.769618  \n",
       "20095      872.886902  \n",
       "24633      859.305326  \n",
       "24879      854.306704  \n",
       "24296      846.405380  \n",
       "19827      836.160958  \n",
       "22864      823.573727  \n",
       "23040      803.040771  \n",
       "23942      801.500779  "
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_fifty_byscore_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>bin</th>\n",
       "      <th>adjusted_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11063</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_1_8</td>\n",
       "      <td>29196.217453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22331</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_2_7</td>\n",
       "      <td>7185.850801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9811</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_1_7</td>\n",
       "      <td>6120.227832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21079</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_2_6</td>\n",
       "      <td>6021.470935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12315</th>\n",
       "      <td>lead SNPs from Neale v2 analysis of UKBB data,...</td>\n",
       "      <td>pc_1_9</td>\n",
       "      <td>4750.949892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6053</th>\n",
       "      <td>laminB1 over3</td>\n",
       "      <td>pc_1_4</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>Hashimoto, 1.2 fold up</td>\n",
       "      <td>pc_1_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>Harvard home work</td>\n",
       "      <td>pc_1_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>HYDROXYCHLOROQUINE-UP SARS-COV-2-UP-Cluster-1</td>\n",
       "      <td>pc_1_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>HYDROXYCHLOROQUINE-UP SARS-COV-2-S7-1-UP</td>\n",
       "      <td>pc_1_0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 pathway     bin  \\\n",
       "11063  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_8   \n",
       "22331  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_7   \n",
       "9811   lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_7   \n",
       "21079  lead SNPs from Neale v2 analysis of UKBB data,...  pc_2_6   \n",
       "12315  lead SNPs from Neale v2 analysis of UKBB data,...  pc_1_9   \n",
       "...                                                  ...     ...   \n",
       "6053                                       laminB1 over3  pc_1_4   \n",
       "382                               Hashimoto, 1.2 fold up  pc_1_0   \n",
       "381                                    Harvard home work  pc_1_0   \n",
       "380        HYDROXYCHLOROQUINE-UP SARS-COV-2-UP-Cluster-1  pc_1_0   \n",
       "379             HYDROXYCHLOROQUINE-UP SARS-COV-2-S7-1-UP  pc_1_0   \n",
       "\n",
       "       adjusted_score  \n",
       "11063    29196.217453  \n",
       "22331     7185.850801  \n",
       "9811      6120.227832  \n",
       "21079     6021.470935  \n",
       "12315     4750.949892  \n",
       "...               ...  \n",
       "6053         0.000000  \n",
       "382          0.000000  \n",
       "381          0.000000  \n",
       "380          0.000000  \n",
       "379          0.000000  \n",
       "\n",
       "[100 rows x 3 columns]"
      ]
     },
     "execution_count": 266,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_five_perclust_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Heatmap plots for comparing cluster pathway association."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "def chart_cluster_pathway(data_array, x_lab, y_lab, z_lab, title_str, text_precision = \".0f\"):\n",
    "    # Convert this grid to columnar data expected by Altair\n",
    "    title = alt.TitleParams(title_str, anchor='middle')\n",
    "    chart = alt.Chart(data_array, title = title).mark_rect().encode(\n",
    "        alt.X(x_lab+\":O\").title(x_lab),\n",
    "        alt.Y(y_lab+\":O\").title(y_lab),\n",
    "        alt.Color(z_lab+\":Q\").title(z_lab)\n",
    "        ).properties(\n",
    "            width=400,\n",
    "            height=400\n",
    "        )\n",
    "    text = chart.mark_text(baseline='middle').encode(\n",
    "    alt.Text(z_lab+':Q', format = text_precision),\n",
    "    color=alt.value('black')\n",
    ")\n",
    "    return(chart+text)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot the top five for each cluster and the top fifty pathways overall. For Odds-ratio and combined-score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "  #altair-viz-49aaf0d643724edeb76b904225feae65.vega-embed {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "  }\n",
       "\n",
       "  #altair-viz-49aaf0d643724edeb76b904225feae65.vega-embed details,\n",
       "  #altair-viz-49aaf0d643724edeb76b904225feae65.vega-embed details summary {\n",
       "    position: relative;\n",
       "  }\n",
       "</style>\n",
       "<div id=\"altair-viz-49aaf0d643724edeb76b904225feae65\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-49aaf0d643724edeb76b904225feae65\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-49aaf0d643724edeb76b904225feae65\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}}, \"layer\": [{\"mark\": {\"type\": \"rect\"}, \"encoding\": {\"color\": {\"field\": \"combined_score\", \"title\": \"combined_score\", \"type\": \"quantitative\"}, \"x\": {\"field\": \"cluster_number\", \"title\": \"cluster_number\", \"type\": \"ordinal\"}, \"y\": {\"field\": \"pathway\", \"title\": \"pathway\", \"type\": \"ordinal\"}}, \"title\": {\"text\": \"top_fifty pathways over all clusters for cluster method KmedoidsCosineThreeMinAICProposed\", \"anchor\": \"middle\"}}, {\"mark\": {\"type\": \"text\", \"baseline\": \"middle\"}, \"encoding\": {\"color\": {\"value\": \"black\"}, \"text\": {\"field\": \"combined_score\", \"format\": \".0f\", \"type\": \"quantitative\"}, \"x\": {\"field\": \"cluster_number\", \"title\": \"cluster_number\", \"type\": \"ordinal\"}, \"y\": {\"field\": \"pathway\", \"title\": \"pathway\", \"type\": \"ordinal\"}}, \"title\": {\"text\": \"top_fifty pathways over all clusters for cluster method KmedoidsCosineThreeMinAICProposed\", \"anchor\": \"middle\"}}], \"data\": {\"name\": \"data-1553c6a9a2e1ddd7b01cd4311a1319cb\"}, \"height\": 400, \"width\": 400, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.17.0.json\", \"datasets\": {\"data-1553c6a9a2e1ddd7b01cd4311a1319cb\": [{\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"cluster_number\": 1, \"combined_score\": 4196.71220978325}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"cluster_number\": 2, \"combined_score\": 1837.225580570573}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"cluster_number\": 0, \"combined_score\": 786.4797974365008}, {\"pathway\": \"child obesity\", \"cluster_number\": 2, \"combined_score\": 307.5632370311696}, {\"pathway\": \"cad&sleep gene list\", \"cluster_number\": 0, \"combined_score\": 154.44185858500512}, {\"pathway\": \"inflammation\", \"cluster_number\": 1, \"combined_score\": 145.17011298794554}, {\"pathway\": \"Genes Associated With Rare Diseases\", \"cluster_number\": 0, \"combined_score\": 113.61905990371194}, {\"pathway\": \"https://science.sciencemag.org/content/365/6456/eaat7693, dataset: faa3e04654acf5afe80cc00d5692f6e3\", \"cluster_number\": 0, \"combined_score\": 86.98846486773482}, {\"pathway\": \"The genes about major depressive disorder analysis.\", \"cluster_number\": 0, \"combined_score\": 77.74517881090729}, {\"pathway\": \"10 percent of the longest transcripts of the human genome\", \"cluster_number\": 1, \"combined_score\": 77.34044989792498}, {\"pathway\": \"genes closest to lead SNPs from Neale analysis of UKBB self reported hypertension data\", \"cluster_number\": 0, \"combined_score\": 76.18428517005925}, {\"pathway\": \"Genes closest to the lead SNPs from Ben Neale's v2 analysis of the UK biobank response to the Worrier question\", \"cluster_number\": 1, \"combined_score\": 75.06009633746545}, {\"pathway\": \"transposon\", \"cluster_number\": 0, \"combined_score\": 71.05961873492295}, {\"pathway\": \"mouse aging anti genage\", \"cluster_number\": 0, \"combined_score\": 67.86264560355225}, {\"pathway\": \"557gene litte list, dataset: c31f4fef2910f4f501a50c2946dd22d9\", \"cluster_number\": 1, \"combined_score\": 67.14583808377866}, {\"pathway\": \"TF OF H3.2C\", \"cluster_number\": 0, \"combined_score\": 63.14817193770837}, {\"pathway\": \"child obesity\", \"cluster_number\": 0, \"combined_score\": 63.14817193770837}, {\"pathway\": \"M2 POLARIZATION\", \"cluster_number\": 0, \"combined_score\": 54.56336312027448}, {\"pathway\": \"Genes closest to the lead SNPs from Ben Neale's v2 analysis of the UK biobank response to the Worrier question\", \"cluster_number\": 0, \"combined_score\": 53.60492452410381}, {\"pathway\": \"Looking for EDS and Schizophrenia linkers\", \"cluster_number\": 0, \"combined_score\": 53.60492452410381}, {\"pathway\": \"dehghan, dataset: a1271dcb68037de75ae77ffe8d628074\", \"cluster_number\": 0, \"combined_score\": 52.93128470784321}, {\"pathway\": \"Genes associated with T1D from the genomics component of TEDDY\", \"cluster_number\": 1, \"combined_score\": 50.914090121313045}, {\"pathway\": \"An ko mice\", \"cluster_number\": 0, \"combined_score\": 48.75760415302472}, {\"pathway\": \"GWAS AD GENES\", \"cluster_number\": 0, \"combined_score\": 48.20808001291636}, {\"pathway\": \"Enriched genes in SCLC compared with NSCLC among CCLE\", \"cluster_number\": 0, \"combined_score\": 46.23495322747936}, {\"pathway\": \"everolimus\", \"cluster_number\": 0, \"combined_score\": 46.23495322747936}, {\"pathway\": \"microRNA in tubercuosis\", \"cluster_number\": 0, \"combined_score\": 45.83014329116797}, {\"pathway\": \"brain pleio\", \"cluster_number\": 1, \"combined_score\": 40.84978853558983}, {\"pathway\": \"testis shotgun\", \"cluster_number\": 0, \"combined_score\": 40.43384010771743}, {\"pathway\": \"PCa signatures\", \"cluster_number\": 1, \"combined_score\": 40.21400796536459}, {\"pathway\": \"https://science.sciencemag.org/content/365/6456/eaat7693, dataset: 47a57d04565659137c0f26b29198bac4\", \"cluster_number\": 2, \"combined_score\": 38.982760588562776}, {\"pathway\": \"MIBEFRADIL-UP SARS-COV-2-S3-3-DOWN\", \"cluster_number\": 0, \"combined_score\": 37.966846321627926}, {\"pathway\": \"Asd 102 Satterstrom Cell\", \"cluster_number\": 2, \"combined_score\": 37.37455970798654}, {\"pathway\": \"genes closest to lead SNPs from Neale analysis of UKBB self reported hypertension data\", \"cluster_number\": 1, \"combined_score\": 36.44568848620342}, {\"pathway\": \"Pi Gene list\", \"cluster_number\": 0, \"combined_score\": 35.36672293621763}, {\"pathway\": \"Gdgtdrhyjgk\", \"cluster_number\": 1, \"combined_score\": 34.57240879920028}, {\"pathway\": \"genage all bedoone tekrari\", \"cluster_number\": 0, \"combined_score\": 34.527731324990114}, {\"pathway\": \"Achilles Genes with negative correlation (< -0.3) between CRISPR data and self-mRNA\", \"cluster_number\": 0, \"combined_score\": 34.527731324990114}, {\"pathway\": \"+E2, CtS +E2, si10 -E2, si10\", \"cluster_number\": 2, \"combined_score\": 34.254629257210226}, {\"pathway\": \"Nemo knockout\", \"cluster_number\": 0, \"combined_score\": 33.75754823262795}, {\"pathway\": \"autism confidence gene from SFARI\", \"cluster_number\": 1, \"combined_score\": 33.69659810928597}, {\"pathway\": \"Upregulated gene in Asthma\", \"cluster_number\": 0, \"combined_score\": 33.38536917576271}, {\"pathway\": \"SPMs and Splicing in Non-Obstructive CAD 300 DEGs in INOCA vs healthy\", \"cluster_number\": 0, \"combined_score\": 33.332088019358466}, {\"pathway\": \"Neale v2 UKBB GWAS, top SNPs selected by distance pruning with distance < 500kb and mapped to closest gene.\", \"cluster_number\": 0, \"combined_score\": 32.05010754891044}, {\"pathway\": \"plts-Sil-WTonly Common in 2+3\", \"cluster_number\": 0, \"combined_score\": 32.05010754891044}, {\"pathway\": \"Genes associated with osteonecrosis in ALL children\", \"cluster_number\": 0, \"combined_score\": 31.929484782564348}, {\"pathway\": \"hsa-miR-590-5p\", \"cluster_number\": 2, \"combined_score\": 31.775886574668068}, {\"pathway\": \"Pi Gene list\", \"cluster_number\": 1, \"combined_score\": 30.860918048217417}, {\"pathway\": \"plts, sil- wt2+3 only\", \"cluster_number\": 0, \"combined_score\": 30.79702707095102}, {\"pathway\": \"Am Not Sure Ture Or Not\", \"cluster_number\": 0, \"combined_score\": 30.260072343994267}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 268,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_lab = \"cluster_number\"\n",
    "y_lab = \"pathway\"\n",
    "tex_list = []\n",
    "for lab in cluster_method_labels:\n",
    "    lab = lab.replace(\"_\",\" \").title().replace(\" \",\"\")\n",
    "    title_str_per = '{} pathways per cluster for cluster method {}'\n",
    "    title_str_ove = '{} pathways over all clusters for cluster method {}'\n",
    "    save_str = path_plots+ \"/{}ClusterPaths{}{}.png\"\n",
    "    # OR plots \n",
    "    # - Top 5 per clust\n",
    "    top_five_or_title = title_str_per.format(top_five_lab,lab)\n",
    "    df = path_tops_dict[lab][OR_lab][top_five_lab]\n",
    "    chart = chart_cluster_pathway(df,x_lab, y_lab, OR_lab, top_five_or_title)\n",
    "    fname = save_str.format(top_five_lab, OR_lab,lab)\n",
    "    chart.save(fname)\n",
    "    tex_list +=[pathway_figure_tex(lab,  fname= fname, score = \"Odds Ratio\", filter = \"top five per cluster\")]\n",
    "    # - Top 50\n",
    "    top_fifty_or_title = title_str_ove.format(top_fifty_lab, lab)\n",
    "    chart = chart_cluster_pathway(path_tops_dict[lab][OR_lab][top_fifty_lab],x_lab, y_lab, OR_lab, top_fifty_or_title)\n",
    "    fname = save_str.format(top_fifty_lab, OR_lab, lab)\n",
    "    chart.save(fname)\n",
    "    tex_list +=[pathway_figure_tex(lab, fname= fname, score = \"Odds Ratio\", filter = \"top fifty overall\")]\n",
    "    # Combined score\n",
    "    # - Top 5 per clust\n",
    "    top_five_score_title = title_str_per.format(top_five_lab, lab)\n",
    "    chart = chart_cluster_pathway(path_tops_dict[lab][score_lab][top_five_lab],x_lab, y_lab, score_lab, top_five_score_title)\n",
    "    fname = save_str.format(top_five_lab, score_lab,lab)\n",
    "    chart.save(fname)\n",
    "    tex_list +=[pathway_figure_tex(lab,  fname= fname, score = \"Combined Score\", filter = \"top five per cluster\")]\n",
    "    # - Top 50\n",
    "    top_fifty_score_title = title_str_ove.format(top_fifty_lab, lab)\n",
    "    chart = chart_cluster_pathway(path_tops_dict[lab][score_lab][top_fifty_lab],x_lab, y_lab, score_lab, top_fifty_score_title)\n",
    "    fname= save_str.format(top_fifty_lab, score_lab,lab)\n",
    "    chart.save(fname)\n",
    "    tex_list +=[pathway_figure_tex(lab, fname= fname, score = \"Combined Score\", filter = \"top fifty overall\")]\n",
    "chart\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "  #altair-viz-9e3d51667418452fa5c01d2408647e0d.vega-embed {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "  }\n",
       "\n",
       "  #altair-viz-9e3d51667418452fa5c01d2408647e0d.vega-embed details,\n",
       "  #altair-viz-9e3d51667418452fa5c01d2408647e0d.vega-embed details summary {\n",
       "    position: relative;\n",
       "  }\n",
       "</style>\n",
       "<div id=\"altair-viz-9e3d51667418452fa5c01d2408647e0d\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-9e3d51667418452fa5c01d2408647e0d\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-9e3d51667418452fa5c01d2408647e0d\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}}, \"layer\": [{\"mark\": {\"type\": \"rect\"}, \"encoding\": {\"color\": {\"field\": \"adjusted_score\", \"title\": \"adjusted_score\", \"type\": \"quantitative\"}, \"x\": {\"field\": \"bin\", \"title\": \"bin\", \"type\": \"ordinal\"}, \"y\": {\"field\": \"pathway\", \"title\": \"pathway\", \"type\": \"ordinal\"}}, \"title\": {\"text\": \"top_fifty pathways per PCA bin for transform ours_with_exp\", \"anchor\": \"middle\"}}, {\"mark\": {\"type\": \"text\", \"baseline\": \"middle\"}, \"encoding\": {\"color\": {\"value\": \"black\"}, \"text\": {\"field\": \"adjusted_score\", \"format\": \".0f\", \"type\": \"quantitative\"}, \"x\": {\"field\": \"bin\", \"title\": \"bin\", \"type\": \"ordinal\"}, \"y\": {\"field\": \"pathway\", \"title\": \"pathway\", \"type\": \"ordinal\"}}, \"title\": {\"text\": \"top_fifty pathways per PCA bin for transform ours_with_exp\", \"anchor\": \"middle\"}}], \"data\": {\"name\": \"data-bcc3b8b9b8a2e9343e7642592bc7b6e9\"}, \"height\": 400, \"width\": 400, \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.17.0.json\", \"datasets\": {\"data-bcc3b8b9b8a2e9343e7642592bc7b6e9\": [{\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"bin\": \"pc_1_8\", \"adjusted_score\": 29196.217452526867}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"bin\": \"pc_2_7\", \"adjusted_score\": 7185.850800794302}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"bin\": \"pc_1_7\", \"adjusted_score\": 6120.227831948536}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"bin\": \"pc_2_6\", \"adjusted_score\": 6021.470935448252}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"bin\": \"pc_1_9\", \"adjusted_score\": 4750.949891715511}, {\"pathway\": \"Basal 2 control Vs COPD\", \"bin\": \"pc_2_9\", \"adjusted_score\": 3273.0730212320955}, {\"pathway\": \"SLE risk genes\", \"bin\": \"pc_2_8\", \"adjusted_score\": 2880.32098085159}, {\"pathway\": \"Rat proximal tubule TF network identified by wTO.\", \"bin\": \"pc_2_8\", \"adjusted_score\": 2766.612886293716}, {\"pathway\": \"HMGB1-interacted proteins pathway analysis\", \"bin\": \"pc_2_9\", \"adjusted_score\": 2599.9798748216904}, {\"pathway\": \"ALX1 target genes\", \"bin\": \"pc_1_9\", \"adjusted_score\": 2430.301929701677}, {\"pathway\": \"common genes from iCLIP and RNA-seq data sets\", \"bin\": \"pc_2_9\", \"adjusted_score\": 2100.832960258872}, {\"pathway\": \"child obesity\", \"bin\": \"pc_1_6\", \"adjusted_score\": 2040.5347347503123}, {\"pathway\": \"SLE risk genes\", \"bin\": \"pc_1_9\", \"adjusted_score\": 1927.7102418224065}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"bin\": \"pc_1_6\", \"adjusted_score\": 1913.0930112635015}, {\"pathway\": \"Depending on gender, these genes can have varying effects on body size\", \"bin\": \"pc_1_5\", \"adjusted_score\": 1728.0670871372397}, {\"pathway\": \"mouse aging anti gomperz\", \"bin\": \"pc_2_8\", \"adjusted_score\": 1706.3655999274886}, {\"pathway\": \"RPE reprogramming, dataset: 2185b5e32d0c3edf0ce0026127b45e7e\", \"bin\": \"pc_2_9\", \"adjusted_score\": 1693.0056123316358}, {\"pathway\": \"inflammation\", \"bin\": \"pc_1_9\", \"adjusted_score\": 1639.581538008398}, {\"pathway\": \"Identified by the KroganLab in 2020 (nature)\", \"bin\": \"pc_2_9\", \"adjusted_score\": 1576.6782694045164}, {\"pathway\": \"Sofosbuvir-Down Sars-Cov-2-S7-1-Up\", \"bin\": \"pc_1_9\", \"adjusted_score\": 1456.9856127225098}, {\"pathway\": \"Prlr +/- from beta cells\", \"bin\": \"pc_2_9\", \"adjusted_score\": 1444.8091128091264}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"bin\": \"pc_2_8\", \"adjusted_score\": 1423.053841608442}, {\"pathway\": \"GE-K9-4-bed\", \"bin\": \"pc_2_8\", \"adjusted_score\": 1385.8405738774604}, {\"pathway\": \"capan1 downregulated p<0.05\", \"bin\": \"pc_2_9\", \"adjusted_score\": 1265.565727233426}, {\"pathway\": \"KHAZAYI\", \"bin\": \"pc_1_9\", \"adjusted_score\": 1240.747071266027}, {\"pathway\": \"diabetes type 1\", \"bin\": \"pc_2_9\", \"adjusted_score\": 1223.4446664499092}, {\"pathway\": \"TNBC vs Normal Differential Gene Expression Analysis\", \"bin\": \"pc_2_9\", \"adjusted_score\": 1112.628727934676}, {\"pathway\": \"XP8GY-12H\", \"bin\": \"pc_1_8\", \"adjusted_score\": 1109.8922657270953}, {\"pathway\": \"child obesity\", \"bin\": \"pc_1_9\", \"adjusted_score\": 1095.5988660059168}, {\"pathway\": \"AD candidates\", \"bin\": \"pc_2_8\", \"adjusted_score\": 1093.4110968722978}, {\"pathway\": \"Genes of IMF\", \"bin\": \"pc_2_9\", \"adjusted_score\": 1089.956635533727}, {\"pathway\": \"Up in vitamin C in endothelial cells human\", \"bin\": \"pc_2_9\", \"adjusted_score\": 1064.1147810967263}, {\"pathway\": \"kc exo mass spec\", \"bin\": \"pc_1_8\", \"adjusted_score\": 1054.2977270840831}, {\"pathway\": \"Biological insights from 108 schizophrenia-associated genetic loci\", \"bin\": \"pc_2_8\", \"adjusted_score\": 1046.9792027788023}, {\"pathway\": \"CancerGenes\", \"bin\": \"pc_2_7\", \"adjusted_score\": 1044.1410420291377}, {\"pathway\": \"SCD survival\", \"bin\": \"pc_1_9\", \"adjusted_score\": 1008.0138923854535}, {\"pathway\": \"10 percent of the longest transcripts of the human genome\", \"bin\": \"pc_1_8\", \"adjusted_score\": 1006.7716213095556}, {\"pathway\": \"inflammation\", \"bin\": \"pc_1_8\", \"adjusted_score\": 973.2371611059298}, {\"pathway\": \"RPE reprogramming, dataset: 1642a68a2db75afc3f8b6e9b1ae0587e\", \"bin\": \"pc_2_9\", \"adjusted_score\": 940.7917183525644}, {\"pathway\": \"plts-Sil-tg only, common 1+4\", \"bin\": \"pc_1_9\", \"adjusted_score\": 931.7821158652412}, {\"pathway\": \"Pi Gene list\", \"bin\": \"pc_2_9\", \"adjusted_score\": 931.482218134112}, {\"pathway\": \"damage\", \"bin\": \"pc_2_7\", \"adjusted_score\": 925.7696176861544}, {\"pathway\": \"ALX1 target genes\", \"bin\": \"pc_2_6\", \"adjusted_score\": 872.8869021109456}, {\"pathway\": \"clay dawn\", \"bin\": \"pc_2_9\", \"adjusted_score\": 859.305325682976}, {\"pathway\": \"modulatity\", \"bin\": \"pc_2_9\", \"adjusted_score\": 854.306704161513}, {\"pathway\": \"ModularityInvasive and Stage2Ac\", \"bin\": \"pc_2_9\", \"adjusted_score\": 846.4053803508858}, {\"pathway\": \"lead SNPs from Neale v2 analysis of UKBB data, pruned with distance cutoff of 500kb and mapped to nearest gene\", \"bin\": \"pc_2_5\", \"adjusted_score\": 836.1609583724857}, {\"pathway\": \"Genes associated with osteonecrosis in ALL children\", \"bin\": \"pc_2_8\", \"adjusted_score\": 823.5737266711044}, {\"pathway\": \"Methylated-Diferentially Expressed Genes\", \"bin\": \"pc_2_8\", \"adjusted_score\": 803.0407707560108}, {\"pathway\": \"CHDI mouse liver\", \"bin\": \"pc_2_9\", \"adjusted_score\": 801.5007792883604}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_lab = \"bin\"\n",
    "y_lab = \"pathway\"\n",
    "for lab in path_pca_tops_dict.keys():\n",
    "    title_str = '{} pathways per PCA bin for transform {}'\n",
    "    save_str = path_plots+ \"/{}_pca_paths_{}_{}.png\"\n",
    "    # OR plots \n",
    "    # - Top 5 per clust\n",
    "    top_five_or_title = title_str.format(top_five_lab,lab)\n",
    "    df = path_pca_tops_dict[lab][OR_lab][top_five_lab]\n",
    "    chart = chart_cluster_pathway(df,x_lab, y_lab, OR_lab, top_five_or_title)\n",
    "    fname = save_str.format(top_five_lab, OR_lab, lab)\n",
    "    chart.save(fname)\n",
    "    # # - Top 50\n",
    "    top_fifty_or_title = title_str.format(top_fifty_lab, lab)\n",
    "    chart = chart_cluster_pathway(path_pca_tops_dict[lab][OR_lab][top_fifty_lab],x_lab, y_lab, OR_lab, top_fifty_or_title)\n",
    "    fname = save_str.format(top_fifty_lab, OR_lab, lab)\n",
    "    chart.save(fname)\n",
    "    # Combined score\n",
    "    # - Top 5 per clust\n",
    "    top_five_score_title = title_str.format(top_five_lab, lab)\n",
    "    chart = chart_cluster_pathway(path_pca_tops_dict[lab][score_lab][top_five_lab],x_lab, y_lab, score_lab, top_five_score_title)\n",
    "    fname = save_str.format(top_five_lab, score_lab,lab)\n",
    "    chart.save(fname)\n",
    "    # - Top 50\n",
    "    top_fifty_score_title = title_str.format(top_fifty_lab, lab)\n",
    "    chart = chart_cluster_pathway(path_pca_tops_dict[lab][score_lab][top_fifty_lab],x_lab, y_lab, score_lab, top_fifty_score_title)\n",
    "    fname = save_str.format(top_fifty_lab, score_lab,lab)\n",
    "    chart.save(fname)\n",
    "    # Adjusted Combined score\n",
    "    # - Top 5 per clust\n",
    "    top_five_score_title = title_str.format(top_five_lab, lab)\n",
    "    chart = chart_cluster_pathway(path_pca_tops_dict[lab][adj_score_lab][top_five_lab],x_lab, y_lab, adj_score_lab, top_five_score_title)\n",
    "    fname = save_str.format(top_five_lab, adj_score_lab,lab)\n",
    "    chart.save(fname)\n",
    "    # - Top 50\n",
    "    top_fifty_score_title = title_str.format(top_fifty_lab, lab)\n",
    "    chart = chart_cluster_pathway(path_pca_tops_dict[lab][adj_score_lab][top_fifty_lab],x_lab, y_lab, adj_score_lab, top_fifty_score_title)\n",
    "    fname = save_str.format(top_fifty_lab, adj_score_lab,lab)\n",
    "    chart.save(fname)\n",
    "chart\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparing the cluster methods\n",
    "In an ideal scenario our clustering method would:\n",
    "* identify pathways which have not been split accross other clusters. \n",
    "* separate the pathways from each other so they can be identified.\n",
    "\n",
    "In addition to this we also want to consider whether the pathways identified have significant enrichment and if so utilise this to inform further treatment investigation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How well has each cluster methods performed on not splitting clusters up?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ssd(A,B):\n",
    "  dif = A.ravel() - B.ravel()\n",
    "  return abs(dif).sum()/(len(A.ravel()))\n",
    "def uniqueness(df, axis = 0, score_lab = \"combined_score\"):\n",
    "    df_wide = df.pivot_table(index='pathway', columns='cluster_number', values=score_lab)\n",
    "    if df_wide.shape[1]==1 and axis ==1: return(\"NaN\")\n",
    "    mat = np.nan_to_num(df_wide.to_numpy())\n",
    "    if axis == 1:\n",
    "      mat = mat.T\n",
    "    mat_norm = (mat - mat.min(0))/ mat.ptp(0)\n",
    "    max_mat = np.zeros(mat_norm.shape)\n",
    "    max_mat[np.argmax(mat_norm, axis = 0)] = 1\n",
    "    score = ssd(max_mat, mat_norm)\n",
    "    return(score)\n",
    "def clust_path_score(df, score_lab = score_lab):\n",
    "  path_contain_score = uniqueness(df, axis = 0, score_lab=score_lab)\n",
    "  path_separate_score = uniqueness(df, axis = 1, score_lab=score_lab)\n",
    "  out_dict = {\"PathContaining\": path_contain_score,\n",
    "               \"PathSeparating\": path_separate_score}\n",
    "  return(out_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rescale_col (df, col_lab):\n",
    "  num_col = pd.to_numeric(df[col_lab], errors='coerce')\n",
    "  not_nan_rows = num_col.notnull()\n",
    "  num_col[not_nan_rows] = (num_col[not_nan_rows] - num_col[not_nan_rows].min())/ (num_col[not_nan_rows].max()- num_col[not_nan_rows].min())\n",
    "  return(num_col[not_nan_rows])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8     0.000000\n",
       "17    0.168617\n",
       "16    0.168617\n",
       "18    0.168617\n",
       "19    0.192361\n",
       "14    0.081227\n",
       "13    0.263454\n",
       "11    0.189508\n",
       "7     0.202920\n",
       "12    0.106824\n",
       "6     0.189003\n",
       "5     0.192186\n",
       "3     0.211330\n",
       "2     0.249573\n",
       "0     0.249573\n",
       "4     0.179412\n",
       "1     0.255066\n",
       "15    1.000000\n",
       "9     0.081976\n",
       "10    0.081976\n",
       "Name: PathContaining, dtype: float64"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clust_method_scores_df.PathContaining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PathContaining</th>\n",
       "      <th>PathSeparating</th>\n",
       "      <th>ClusterMethod</th>\n",
       "      <th>OverallPathwayIdentifying</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>DBScanZeroProposed</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.081391</td>\n",
       "      <td>0.452547</td>\n",
       "      <td>GMMOneProposed</td>\n",
       "      <td>0.118507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.168957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>BirchTwoProposed</td>\n",
       "      <td>0.152061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.168957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>BirchOneProposed</td>\n",
       "      <td>0.152061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.168957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>KmeansCosineTwoMinAICProposed</td>\n",
       "      <td>0.152061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.107039</td>\n",
       "      <td>0.602011</td>\n",
       "      <td>KmedoidsCosineSixProposed</td>\n",
       "      <td>0.156536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.192749</td>\n",
       "      <td>0.159654</td>\n",
       "      <td>KmedoidsCosineThreeMinAICProposed</td>\n",
       "      <td>0.189440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.189384</td>\n",
       "      <td>0.250035</td>\n",
       "      <td>KmedoidsEuclideanProposedFour</td>\n",
       "      <td>0.195449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.192573</td>\n",
       "      <td>0.230169</td>\n",
       "      <td>KmeansElkanStdFourProposed</td>\n",
       "      <td>0.196333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.192573</td>\n",
       "      <td>0.230169</td>\n",
       "      <td>KmeansLloydStdFourProposed</td>\n",
       "      <td>0.196333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.179774</td>\n",
       "      <td>0.468810</td>\n",
       "      <td>KmeansLloydStdSixProposed</td>\n",
       "      <td>0.208678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.189890</td>\n",
       "      <td>0.513248</td>\n",
       "      <td>KmeansCosineMinibatchSixProposed</td>\n",
       "      <td>0.222226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.203330</td>\n",
       "      <td>0.457834</td>\n",
       "      <td>KmeansCosineSixProposed</td>\n",
       "      <td>0.228780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.263985</td>\n",
       "      <td>0.115324</td>\n",
       "      <td>GMMZeroProposed</td>\n",
       "      <td>0.249119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.250076</td>\n",
       "      <td>0.286592</td>\n",
       "      <td>KmeansElkanStdFourDarrous</td>\n",
       "      <td>0.253728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.250076</td>\n",
       "      <td>0.286592</td>\n",
       "      <td>KmeansLloydStdFourDarrous</td>\n",
       "      <td>0.253728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.255581</td>\n",
       "      <td>0.476932</td>\n",
       "      <td>KmeansLloydStdSixDarrous</td>\n",
       "      <td>0.277716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>BirchZeroProposed</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.082141</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DBScanOneProposed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.082141</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DBScanTwoProposed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PathContaining  PathSeparating                      ClusterMethod  \\\n",
       "8         0.000000        0.000000                 DBScanZeroProposed   \n",
       "14        0.081391        0.452547                     GMMOneProposed   \n",
       "17        0.168957        0.000000                   BirchTwoProposed   \n",
       "16        0.168957        0.000000                   BirchOneProposed   \n",
       "18        0.168957        0.000000      KmeansCosineTwoMinAICProposed   \n",
       "12        0.107039        0.602011          KmedoidsCosineSixProposed   \n",
       "19        0.192749        0.159654  KmedoidsCosineThreeMinAICProposed   \n",
       "6         0.189384        0.250035      KmedoidsEuclideanProposedFour   \n",
       "5         0.192573        0.230169         KmeansElkanStdFourProposed   \n",
       "3         0.192573        0.230169         KmeansLloydStdFourProposed   \n",
       "4         0.179774        0.468810          KmeansLloydStdSixProposed   \n",
       "11        0.189890        0.513248   KmeansCosineMinibatchSixProposed   \n",
       "7         0.203330        0.457834            KmeansCosineSixProposed   \n",
       "13        0.263985        0.115324                    GMMZeroProposed   \n",
       "2         0.250076        0.286592          KmeansElkanStdFourDarrous   \n",
       "0         0.250076        0.286592          KmeansLloydStdFourDarrous   \n",
       "1         0.255581        0.476932           KmeansLloydStdSixDarrous   \n",
       "15        1.000000        1.000000                  BirchZeroProposed   \n",
       "9         0.082141             NaN                  DBScanOneProposed   \n",
       "10        0.082141             NaN                  DBScanTwoProposed   \n",
       "\n",
       "    OverallPathwayIdentifying  \n",
       "8                    0.000000  \n",
       "14                   0.118507  \n",
       "17                   0.152061  \n",
       "16                   0.152061  \n",
       "18                   0.152061  \n",
       "12                   0.156536  \n",
       "19                   0.189440  \n",
       "6                    0.195449  \n",
       "5                    0.196333  \n",
       "3                    0.196333  \n",
       "4                    0.208678  \n",
       "11                   0.222226  \n",
       "7                    0.228780  \n",
       "13                   0.249119  \n",
       "2                    0.253728  \n",
       "0                    0.253728  \n",
       "1                    0.277716  \n",
       "15                   1.000000  \n",
       "9                         NaN  \n",
       "10                        NaN  "
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clust_method_scores = []\n",
    "for lab in cluster_method_labels:\n",
    "    clust_path_dict = clust_path_score(path_clust_OR_dict[lab], score_lab=OR_lab)\n",
    "    clust_path_dict[\"ClusterMethod\"]=  lab\n",
    "    clust_method_scores.append(clust_path_dict)\n",
    "# Rescale the cluster scores.\n",
    "cont_weight = 0.90\n",
    "sep_weight = 0.1\n",
    "clust_method_scores_df = pd.DataFrame(clust_method_scores)\n",
    "clust_method_scores_df.PathContaining = rescale_col(clust_method_scores_df, \"PathContaining\")\n",
    "clust_method_scores_df.PathSeparating = rescale_col(clust_method_scores_df, \"PathSeparating\")\n",
    "clust_method_scores_df[\"OverallPathwayIdentifying\"] = cont_weight * clust_method_scores_df.PathContaining + sep_weight * clust_method_scores_df.PathSeparating\n",
    "clust_method_scores_df.sort_values(by=\"OverallPathwayIdentifying\", ascending=True, na_position=\"last\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Account for overfitting\n",
    "Divide the cluster pathway score by the `AIC`. This will adjust for methods which are overfitting clusters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PathContaining</th>\n",
       "      <th>PathSeparating</th>\n",
       "      <th>ClusterMethod</th>\n",
       "      <th>OverallPathwayIdentifying</th>\n",
       "      <th>aic</th>\n",
       "      <th>AdjustedPathContaining</th>\n",
       "      <th>AdjustedPathSeparating</th>\n",
       "      <th>AdjustedOverall</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>DBScanZeroProposed</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5698.360008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.168957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>BirchTwoProposed</td>\n",
       "      <td>0.152061</td>\n",
       "      <td>1959.091867</td>\n",
       "      <td>0.001250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.168957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>BirchOneProposed</td>\n",
       "      <td>0.152061</td>\n",
       "      <td>1959.091867</td>\n",
       "      <td>0.001250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.168957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>KmeansCosineTwoMinAICProposed</td>\n",
       "      <td>0.152061</td>\n",
       "      <td>2953.690816</td>\n",
       "      <td>0.001885</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.192749</td>\n",
       "      <td>0.159654</td>\n",
       "      <td>KmedoidsCosineThreeMinAICProposed</td>\n",
       "      <td>0.189440</td>\n",
       "      <td>6053.822311</td>\n",
       "      <td>0.004408</td>\n",
       "      <td>0.003651</td>\n",
       "      <td>0.004332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.081391</td>\n",
       "      <td>0.452547</td>\n",
       "      <td>GMMOneProposed</td>\n",
       "      <td>0.118507</td>\n",
       "      <td>9840.977985</td>\n",
       "      <td>0.003026</td>\n",
       "      <td>0.016824</td>\n",
       "      <td>0.004406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.263985</td>\n",
       "      <td>0.115324</td>\n",
       "      <td>GMMZeroProposed</td>\n",
       "      <td>0.249119</td>\n",
       "      <td>5537.238223</td>\n",
       "      <td>0.005522</td>\n",
       "      <td>0.002412</td>\n",
       "      <td>0.005211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.189890</td>\n",
       "      <td>0.513248</td>\n",
       "      <td>KmeansCosineMinibatchSixProposed</td>\n",
       "      <td>0.222226</td>\n",
       "      <td>6597.522864</td>\n",
       "      <td>0.004733</td>\n",
       "      <td>0.012792</td>\n",
       "      <td>0.005539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.203330</td>\n",
       "      <td>0.457834</td>\n",
       "      <td>KmeansCosineSixProposed</td>\n",
       "      <td>0.228780</td>\n",
       "      <td>7170.450067</td>\n",
       "      <td>0.005508</td>\n",
       "      <td>0.012402</td>\n",
       "      <td>0.006197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.107039</td>\n",
       "      <td>0.602011</td>\n",
       "      <td>KmedoidsCosineSixProposed</td>\n",
       "      <td>0.156536</td>\n",
       "      <td>10553.675863</td>\n",
       "      <td>0.004268</td>\n",
       "      <td>0.024002</td>\n",
       "      <td>0.006241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.189384</td>\n",
       "      <td>0.250035</td>\n",
       "      <td>KmedoidsEuclideanProposedFour</td>\n",
       "      <td>0.195449</td>\n",
       "      <td>14910.143710</td>\n",
       "      <td>0.010667</td>\n",
       "      <td>0.014084</td>\n",
       "      <td>0.011009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.192573</td>\n",
       "      <td>0.230169</td>\n",
       "      <td>KmeansElkanStdFourProposed</td>\n",
       "      <td>0.196333</td>\n",
       "      <td>47432.884363</td>\n",
       "      <td>0.034507</td>\n",
       "      <td>0.041244</td>\n",
       "      <td>0.035181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.192573</td>\n",
       "      <td>0.230169</td>\n",
       "      <td>KmeansLloydStdFourProposed</td>\n",
       "      <td>0.196333</td>\n",
       "      <td>47432.884363</td>\n",
       "      <td>0.034507</td>\n",
       "      <td>0.041244</td>\n",
       "      <td>0.035181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.250076</td>\n",
       "      <td>0.286592</td>\n",
       "      <td>KmeansElkanStdFourDarrous</td>\n",
       "      <td>0.253728</td>\n",
       "      <td>40169.834883</td>\n",
       "      <td>0.037949</td>\n",
       "      <td>0.043491</td>\n",
       "      <td>0.038504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.250076</td>\n",
       "      <td>0.286592</td>\n",
       "      <td>KmeansLloydStdFourDarrous</td>\n",
       "      <td>0.253728</td>\n",
       "      <td>40169.834883</td>\n",
       "      <td>0.037949</td>\n",
       "      <td>0.043491</td>\n",
       "      <td>0.038504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.179774</td>\n",
       "      <td>0.468810</td>\n",
       "      <td>KmeansLloydStdSixProposed</td>\n",
       "      <td>0.208678</td>\n",
       "      <td>49809.327363</td>\n",
       "      <td>0.033828</td>\n",
       "      <td>0.088215</td>\n",
       "      <td>0.039266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.255581</td>\n",
       "      <td>0.476932</td>\n",
       "      <td>KmeansLloydStdSixDarrous</td>\n",
       "      <td>0.277716</td>\n",
       "      <td>40016.885845</td>\n",
       "      <td>0.038637</td>\n",
       "      <td>0.072100</td>\n",
       "      <td>0.041983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>BirchZeroProposed</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>264708.032968</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.082141</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DBScanOneProposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3973.980002</td>\n",
       "      <td>0.001233</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.082141</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DBScanTwoProposed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3973.980002</td>\n",
       "      <td>0.001233</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PathContaining  PathSeparating                      ClusterMethod  \\\n",
       "8         0.000000        0.000000                 DBScanZeroProposed   \n",
       "17        0.168957        0.000000                   BirchTwoProposed   \n",
       "16        0.168957        0.000000                   BirchOneProposed   \n",
       "18        0.168957        0.000000      KmeansCosineTwoMinAICProposed   \n",
       "19        0.192749        0.159654  KmedoidsCosineThreeMinAICProposed   \n",
       "14        0.081391        0.452547                     GMMOneProposed   \n",
       "13        0.263985        0.115324                    GMMZeroProposed   \n",
       "11        0.189890        0.513248   KmeansCosineMinibatchSixProposed   \n",
       "7         0.203330        0.457834            KmeansCosineSixProposed   \n",
       "12        0.107039        0.602011          KmedoidsCosineSixProposed   \n",
       "6         0.189384        0.250035      KmedoidsEuclideanProposedFour   \n",
       "5         0.192573        0.230169         KmeansElkanStdFourProposed   \n",
       "3         0.192573        0.230169         KmeansLloydStdFourProposed   \n",
       "2         0.250076        0.286592          KmeansElkanStdFourDarrous   \n",
       "0         0.250076        0.286592          KmeansLloydStdFourDarrous   \n",
       "4         0.179774        0.468810          KmeansLloydStdSixProposed   \n",
       "1         0.255581        0.476932           KmeansLloydStdSixDarrous   \n",
       "15        1.000000        1.000000                  BirchZeroProposed   \n",
       "9         0.082141             NaN                  DBScanOneProposed   \n",
       "10        0.082141             NaN                  DBScanTwoProposed   \n",
       "\n",
       "    OverallPathwayIdentifying            aic  AdjustedPathContaining  \\\n",
       "8                    0.000000    5698.360008                0.000000   \n",
       "17                   0.152061    1959.091867                0.001250   \n",
       "16                   0.152061    1959.091867                0.001250   \n",
       "18                   0.152061    2953.690816                0.001885   \n",
       "19                   0.189440    6053.822311                0.004408   \n",
       "14                   0.118507    9840.977985                0.003026   \n",
       "13                   0.249119    5537.238223                0.005522   \n",
       "11                   0.222226    6597.522864                0.004733   \n",
       "7                    0.228780    7170.450067                0.005508   \n",
       "12                   0.156536   10553.675863                0.004268   \n",
       "6                    0.195449   14910.143710                0.010667   \n",
       "5                    0.196333   47432.884363                0.034507   \n",
       "3                    0.196333   47432.884363                0.034507   \n",
       "2                    0.253728   40169.834883                0.037949   \n",
       "0                    0.253728   40169.834883                0.037949   \n",
       "4                    0.208678   49809.327363                0.033828   \n",
       "1                    0.277716   40016.885845                0.038637   \n",
       "15                   1.000000  264708.032968                1.000000   \n",
       "9                         NaN    3973.980002                0.001233   \n",
       "10                        NaN    3973.980002                0.001233   \n",
       "\n",
       "    AdjustedPathSeparating  AdjustedOverall  \n",
       "8                 0.000000         0.000000  \n",
       "17                0.000000         0.001125  \n",
       "16                0.000000         0.001125  \n",
       "18                0.000000         0.001697  \n",
       "19                0.003651         0.004332  \n",
       "14                0.016824         0.004406  \n",
       "13                0.002412         0.005211  \n",
       "11                0.012792         0.005539  \n",
       "7                 0.012402         0.006197  \n",
       "12                0.024002         0.006241  \n",
       "6                 0.014084         0.011009  \n",
       "5                 0.041244         0.035181  \n",
       "3                 0.041244         0.035181  \n",
       "2                 0.043491         0.038504  \n",
       "0                 0.043491         0.038504  \n",
       "4                 0.088215         0.039266  \n",
       "1                 0.072100         0.041983  \n",
       "15                1.000000         1.000000  \n",
       "9                      NaN              NaN  \n",
       "10                     NaN              NaN  "
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clust_method_scores_df[\"aic\"] = [clustering_info.loc[lab, \"aic\"] for lab in clust_method_scores_df.ClusterMethod]\n",
    "clust_method_scores_df[\"AdjustedPathContaining\"] = clust_method_scores_df.PathContaining*clust_method_scores_df.aic\n",
    "clust_method_scores_df[\"AdjustedPathSeparating\"] = clust_method_scores_df.PathSeparating*clust_method_scores_df.aic\n",
    "clust_method_scores_df.AdjustedPathContaining = rescale_col(clust_method_scores_df, \"AdjustedPathContaining\")\n",
    "clust_method_scores_df.AdjustedPathSeparating = rescale_col(clust_method_scores_df, \"AdjustedPathSeparating\")\n",
    "clust_method_scores_df[\"AdjustedOverall\"] = cont_weight*clust_method_scores_df.AdjustedPathContaining + sep_weight*clust_method_scores_df.AdjustedPathSeparating\n",
    "\n",
    "clust_method_scores_df.sort_values(by=\"AdjustedOverall\", ascending=True, na_position=\"last\",inplace=True)\n",
    "clust_method_scores_df.to_csv(tex_dir+\"/ClusterPathwayComparions.csv\", \n",
    "                              float_format='%.3f')\n",
    "clust_method_scores_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(tex_dir+\"/cluster_pathway_figures_text.tex\", \"w\") as text_file:\n",
    "    for tex_item in tex_list:\n",
    "        text_file.write(tex_item)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
